
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Plotting functions in PIASO &#8212; PIASO 0.1.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css?v=2aa19091" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=0330b353" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/documentation_options.js?v=01f34227"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/01_showThePlotFunction';</script>
    <script src="../_static/sidebar_toggle.js?v=3b053ec6"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Built-in color palettes in PIASO" href="colorPalettes.html" />
    <link rel="prev" title="Finer clustering on selected cluster(s)" href="LeidenLocalTutorial.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/PIASO_logo.png" class="logo__image only-light" alt="PIASO 0.1.0 documentation - Home"/>
    <img src="../_static/PIASO_logo.png" class="logo__image only-dark pst-js-only" alt="PIASO 0.1.0 documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../installation.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../tutorials.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../modules.html">
    API
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../references.html">
    References
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../releaseNotes.html">
    Release notes
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/genecell/PIASO" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../installation.html">
    Installation
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="../tutorials.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../modules.html">
    API
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../references.html">
    References
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../releaseNotes.html">
    Release notes
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/genecell/PIASO" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<h3><a href="../index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../modules.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="../releaseNotes.html">Release notes</a></li>
</ul>
</div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../tutorials.html" class="nav-link">Tutorials</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Plotting functions in PIASO</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="Plotting-functions-in-PIASO">
<h1>Plotting functions in PIASO<a class="headerlink" href="#Plotting-functions-in-PIASO" title="Link to this heading">#</a></h1>
<p>This notebook will demonstrate the different plotting functions and their application to sample data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>path = &#39;/home/vas744/Analysis/Python/Packages/PIASO&#39;
import sys
sys.path.append(path)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import importlib
from scipy import sparse
import piaso
import scanpy as sc
import numpy as np
import matplotlib.pyplot as plt
import warnings
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.set_figure_params(dpi=80,dpi_save=300, color_map=&#39;viridis&#39;,facecolor=&#39;white&#39;)
from matplotlib import rcParams
rcParams[&#39;figure.figsize&#39;] = 4, 4
save_dir=&#39;/data1/vallaris/Results/single-cell/Methods/PIASO&#39;
sc.settings.figdir = save_dir
prefix=&#39;GitHub_testing&#39;
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>warnings.simplefilter(action=&#39;ignore&#39;, category=FutureWarning)
</pre></div>
</div>
</div>
<section id="Load-the-data">
<h2>Load the data<a class="headerlink" href="#Load-the-data" title="Link to this heading">#</a></h2>
<p>Let us load a 20k subsampled version of the Seattle Alzheimer’s Disease Brain Cell Atlas (SEA-AD) project dataset described in detail in <a class="reference external" href="../references.html#cell1">Gabitto et. al. (2024)</a>. We will be using the scRNA-seq data from the dataset in this tutorial.</p>
<p>Download the subsampled dataset from Google Drive: <a class="reference external" href="https://drive.google.com/file/d/1EdRA0ECvPlEnNaOzKqj19GrEtucB7hmE/view?usp=drive_link">https://drive.google.com/file/d/1EdRA0ECvPlEnNaOzKqj19GrEtucB7hmE/view?usp=drive_link</a></p>
<p>The original data is available on <a class="reference external" href="https://portal.brain-map.org/explore/seattle-alzheimers-disease">https://portal.brain-map.org/explore/seattle-alzheimers-disease</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>!/home/vas744/Software/gdrive files download --destination /home/vas744/Data/Public/PIASO 1EdRA0ECvPlEnNaOzKqj19GrEtucB7hmE
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata=sc.read(&#39;/n/scratch/users/v/vas744/Data/Public/PIASO/SEA-AD_RNA_MTG_subsample_excludeReference_20k_piaso.h5ad&#39;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 20000 × 36601
    obs: &#39;sample_id&#39;, &#39;Neurotypical reference&#39;, &#39;Donor ID&#39;, &#39;Organism&#39;, &#39;Brain Region&#39;, &#39;Sex&#39;, &#39;Gender&#39;, &#39;Age at Death&#39;, &#39;Race (choice=White)&#39;, &#39;Race (choice=Black/ African American)&#39;, &#39;Race (choice=Asian)&#39;, &#39;Race (choice=American Indian/ Alaska Native)&#39;, &#39;Race (choice=Native Hawaiian or Pacific Islander)&#39;, &#39;Race (choice=Unknown or unreported)&#39;, &#39;Race (choice=Other)&#39;, &#39;specify other race&#39;, &#39;Hispanic/Latino&#39;, &#39;Highest level of education&#39;, &#39;Years of education&#39;, &#39;PMI&#39;, &#39;Fresh Brain Weight&#39;, &#39;Brain pH&#39;, &#39;Overall AD neuropathological Change&#39;, &#39;Thal&#39;, &#39;Braak&#39;, &#39;CERAD score&#39;, &#39;Overall CAA Score&#39;, &#39;Highest Lewy Body Disease&#39;, &#39;Total Microinfarcts (not observed grossly)&#39;, &#39;Total microinfarcts in screening sections&#39;, &#39;Atherosclerosis&#39;, &#39;Arteriolosclerosis&#39;, &#39;LATE&#39;, &#39;Cognitive Status&#39;, &#39;Last CASI Score&#39;, &#39;Interval from last CASI in months&#39;, &#39;Last MMSE Score&#39;, &#39;Interval from last MMSE in months&#39;, &#39;Last MOCA Score&#39;, &#39;Interval from last MOCA in months&#39;, &#39;APOE Genotype&#39;, &#39;Primary Study Name&#39;, &#39;Secondary Study Name&#39;, &#39;NeuN positive fraction on FANS&#39;, &#39;RIN&#39;, &#39;cell_prep_type&#39;, &#39;facs_population_plan&#39;, &#39;rna_amplification&#39;, &#39;sample_name&#39;, &#39;sample_quantity_count&#39;, &#39;expc_cell_capture&#39;, &#39;method&#39;, &#39;pcr_cycles&#39;, &#39;percent_cdna_longer_than_400bp&#39;, &#39;rna_amplification_pass_fail&#39;, &#39;amplified_quantity_ng&#39;, &#39;load_name&#39;, &#39;library_prep&#39;, &#39;library_input_ng&#39;, &#39;r1_index&#39;, &#39;avg_size_bp&#39;, &#39;quantification_fmol&#39;, &#39;library_prep_pass_fail&#39;, &#39;exp_component_vendor_name&#39;, &#39;batch_vendor_name&#39;, &#39;experiment_component_failed&#39;, &#39;alignment&#39;, &#39;Genome&#39;, &#39;ar_id&#39;, &#39;bc&#39;, &#39;GEX_Estimated_number_of_cells&#39;, &#39;GEX_number_of_reads&#39;, &#39;GEX_sequencing_saturation&#39;, &#39;GEX_Mean_raw_reads_per_cell&#39;, &#39;GEX_Q30_bases_in_barcode&#39;, &#39;GEX_Q30_bases_in_read_2&#39;, &#39;GEX_Q30_bases_in_UMI&#39;, &#39;GEX_Percent_duplicates&#39;, &#39;GEX_Q30_bases_in_sample_index_i1&#39;, &#39;GEX_Q30_bases_in_sample_index_i2&#39;, &#39;GEX_Reads_with_TSO&#39;, &#39;GEX_Sequenced_read_pairs&#39;, &#39;GEX_Valid_UMIs&#39;, &#39;GEX_Valid_barcodes&#39;, &#39;GEX_Reads_mapped_to_genome&#39;, &#39;GEX_Reads_mapped_confidently_to_genome&#39;, &#39;GEX_Reads_mapped_confidently_to_intergenic_regions&#39;, &#39;GEX_Reads_mapped_confidently_to_intronic_regions&#39;, &#39;GEX_Reads_mapped_confidently_to_exonic_regions&#39;, &#39;GEX_Reads_mapped_confidently_to_transcriptome&#39;, &#39;GEX_Reads_mapped_antisense_to_gene&#39;, &#39;GEX_Fraction_of_transcriptomic_reads_in_cells&#39;, &#39;GEX_Total_genes_detected&#39;, &#39;GEX_Median_UMI_counts_per_cell&#39;, &#39;GEX_Median_genes_per_cell&#39;, &#39;Multiome_Feature_linkages_detected&#39;, &#39;Multiome_Linked_genes&#39;, &#39;Multiome_Linked_peaks&#39;, &#39;ATAC_Confidently_mapped_read_pairs&#39;, &#39;ATAC_Fraction_of_genome_in_peaks&#39;, &#39;ATAC_Fraction_of_high_quality_fragments_in_cells&#39;, &#39;ATAC_Fraction_of_high_quality_fragments_overlapping_TSS&#39;, &#39;ATAC_Fraction_of_high_quality_fragments_overlapping_peaks&#39;, &#39;ATAC_Fraction_of_transposition_events_in_peaks_in_cells&#39;, &#39;ATAC_Mean_raw_read_pairs_per_cell&#39;, &#39;ATAC_Median_high_quality_fragments_per_cell&#39;, &#39;ATAC_Non-nuclear_read_pairs&#39;, &#39;ATAC_Number_of_peaks&#39;, &#39;ATAC_Percent_duplicates&#39;, &#39;ATAC_Q30_bases_in_barcode&#39;, &#39;ATAC_Q30_bases_in_read_1&#39;, &#39;ATAC_Q30_bases_in_read_2&#39;, &#39;ATAC_Q30_bases_in_sample_index_i1&#39;, &#39;ATAC_Sequenced_read_pairs&#39;, &#39;ATAC_TSS_enrichment_score&#39;, &#39;ATAC_Unmapped_read_pairs&#39;, &#39;ATAC_Valid_barcodes&#39;, &#39;Number of mapped reads&#39;, &#39;Number of unmapped reads&#39;, &#39;Number of multimapped reads&#39;, &#39;Number of reads&#39;, &#39;Number of UMIs&#39;, &#39;Genes detected&#39;, &#39;Doublet score&#39;, &#39;Fraction mitochondrial UMIs&#39;, &#39;Used in analysis&#39;, &#39;Class confidence&#39;, &#39;Class&#39;, &#39;Subclass confidence&#39;, &#39;Subclass&#39;, &#39;Supertype confidence&#39;, &#39;Supertype (non-expanded)&#39;, &#39;Supertype&#39;, &#39;Continuous Pseudo-progression Score&#39;, &#39;Severely Affected Donor&#39;
    var: &#39;gene_ids&#39;
    uns: &#39;APOE4 Status_colors&#39;, &#39;Braak_colors&#39;, &#39;CERAD score_colors&#39;, &#39;Cognitive Status_colors&#39;, &#39;Great Apes Metadata&#39;, &#39;Highest Lewy Body Disease_colors&#39;, &#39;LATE_colors&#39;, &#39;Overall AD neuropathological Change_colors&#39;, &#39;Sex_colors&#39;, &#39;Subclass_colors&#39;, &#39;Supertype_colors&#39;, &#39;Thal_colors&#39;, &#39;UW Clinical Metadata&#39;, &#39;X_normalization&#39;, &#39;batch_condition&#39;, &#39;default_embedding&#39;, &#39;neighbors&#39;, &#39;title&#39;, &#39;umap&#39;
    obsm: &#39;X_scVI&#39;, &#39;X_umap&#39;
    layers: &#39;UMIs&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div></div>
</div>
</section>
<section id="Visualize-with-a-discrete-color-map">
<h2>Visualize with a discrete color map<a class="headerlink" href="#Visualize-with-a-discrete-color-map" title="Link to this heading">#</a></h2>
<p>The discrete color map can be used to visualize the categorical variables such as cell subclass.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.embedding(adata,
                basis=&#39;X_umap&#39;,
                color=[&#39;Subclass&#39;],
                palette=piaso.pl.color.d_color3,
                legend_fontoutline=2,
                legend_fontweight=5,
                cmap=&#39;Spectral_r&#39;,
                size=10,
                frameon=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_12_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_12_0.png" style="width: 522px; height: 284px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.embedding(adata,
                basis=&#39;X_umap&#39;,
                color=[&#39;Subclass&#39;],
                palette=piaso.pl.color.d_color3,
                legend_fontoutline=2,
                legend_fontsize=7,
                legend_fontweight=5,
                legend_loc=&#39;on data&#39;,
                cmap=&#39;Spectral_r&#39;,
                size=10,
                frameon=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_13_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_13_0.png" style="width: 265px; height: 277px;" />
</div>
</div>
</section>
<section id="Visualize-with-a-continuous-color-map">
<h2>Visualize with a continuous color map<a class="headerlink" href="#Visualize-with-a-continuous-color-map" title="Link to this heading">#</a></h2>
<p>The continuous color map can be used to visualize the continuous variables such as gene expression.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata,
           color=[&#39;GAD1&#39;, &#39;PVALB&#39;, &#39;SST&#39;, &#39;PDGFRA&#39;],
           cmap=piaso.pl.color.c_color1,
           legend_fontsize=10,
           legend_fontoutline=3,
           ncols=2,
           size=30,
           frameon=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_15_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_15_0.png" style="width: 645px; height: 567px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.umap(adata,
           color=[&#39;GAD1&#39;, &#39;PVALB&#39;, &#39;SST&#39;, &#39;PDGFRA&#39;],
           cmap=piaso.pl.color.c_color4,
           legend_fontsize=10,
           legend_fontoutline=3,
           ncols=2,
           size=30,
           frameon=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_16_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_16_0.png" style="width: 645px; height: 567px;" />
</div>
</div>
</section>
<section id="Split-the-UMAP-by-condition">
<h2>Split the UMAP by condition<a class="headerlink" href="#Split-the-UMAP-by-condition" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping_dict=dict(zip(adata.obs[&#39;CERAD score&#39;], adata.obs[&#39;CERAD score&#39;]))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping_dict={&#39;Absent&#39;: &#39;Absent&#39;,
 &#39;Sparse&#39;: &#39;Disease&#39;,
 &#39;Moderate&#39;: &#39;Disease&#39;,
 &#39;Frequent&#39;: &#39;Disease&#39;}
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&#39;Condition&#39;]=adata.obs[&#39;CERAD score&#39;].map(mapping_dict)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.embedding(adata,
                basis=&#39;X_umap&#39;,
                color=[&#39;Condition&#39;],
                palette=piaso.pl.color.d_color1,
                legend_fontoutline=2,
                legend_fontweight=5,
                cmap=&#39;Spectral_r&#39;,
                size=10,
                frameon=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_21_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_21_0.png" style="width: 348px; height: 277px;" />
</div>
</div>
<section id="Visualizing-the-expression-of-a-specific-gene-across-two-different-experimental-conditions">
<h3>Visualizing the expression of a specific gene across two different experimental conditions<a class="headerlink" href="#Visualizing-the-expression-of-a-specific-gene-across-two-different-experimental-conditions" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_embeddings_split(adata,
                               color=&#39;ADGRG7&#39;,
                               layer=None,
                               splitby=&#39;Condition&#39;,
                               color_map=piaso.pl.color.c_color1,
                               size=100,
                               frameon=False,
                               legend_loc=None,)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_23_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_23_0.png" style="width: 669px; height: 310px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_embeddings_split(adata,
                               color=&#39;FEZF2&#39;,
                               layer=None,
                               splitby=&#39;Condition&#39;,
                               color_map=piaso.pl.color.c_color1,
                               size=100,
                               frameon=False,
                               legend_loc=None,)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_24_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_24_0.png" style="width: 678px; height: 310px;" />
</div>
</div>
</section>
<section id="Visulaizing-continuous-variables-(Number-of-UMIs)-across-different-experimental-conditions">
<h3>Visulaizing continuous variables (Number of UMIs) across different experimental conditions<a class="headerlink" href="#Visulaizing-continuous-variables-(Number-of-UMIs)-across-different-experimental-conditions" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_embeddings_split(adata,
                               color=&#39;Number of UMIs&#39;,
                               layer=None,
                               splitby=&#39;Condition&#39;,
                               color_map=piaso.pl.color.c_color4,
                               size=100,
                               frameon=False,
                               legend_loc=None)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_26_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_26_0.png" style="width: 699px; height: 301px;" />
</div>
</div>
</section>
<section id="Visualizing-categorical-variables-(Cell-subclass)-across-different-experimental-conditions">
<h3>Visualizing categorical variables (Cell subclass) across different experimental conditions<a class="headerlink" href="#Visualizing-categorical-variables-(Cell-subclass)-across-different-experimental-conditions" title="Link to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_embeddings_split(adata,
                               color=&#39;Subclass&#39;,
                               layer=None,
                               splitby=&#39;Condition&#39;,
                               size=100,
                               frameon=False,)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_28_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_28_0.png" style="width: 1033px; height: 381px;" />
</div>
</div>
<p>We can visualize these cell subclasses alongside the subclass legend. To enhance clarity, the ncol parameter can be used to adjust the number of subplots per row.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_embeddings_split(adata,
                               color=&#39;Subclass&#39;,
                               layer=None,
                               splitby=&#39;Condition&#39;,
                               size=100,
                               frameon=False,
                               ncol=1)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_30_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_30_0.png" style="width: 639px; height: 781px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_embeddings_split(adata,
                               color=&#39;Subclass&#39;,
                               layer=None,
                               splitby=&#39;Condition&#39;,
                               size=100,
                               frameon=False,
                               legend_fontsize=7,
                               legend_fontoutline=2,
                               legend_fontweight=5,
                               legend_loc=&#39;on data&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_31_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_31_0.png" style="width: 640px; height: 319px;" />
</div>
</div>
</section>
</section>
<section id="Violin-plots-of-grouped-data">
<h2>Violin plots of grouped data<a class="headerlink" href="#Violin-plots-of-grouped-data" title="Link to this heading">#</a></h2>
<p>Visulaize how the groups in the dataset vary across a list of different features. Here, you can plot a different subplot for each feature of interest.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_features_violin(adata,
                              feature_list=[ &#39;CALB2&#39;, &#39;GAD2&#39;, &#39;SATB2&#39;, &#39;TBR1&#39;, &#39;Number of UMIs&#39;],
                              width_single=8,
                              height_single=2.3,
                              groupby=&#39;Subclass&#39;,
                              show_grid=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_33_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_33_0.png" style="width: 593px; height: 854px;" />
</div>
</div>
<p>Adding the horizontal grid line</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_features_violin(adata,
                              feature_list=[ &#39;CALB2&#39;, &#39;GAD2&#39;, &#39;SATB2&#39;, &#39;TBR1&#39;, &#39;Number of UMIs&#39;],
                              width_single=8,
                              height_single=2.3,
                              groupby=&#39;Subclass&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_35_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_35_0.png" style="width: 593px; height: 854px;" />
</div>
</div>
<p>The height and widhts of these plots can be adjusted according to the number of features and groups you have for better visualization.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_features_violin(adata,
                              feature_list=[ &#39;CALB2&#39;, &#39;GAD2&#39;, &#39;SATB2&#39;, &#39;TBR1&#39;, &#39;Number of UMIs&#39;],
                              width_single=20,
                              height_single=3,
                              groupby=&#39;Subclass&#39;,
                              show_grid=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_37_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_37_0.png" style="width: 1342px; height: 1067px;" />
</div>
</div>
</section>
<section id="Violin-plots-of-entire-data">
<h2>Violin plots of entire data<a class="headerlink" href="#Violin-plots-of-entire-data" title="Link to this heading">#</a></h2>
<p>We can visualize the entire dataset as a single group, adjusting the dot size to plot all the cells onto the violin plot.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_features_violin(adata,
                              feature_list=[ &#39;Genes detected&#39;],
                              width_single=4,
                              height_single=2.0,
                              size=1)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_40_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_40_0.png" style="width: 335px; height: 216px;" />
</div>
</div>
<p>We can visualize the distributions of the cells across various features together.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_features_violin(adata,
                              feature_list=[ &#39;MALAT1&#39;, &#39;GAD2&#39;, &#39;Number of UMIs&#39;],
                              width_single=4,
                              height_single=2.3,
                              size=0.5,
                              show_grid=False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_42_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_42_0.png" style="width: 344px; height: 514px;" />
</div>
</div>
</section>
<section id="Save-output-plots">
<h2>Save output plots<a class="headerlink" href="#Save-output-plots" title="Link to this heading">#</a></h2>
<p>Use the <code class="docutils literal notranslate"><span class="pre">save</span></code> parameter to specific the file name and path to save:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_embeddings_split(adata,
                               color=&#39;ADGRG7&#39;,
                               layer=None,
                               splitby=&#39;Condition&#39;,
                               color_map=piaso.pl.color.c_color1,
                               size=100,
                               frameon=False,
                               legend_loc=None,
                               save=&#39;./PIASO_UMAP_split_by_condition.pdf&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_44_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_44_0.png" style="width: 669px; height: 310px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Figure saved to:  ./PIASO_UMAP_split_by_condition.pdf
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pl.plot_features_violin(adata,
                              feature_list=[ &#39;CALB2&#39;, &#39;GAD2&#39;, &#39;SATB2&#39;, &#39;TBR1&#39;, &#39;Number of UMIs&#39;],
                              width_single=8,
                              height_single=2.3,
                              groupby=&#39;Subclass&#39;,
                              show_grid=False,
                              save=&#39;./violin_plot_piaso.pdf&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_45_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_45_0.png" style="width: 593px; height: 854px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Figure saved to: ./violin_plot_piaso.pdf
</pre></div></div>
</div>
</section>
<section id="Combine-categorical-variables">
<h2>Combine categorical variables<a class="headerlink" href="#Combine-categorical-variables" title="Link to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping_dict={&#39;Absent&#39;: &#39;Control&#39;,
 &#39;Sparse&#39;: &#39;Disease&#39;,
 &#39;Moderate&#39;: &#39;Disease&#39;,
 &#39;Frequent&#39;: &#39;Disease&#39;}
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&#39;Condition&#39;]=adata.obs[&#39;CERAD score&#39;].map(mapping_dict)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&#39;SubclassXCondition&#39;] = piaso.pp.getCrossCategories(adata.obs, &#39;Subclass&#39;, &#39;Condition&#39;, )
</pre></div>
</div>
</div>
<p>By combining multiple categories, while maintaining the original orders of the categories, you could show the gene expression changes at cell type level across different conditions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.dotplot(
    adata,
    var_names=[&#39;SERPINE1&#39;],
    groupby=&#39;SubclassXCondition&#39;,
    swap_axes=True,
    cmap=piaso.pl.color.c_color4
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_51_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_51_0.png" style="width: 1293px; height: 330px;" />
</div>
</div>
<p>We can also switch the order in which we combine the categorical variables.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs[&#39;ConditionXSubclass&#39;] = piaso.pp.getCrossCategories(adata.obs, &#39;Condition&#39;, &#39;Subclass&#39;,)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.dotplot(
    adata,
    var_names=[&#39;SERPINE1&#39;],
    groupby=&#39;ConditionXSubclass&#39;,
    swap_axes=True,
    cmap=piaso.pl.color.c_color4
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_01_showThePlotFunction_54_0.png" class="no-scaled-link" src="../_images/notebooks_01_showThePlotFunction_54_0.png" style="width: 1293px; height: 330px;" />
</div>
</div>
</section>
<section id="Count-the-number-of-cells-in-each-category">
<h2>Count the number of cells in each category<a class="headerlink" href="#Count-the-number-of-cells-in-each-category" title="Link to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pp.table(adata.obs[&#39;Subclass&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Microglia-PVM&#39;: 638,
 &#39;Lamp5&#39;: 653,
 &#39;OPC&#39;: 496,
 &#39;L5/6 NP&#39;: 299,
 &#39;Oligodendrocyte&#39;: 1663,
 &#39;L2/3 IT&#39;: 4826,
 &#39;Vip&#39;: 1551,
 &#39;Lamp5 Lhx6&#39;: 305,
 &#39;L5 IT&#39;: 1797,
 &#39;L4 IT&#39;: 2388,
 &#39;Pvalb&#39;: 1299,
 &#39;Astrocyte&#39;: 1080,
 &#39;L6 IT&#39;: 605,
 &#39;Sst&#39;: 780,
 &#39;L6 IT Car3&#39;: 354,
 &#39;VLMC&#39;: 86,
 &#39;L6b&#39;: 224,
 &#39;Pax6&#39;: 140,
 &#39;L6 CT&#39;: 231,
 &#39;Sncg&#39;: 329,
 &#39;L5 ET&#39;: 37,
 &#39;Chandelier&#39;: 157,
 &#39;Sst Chodl&#39;: 31,
 &#39;Endothelial&#39;: 31}
</pre></div></div>
</div>
<p>You can use the rank parameter to order the results, and the ascending parameter to determine the order of the ranking.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pp.table(adata.obs[&#39;Subclass&#39;], rank = True, ascending = True)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Sst Chodl&#39;: 31,
 &#39;Endothelial&#39;: 31,
 &#39;L5 ET&#39;: 37,
 &#39;VLMC&#39;: 86,
 &#39;Pax6&#39;: 140,
 &#39;Chandelier&#39;: 157,
 &#39;L6b&#39;: 224,
 &#39;L6 CT&#39;: 231,
 &#39;L5/6 NP&#39;: 299,
 &#39;Lamp5 Lhx6&#39;: 305,
 &#39;Sncg&#39;: 329,
 &#39;L6 IT Car3&#39;: 354,
 &#39;OPC&#39;: 496,
 &#39;L6 IT&#39;: 605,
 &#39;Microglia-PVM&#39;: 638,
 &#39;Lamp5&#39;: 653,
 &#39;Sst&#39;: 780,
 &#39;Astrocyte&#39;: 1080,
 &#39;Pvalb&#39;: 1299,
 &#39;Vip&#39;: 1551,
 &#39;Oligodendrocyte&#39;: 1663,
 &#39;L5 IT&#39;: 1797,
 &#39;L4 IT&#39;: 2388,
 &#39;L2/3 IT&#39;: 4826}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>piaso.pp.table(adata.obs[&#39;Subclass&#39;], rank = True, ascending = False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;L2/3 IT&#39;: 4826,
 &#39;L4 IT&#39;: 2388,
 &#39;L5 IT&#39;: 1797,
 &#39;Oligodendrocyte&#39;: 1663,
 &#39;Vip&#39;: 1551,
 &#39;Pvalb&#39;: 1299,
 &#39;Astrocyte&#39;: 1080,
 &#39;Sst&#39;: 780,
 &#39;Lamp5&#39;: 653,
 &#39;Microglia-PVM&#39;: 638,
 &#39;L6 IT&#39;: 605,
 &#39;OPC&#39;: 496,
 &#39;L6 IT Car3&#39;: 354,
 &#39;Sncg&#39;: 329,
 &#39;Lamp5 Lhx6&#39;: 305,
 &#39;L5/6 NP&#39;: 299,
 &#39;L6 CT&#39;: 231,
 &#39;L6b&#39;: 224,
 &#39;Chandelier&#39;: 157,
 &#39;Pax6&#39;: 140,
 &#39;VLMC&#39;: 86,
 &#39;L5 ET&#39;: 37,
 &#39;Sst Chodl&#39;: 31,
 &#39;Endothelial&#39;: 31}
</pre></div></div>
</div>
<p>In order to store the results as a dataframe, set as_dataframe = True.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>subclass_df = piaso.pp.table(adata.obs[&#39;Subclass&#39;], rank = True, ascending = False, as_dataframe = True)
</pre></div>
</div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="LeidenLocalTutorial.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Finer clustering on selected cluster(s)</p>
      </div>
    </a>
    <a class="right-next"
       href="colorPalettes.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Built-in color palettes in PIASO</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Load-the-data">Load the data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Visualize-with-a-discrete-color-map">Visualize with a discrete color map</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Visualize-with-a-continuous-color-map">Visualize with a continuous color map</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Split-the-UMAP-by-condition">Split the UMAP by condition</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Visualizing-the-expression-of-a-specific-gene-across-two-different-experimental-conditions">Visualizing the expression of a specific gene across two different experimental conditions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Visulaizing-continuous-variables-(Number-of-UMIs)-across-different-experimental-conditions">Visulaizing continuous variables (Number of UMIs) across different experimental conditions</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Visualizing-categorical-variables-(Cell-subclass)-across-different-experimental-conditions">Visualizing categorical variables (Cell subclass) across different experimental conditions</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Violin-plots-of-grouped-data">Violin plots of grouped data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Violin-plots-of-entire-data">Violin plots of entire data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Save-output-plots">Save output plots</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Combine-categorical-variables">Combine categorical variables</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Count-the-number-of-cells-in-each-category">Count the number of cells in each category</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Min Dai.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>