
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>piaso.tools package &#8212; PIASO 0.1.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css?v=0330b353" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=01f34227"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'piaso.tools';</script>
    <script src="_static/sidebar_toggle.js?v=3b053ec6"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="References" href="references.html" />
    <link rel="prev" title="piaso.preprocessing package" href="piaso.preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/PIASO_logo.png" class="logo__image only-light" alt="PIASO 0.1.0 documentation - Home"/>
    <img src="_static/PIASO_logo.png" class="logo__image only-dark pst-js-only" alt="PIASO 0.1.0 documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tutorials.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="modules.html">
    API
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="references.html">
    References
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="releaseNotes.html">
    Release notes
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/genecell/PIASO" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tutorials.html">
    Tutorials
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="modules.html">
    API
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="references.html">
    References
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="releaseNotes.html">
    Release notes
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/genecell/PIASO" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<h3><a href="index.html">Table of Contents</a></h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="releaseNotes.html">Release notes</a></li>
</ul>
</div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="modules.html" class="nav-link">API</a></li>
    
    
    <li class="breadcrumb-item"><a href="piaso.html" class="nav-link">piaso package</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">piaso.tools package</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="module-piaso.tools">
<span id="piaso-tools-package"></span><h1>piaso.tools package<a class="headerlink" href="#module-piaso.tools" title="Link to this heading">#</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runSVD">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runSVD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_iter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_svd'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runSVD" title="Link to this definition">#</a></dt>
<dd><p>Performs Truncated Singular Value Decomposition (SVD) on the specified gene expression matrix (adata.X or a specified layer)
within an AnnData object and stores the resulting low-dimensional representation in <cite>adata.obsm</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – An AnnData object.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – If True, the decomposition is performed only on highly variable genes/features.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=50</em>) – The number of principal components to retain.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=10</em>) – A random seed to ensure reproducibility.</p></li>
<li><p><strong>scale_data</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, standardizes the input data before performing SVD.</p></li>
<li><p><strong>n_iter</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=7</em>) – Number of iterations for randomized SVD solver. The default is larger than the default in randomized_svd to handle sparse matrices that may have large slowly decaying spectrum. Also larger than the <cite>n_iter</cite> default value (5) in the TruncatedSVD function.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='X_svd'</em>) – The key under which the resulting cell embeddings are stored in <cite>adata.obsm</cite>.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies which layer of <cite>adata</cite> to use for the transformation. If None, <cite>adata.X</cite> is used.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=0</em>) – Controls the verbosity of logging messages.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function modifies <cite>adata</cite> in place, storing the cell embeddings in <cite>adata.obsm[key_added]</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runSVD</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
<span class="gp">... </span>       <span class="n">scale_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;X_svd&#39;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the transformed data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_svd&#39;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runSVDLazy">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runSVDLazy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_iter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_trim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_svd'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runSVDLazy" title="Link to this definition">#</a></dt>
<dd><p>Performs Truncated Singular Value Decomposition (SVD) in a “lazy” mode, based on the <cite>piaso.tl.runSVD</cite> function.</p>
<p>Compared to <cite>piaso.tl.runSVD</cite>, this function includes the step of highly variable gene section. If <cite>layer</cite> is set to <cite>infog</cite>,
both the highly variable genes and normalized gene expression values were taken from the INFOG normalization outputs.</p>
<p>This function performs on the specified gene expression matrix (adata.X or a specified layer) within an AnnData object
and stores the resulting low-dimensional representation in <cite>adata.obsm</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – An AnnData object.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, returns a copy of <cite>adata</cite> with the computed embeddings instead of modifying in place.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=50</em>) – The number of singular value decomposition (SVD) components to retain.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – If True, uses only highly variable genes for the decomposition.</p></li>
<li><p><strong>n_top_genes</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=3000</em>) – The number of top highly variable genes to retain before performing SVD.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=0</em>) – Controls the verbosity of logging messages.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies the key in <cite>adata.obs</cite> containing batch labels for <cite>batch_key</cite> used in the <cite>sc.pp.highly_variable_genes</cite> function.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=1927</em>) – A random seed to ensure reproducibility.</p></li>
<li><p><strong>scale_data</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, standardizes the input data before performing SVD.</p></li>
<li><p><strong>n_iter</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=7</em>) – Number of iterations for randomized SVD solver. The default is larger than the default in randomized_svd to handle sparse matrices that may have large slowly decaying spectrum. Also larger than the <cite>n_iter</cite> default value (5) in the TruncatedSVD function.</p></li>
<li><p><strong>infog_trim</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – Used for the <cite>trim</cite> parameter in <cite>piaso.tl.infog</cite> function, effective only when <cite>layer</cite> set to <cite>infog</cite>.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='X_svd'</em>) – The key under which the resulting cell embeddings are stored in <cite>adata.obsm</cite>.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies which layer of <cite>adata</cite> to use for the transformation. If None, <cite>adata.X</cite> is used.</p></li>
<li><p><strong>infog_layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Used for the <cite>layer</cite> parameter in <cite>piaso.tl.infog</cite> function, effective only when <cite>layer</cite> set to <cite>infog</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>copy</cite> is True, returns a modified AnnData object. Otherwise, modifies <cite>adata</cite> in place.</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runSVDLazy</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_svd&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">layer</span><span class="o">=</span><span class="kc">None</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the cell embedding</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_svd&#39;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runGDR">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runGDR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_iter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoring_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runGDR" title="Link to this definition">#</a></dt>
<dd><p>Run GDR (marker Gene-guided dimensionality reduction) on single-cell data.</p>
<p>GDR performs dimensionality reduction guided by marker genes to better preserve biological signals.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obs</cite> representing batch information. Defaults to None. If provided, marker gene identifications will be performed for each batch separately.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obs</cite> to specify which cell group information to use. Defaults to None. If none, de novo clustering will be performed.</p></li>
<li><p><strong>n_gene</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of genes, parameter used in COSG. Defaults to 30.</p></li>
<li><p><strong>mu</strong> (<em>float</em><em>, </em><em>optional</em>) – Gene expression specificity parameter, used in COSG. Defaults to 1.0.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – Layer in <cite>adata.layers</cite> to use for the analysis. Defaults to None, which uses <cite>adata.X</cite>.</p></li>
<li><p><strong>score_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – If specified, the gene scoring will be calculated using this layer of <cite>adata.layers</cite>. Defaults to None.</p></li>
<li><p><strong>infog_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – If specified, INFOG normalization will be applied using this layer, which should
contain the raw UMI count matrix. Defaults to None.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use only highly variable genes when rerunning the dimensionality reduction. Defaults to True. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_highly_variable_genes</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of highly variable genes to use when <cite>use_highly_variable</cite> is True. Defaults to 5000. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_svd_dims</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of dimensions to use for SVD. Defaults to 50. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_svd_iter</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=7</em>) – Number of iterations for randomized SVD solver. The default is larger than the default in randomized_svd to handle sparse matrices that may have large slowly decaying spectrum. Also larger than the <cite>n_iter</cite> default value (5) in the TruncatedSVD function.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em>) – Resolution parameter for de novo clustering. Defaults to 1.0. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>scoring_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Specifies the gene set scoring method used to compute gene scores.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em>) – Key under which the GDR dimensionality reduction results will be stored in <cite>adata.obsm</cite>. If None, results will be saved to <cite>adata.obsm[X_gdr]</cite>.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em>) – Verbosity level of the function. Higher values provide more detailed logs. Defaults to 0.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function modifies <cite>adata</cite> in place by adding GDR dimensionality reduction result to <cite>adata.obsm[key_added]</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;example.h5ad&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runGDR</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;CellTypes&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_gene</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_gdr&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.calculateScoreParallel">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">calculateScoreParallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'scanpy'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'piaso'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'piaso'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_pvals</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.calculateScoreParallel" title="Link to this definition">#</a></dt>
<dd><p>Compute gene set scores in parallel using shared memory for efficiency.</p>
<p>This function processes multiple gene sets in parallel, computing enrichment scores
for each gene set across all cells in the AnnData object. It uses shared memory
to efficiently pass the expression matrix to worker processes, avoiding costly
serialization overhead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The input AnnData object containing gene expression data.</p></li>
<li><p><strong>gene_set</strong> (<em>dict</em><em>, </em><em>list</em><em> of </em><em>lists</em><em>, or </em><em>pandas.DataFrame</em>) – <dl class="simple">
<dt>A collection of gene sets to score. Supported formats:</dt><dd><ul>
<li><p>dict: Keys are gene set names, values are lists of gene names.</p></li>
<li><p>list of lists: Each sublist contains gene names for one gene set.
Gene sets will be named “GeneSet_0”, “GeneSet_1”, etc.</p></li>
<li><p>pandas.DataFrame: Each column represents a gene set, with column names
as gene set names and gene names as values.</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>score_method</strong> (<em>{'scanpy'</em><em>, </em><em>'piaso'}</em><em>, </em><em>default 'piaso'</em>) – <p>The method used for gene set scoring.
- ‘scanpy’: Uses Scanpy’s built-in gene set scoring method.
- ‘piaso’: Uses the PIASO’s gene set scoring method, which is more robust to sequencing</p>
<blockquote>
<div><p>depth variations and provides p-values.</p>
</div></blockquote>
</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>default 1927</em>) – Random seed for reproducibility.</p></li>
<li><p><strong>score_layer</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>default None</em>) – Layer of the AnnData object to use. If None, <cite>adata.X</cite> is used.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>default None</em>) – Number of parallel worker processes to use. If None, defaults to the number
of CPU cores available.</p></li>
<li><p><strong>return_pvals</strong> (<em>bool</em><em>, </em><em>default False</em>) – Whether to return -log10(p-values) when using ‘piaso’ method. Only applicable
when score_method=’piaso’. If True, returns a third array containing p-values.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>default 0</em>) – Level of verbosity for progress reporting.
- 0: Silent (no progress bar)
- &gt;0: Show progress bar during parallel computation</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>score_matrix</strong> (<em>np.ndarray</em>) – A 2D array of shape (n_cells, n_gene_sets) where each column contains
the scores for one gene set across all cells.</p></li>
<li><p><strong>gene_set_names</strong> (<em>list of str</em>) – The names of the gene sets, in the same order as columns in score_matrix.</p></li>
<li><p><strong>nlog10_pval_matrix</strong> (<em>np.ndarray, optional</em>) – Only returned when score_method=’piaso’ and return_pvals=True.
A 2D array of shape (n_cells, n_gene_sets) containing -log10(p-values)
for each gene set score. Returns None if p-values are not available.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load example data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc3k</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Define gene sets</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gene_sets</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>    <span class="s1">&#39;T_cell_markers&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD3D&#39;</span><span class="p">,</span> <span class="s1">&#39;CD3E&#39;</span><span class="p">,</span> <span class="s1">&#39;CD8A&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;B_cell_markers&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;CD79A&#39;</span><span class="p">,</span> <span class="s1">&#39;CD79B&#39;</span><span class="p">,</span> <span class="s1">&#39;MS4A1&#39;</span><span class="p">]</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Compute scores using Scanpy method</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">scores</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">calculateScoreParallel</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">gene_set</span><span class="o">=</span><span class="n">gene_sets</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">score_method</span><span class="o">=</span><span class="s1">&#39;piaso&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">1</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Add scores to AnnData object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">name</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">names</span><span class="p">):</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s1">_score&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">scores</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.calculateScoreParallel_multiBatch">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">calculateScoreParallel_multiBatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_gene_n_groups_indices</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'scanpy'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'piaso'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.calculateScoreParallel_multiBatch" title="Link to this definition">#</a></dt>
<dd><p>Calculate gene set scores for each adata batch in parallel using shared memory. Different marker gene sets will be calculated in parallel as well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The key in <cite>adata.obs</cite> used to identify batches.</p></li>
<li><p><strong>marker_gene</strong> (<em>DataFrame</em>) – The marker gene DataFrame.</p></li>
<li><p><strong>marker_gene_n_groups_indices</strong> (<em>list</em>) – Indices specifying the marker gene set group boundaries, used for score normalization within each marker gene set group.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em>) – Maximum number of parallel workers to use.</p></li>
<li><p><strong>score_layer</strong> (<em>str</em>) – The layer of <cite>adata</cite> to use for scoring.</p></li>
<li><p><strong>score_method</strong> (<em>{'scanpy'</em><em>, </em><em>'piaso'}</em><em>, </em><em>optional</em>) – The method used for gene set scoring. Must be either ‘scanpy’ (default) or ‘piaso’.
- ‘scanpy’: Uses the Scanpy’s built-in gene set scoring method.
- ‘piaso’: Uses the PIASO’s gene set scoring method, which is more robust to sequencing depth variations.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>list: A list of normalized score arrays for each batch.</p></li>
<li><p>list: A list of cell barcodes for each batch.</p></li>
<li><p>list: A list of gene set names.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;example_data.h5ad&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score_list</span><span class="p">,</span> <span class="n">cellbarcode_info</span><span class="p">,</span> <span class="n">gene_set_names</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">calculateScoreParallel_multiBatch</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">marker_gene</span><span class="o">=</span><span class="n">marker_gene</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">marker_gene_n_groups_indices</span><span class="o">=</span><span class="n">marker_gene_n_groups_indices</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">score_layer</span><span class="o">=</span><span class="s1">&#39;piaso&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">max_workers</span><span class="o">=</span><span class="mi">8</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cellbarcode_info</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runGDRParallel">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runGDRParallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_iter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoring_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calculate_score_multiBatch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runGDRParallel" title="Link to this definition">#</a></dt>
<dd><p>Run GDR (marker Gene-guided dimensionality reduction) in parallel using multi-cores and shared memeory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obs</cite> representing batch information. Defaults to None. If specified, different batches will be processed separately and in parallel, otherwise, the input data will be processed as one batch.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obs</cite> to specify which cell group information to use. Defaults to None. If none, de novo clustering will be performed.</p></li>
<li><p><strong>n_gene</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of genes, parameter used in COSG. Defaults to 30.</p></li>
<li><p><strong>mu</strong> (<em>float</em><em>, </em><em>optional</em>) – Gene expression specificity parameter, used in COSG. Defaults to 1.0.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – Layer in <cite>adata.layers</cite> to use for the analysis. Defaults to None, which uses <cite>adata.X</cite>.</p></li>
<li><p><strong>score_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – If specified, the gene scoring will be calculated using this layer of <cite>adata.layers</cite>. Defaults to None.</p></li>
<li><p><strong>infog_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – If specified, the INFOG normalization will be calculated using this layer of <cite>adata.layers</cite>, which is expected to contain the UMI count matrix. Defaults to None.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use only highly variable genes when rerunning the dimensionality reduction. Defaults to True. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_highly_variable_genes</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of highly variable genes to use when <cite>use_highly_variable</cite> is True. Defaults to 5000. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_svd_dims</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of dimensions to use for SVD. Defaults to 50. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_svd_iter</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=7</em>) – Number of iterations for randomized SVD solver. The default is larger than the default in randomized_svd to handle sparse matrices that may have large slowly decaying spectrum. Also larger than the <cite>n_iter</cite> default value (5) in the TruncatedSVD function.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em>) – Resolution parameter for de novo clustering. Defaults to 1.0. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>scoring_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Specifies the gene set scoring method used to compute gene scores. If set to None, use PIASO’s scoring method as default.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em>) – Key under which the GDR dimensionality reduction results will be stored in <cite>adata.obsm</cite>. If None, results will be saved to <cite>adata.obsm[X_gdr]</cite>.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of workers to use for parallel computation. Defaults to 8.</p></li>
<li><p><strong>calculate_score_multiBatch</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate gene scores across multiple adata batches (if <cite>batch_key</cite> is specified). Defaults to False.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em>) – Verbosity level of the function. Higher values provide more detailed logs. Defaults to 0.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function modifies <cite>adata</cite> in place by adding GDR dimensionality reduction result to <cite>adata.obsm[key_added]</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;example.h5ad&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runGDRParallel</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;CellTypes&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_gene</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">max_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_gdr&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runCOSGParallel">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runCOSGParallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_iter</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_gene_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runCOSGParallel" title="Link to this definition">#</a></dt>
<dd><p>Run COSG on batches in parallel using shared memory and multiprocessing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The key in <cite>adata.obs</cite> used to identify batches.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – The key in <cite>adata.obs</cite> used to group observations for clustering. If None, clustering will be performed.</p></li>
<li><p><strong>n_svd_dims</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 50</em><em>)</em>) – Number of SVD components to compute.</p></li>
<li><p><strong>n_svd_iter</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=7</em>) – Number of iterations for randomized SVD solver. The default is larger than the default in randomized_svd to handle sparse matrices that may have large slowly decaying spectrum. Also larger than the <cite>n_iter</cite> default value (5) in the TruncatedSVD function.</p></li>
<li><p><strong>n_highly_variable_genes</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 5000</em><em>)</em>) – Number of highly variable genes to use for SVD.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity for logging information.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – Resolution parameter for clustering.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Layer of the <cite>adata</cite> object to use for COSG.</p></li>
<li><p><strong>infog_layer</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – If specified, the INFOG normalization will be calculated using this layer of <cite>adata.layers</cite>, which is expected to contain the UMI count matrix. Defaults to None.</p></li>
<li><p><strong>mu</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – COSG parameter to control regularization.</p></li>
<li><p><strong>n_gene</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 30</em><em>)</em>) – Number of marker genes to compute for each cluster.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to use highly variable genes for SVD.</p></li>
<li><p><strong>return_gene_names</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to return gene names instead of indices in the marker gene DataFrame.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 8</em><em>)</em>) – Maximum number of parallel workers to use. If None, defaults to the number of available CPU cores.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Combined marker gene DataFrame with batch-specific suffixes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;example_data.h5ad&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">marker_genes</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runCOSGParallel</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_svd_dims</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_highly_variable_genes</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">resolution</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;log1p&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">mu</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_gene</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">return_gene_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">max_workers</span><span class="o">=</span><span class="mi">4</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">marker_genes</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.leiden_local">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">leiden_local</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'each'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Leiden_local'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dr_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gdr_resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.leiden_local" title="Link to this definition">#</a></dt>
<dd><p>Perform Leiden clustering locally, i.e., on selected group(s), on an AnnData object.
This function enables flexible clustering within specified groups, supports batch effect handling, and stores results in the AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>)</p></li>
<li><p><strong>clustering_type</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'each'</em><em>)</em>) – Specifies the clustering approach:
- ‘each’: Perform clustering independently within each group.
- ‘all’: Perform clustering across all selected groups.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'Leiden'</em><em>)</em>) – The key in <cite>adata.obs</cite> specifying the cell labels to be used for selecting groups.</p></li>
<li><p><strong>groups</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – A list of specific group(s) to be clustered. If None, all groups in the <cite>groupby</cite> category will be used.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.25</em><em>)</em>) – Resolution parameter for the Leiden algorithm, controlling clustering granularity.
Higher values result in more clusters.</p></li>
<li><p><strong>batch_key</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Key in <cite>adata.obs</cite> specifying batch labels. If provided, it handles batch effects during clustering. If None, batch effects are ignored.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'Leiden_local'</em><em>)</em>) – The name of the key under which the local Leiden clustering results will be stored in <cite>adata.obs</cite>.</p></li>
<li><p><strong>dr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'X_pca'</em><em>)</em>) – Dimensionality reduction method to be used for local clustering.
Allowed values are: ‘X_pca’, ‘X_gdr’, ‘X_pca_harmony’, ‘X_svd_full’, ‘X_svd_full_harmony’.</p></li>
<li><p><strong>gdr_resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – Resolution parameter for the GDR dimensionality reduction method if ‘dr_method’ is set to ‘X_gdr’.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – If False, the operation is performed in-place. If True, a copy of the <cite>adata</cite> object is returned with the clustering results added.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>If <cite>copy=True</cite>: Returns a new AnnData object with clustering results added to <cite>adata.obs[key_added]</cite>.</p></li>
<li><p>If <cite>copy=False</cite>: Modifies the input <cite>adata</cite> object in-place by adding clustering results to <cite>adata.obs[key_added]</cite>.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData or None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example usage</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">leiden_local</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">clustering_type</span><span class="o">=</span><span class="s1">&#39;each&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;Leiden&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">resolution</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;Leiden_local&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">dr_method</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">copy</span><span class="o">=</span><span class="kc">False</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.infog">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">infog</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'infog'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added_highly_variable_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'highly_variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.infog" title="Link to this definition">#</a></dt>
<dd><p>Performs INFOG normalization of single-cell RNA sequencing data based on “biological information”.</p>
<p>This function outputs the selected highly variable genes and normalized gene expression values based on the raw UMI counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – An AnnData object.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, returns a new AnnData object with the normalized data instead of modifying <cite>adata</cite> in place.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, the normalized data is stored in <cite>adata.X</cite> rather than in <cite>adata.layers[key_added]</cite>.</p></li>
<li><p><strong>n_top_genes</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=3000</em>) – The number of top highly variable genes to select.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='infog'</em>) – The key under which the normalized gene expression matrix is stored in <cite>adata.layers</cite>.</p></li>
<li><p><strong>key_added_highly_variable_gene</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='highly_variable'</em>) – The key under which the selection of highly variable genes is stored in <cite>adata.var</cite>.</p></li>
<li><p><strong>trim</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – If True, trim the normalized gene expression values.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=1</em>) – Controls the level of logging and output messages.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies which layer of <cite>adata</cite> to use for INFOG normalization. If None, <cite>adata.X</cite> is used. Note: the raw UMIs counts should be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>If <cite>copy</cite> is True, returns a modified AnnData object with the normalized expression matrix.</p></li>
<li><p>Otherwise, modifies <cite>adata</cite> in place.</p></li>
<li><p>The normalized gene expression values will be saved in <cite>adata.X</cite> if <cite>inplace</cite> is True, or in <cite>adata.layers</cite></p></li>
<li><p>with the key <cite>key_added</cite> by default if <cite>inplace</cite> is False.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">infog</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;infog&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;raw&quot;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the normalized data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;infog&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the highly variable genes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.score">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_nearest_neighbors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'infog'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_ctrl_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.score" title="Link to this definition">#</a></dt>
<dd><p>For a given gene set, compute gene expression enrichment scores and P values for all the cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The AnnData object for the gene expression matrix.</p></li>
<li><p><strong>gene_list</strong> (<em>list</em><em> of </em><em>str</em>) – A list of gene names for which the score will be computed.</p></li>
<li><p><strong>gene_weights</strong> (<em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – A list of weights corresponding to the genes in <cite>gene_list</cite>. The length of
<cite>gene_weights</cite> must match the length of <cite>gene_list</cite>. If None, all genes in
<cite>gene_list</cite> are weighted equally. Default is None.</p></li>
<li><p><strong>n_nearest_neighbors</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of nearest neighbors to consider for randomly selecting control gene sets based on the similarity of genes’ mean and variance among all cells.
Default is 30.</p></li>
<li><p><strong>leaf_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Leaf size for the KD-tree or Ball-tree used in nearest neighbor calculations. Default is 40.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the layer in <cite>adata.layers</cite> to use for gene expression values. Default is ‘infog’.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
<li><p><strong>n_ctrl_set</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of control gene sets to be used for calculating P values. Default is 100.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em>) – If provided, the computed scores will be stored in <cite>adata.obs[key_added]</cite>. The scores and P values will be stored in <cite>adata.uns[key_added]</cite> as well.
Default is None, and the <cite>INFOG_score</cite> will be used as the key.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity for logging information.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Modifies the <cite>adata</cite> object in-place, see <cite>key_added</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.predictCellTypeByGDR">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">predictCellTypeByGDR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'log1p'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer_reference</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'log1p'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CellTypes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query_groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_integration</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoring_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.predictCellTypeByGDR" title="Link to this definition">#</a></dt>
<dd><p>Predicts cell types in a query dataset (<cite>adata</cite>) using the GDR dimensionality reduction method based on a reference dataset (<cite>adata_ref</cite>).
To use GDR for dimensionality reduction, please refer to <cite>piaso.tl.runGDR</cite> or <cite>piaso.tl.runGDRParallel</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The query single-cell AnnData object for which cell types are to be predicted.</p></li>
<li><p><strong>adata_ref</strong> (<em>AnnData</em>) – The reference single-cell AnnData object with known cell type annotations.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'log1p'</em><em>)</em>) – The layer in <cite>adata</cite> to use for gene expression data. If <cite>None</cite>, uses the <cite>.X</cite> matrix.</p></li>
<li><p><strong>layer_reference</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'log1p'</em><em>)</em>) – The layer in <cite>adata_ref</cite> to use for reference gene expression data. If <cite>None</cite>, uses the <cite>.X</cite> matrix.</p></li>
<li><p><strong>reference_groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'CellTypes'</em><em>)</em>) – The column in <cite>adata_ref.obs</cite> used to define reference cell type groupings.</p></li>
<li><p><strong>query_groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'Leiden'</em><em>)</em>) – The column in <cite>adata.obs</cite> used to for GDR dimensionality reduction, such as clusters identified using Leiden or Louvain algorithms.</p></li>
<li><p><strong>mu</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 10.0</em><em>)</em>) – A regularization parameter for controlling the gene expression specificity, used in COSG (marker gene identification) and GDR.</p></li>
<li><p><strong>n_genes</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 15</em><em>)</em>) – The number of top specific genes per group, used in COSG and GDR.</p></li>
<li><p><strong>return_integration</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – If <cite>True</cite>, the function will return the integrated low-dimensional cell embeddings of the query dataset and reference dataset.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to use highly variable genes, used in GDR.</p></li>
<li><p><strong>n_highly_variable_genes</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 5000</em><em>)</em>) – The number of highly variable genes to select, if <cite>use_highly_variable</cite> is <cite>True</cite>, used in GDR.</p></li>
<li><p><strong>n_svd_dims</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 50</em><em>)</em>) – The number of dimensions to retain during SVD, used in GDR.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – Resolution parameter for clustering, used in GDR.</p></li>
<li><p><strong>scoring_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – The method used for gene set scoring, used in GDR.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – A key to add the predicted cell types or integration results to <cite>adata.obs</cite>. If <cite>None</cite>, <cite>CellTypes_gdr</cite> will be used.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – The level of logging output. Higher values produce more detailed logs for debugging and monitoring progress.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>return_integration</cite> is <cite>True</cite>, returns an AnnData object of merged reference and query datasets with integrated
cell embeddings and predicted cell types. Otherwise, updates <cite>adata</cite> in place with the predicted cell types.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None or AnnData</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load query dataset</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;query_data.h5ad&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load reference dataset with known cell type annotations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata_ref</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;reference_data.h5ad&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Predict cell types for the query dataset</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">predictCellTypeByGDR</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">adata_ref</span><span class="o">=</span><span class="n">adata_ref</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;log1p&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">layer_reference</span><span class="o">=</span><span class="s1">&#39;log1p&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">reference_groupby</span><span class="o">=</span><span class="s1">&#39;CellTypes&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">query_groupby</span><span class="o">=</span><span class="s1">&#39;Leiden&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">mu</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">n_genes</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">return_integration</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">n_highly_variable_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">n_svd_dims</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">resolution</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;CellTypes_gdr&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the predicted cell types in the query dataset</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;CellTypes_gdr&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.smoothCellTypePrediction">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">smoothCellTypePrediction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k_nearest_neighbors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_confidence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_existing_adjacency_graph</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_faiss</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.smoothCellTypePrediction" title="Link to this definition">#</a></dt>
<dd><p>Smooth cell type predictions using k-nearest neighbors in a low-dimensional embedding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object containing single-cell data</p></li>
<li><p><strong>groupby</strong> (<em>str</em>) – Key in adata.obs containing the cell type predictions to smooth</p></li>
<li><p><strong>use_rep</strong> (<em>str</em><em>, </em><em>default='X_pca'</em>) – Key in adata.obsm containing the low-dimensional embedding to use for finding neighbors</p></li>
<li><p><strong>k_nearest_neighbors</strong> (<em>int</em><em>, </em><em>default=5</em>) – Number of neighbors to consider (including the cell itself)</p></li>
<li><p><strong>return_confidence</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Whether to return confidence scores (proportion of neighbors with the majority label)</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to modify adata inplace or return a copy</p></li>
<li><p><strong>use_existing_adjacency_graph</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to use existing neighborhood graph (adata.obsp[‘connectivities’]) if available</p></li>
<li><p><strong>use_faiss</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Whether to use FAISS for faster neighbor search (requires faiss package)</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>default=None</em>) – If provided, use this key as the output key in adata.obs instead of ‘{groupby}_smoothed’</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>default=1</em>) – Level of verbosity (0=no output, 1=basic info, 2=detailed info)</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em><em>, </em><em>default=-1</em>) – Number of jobs for parallel processing. -1 means using all processors.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>If inplace=True</em> – None, but adds ‘groupby_smoothed’ (or key_added) to adata.obs
If return_confidence=True, also adds ‘groupby_confidence’ (or key_added_confidence) to adata.obs</p></li>
<li><p><em>If inplace=False</em> – Copy of adata with added columns</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Basic usage</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">smoothCellTypePrediction</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;CellTypes_pred&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;CellTypes_pred_smoothed&#39;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># With confidence scores</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">smoothCellTypePrediction</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;CellTypes_pred&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">k_nearest_neighbors</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">return_confidence</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;CellTypes_pred_smoothed&#39;</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.predictCellTypeByMarker">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">predictCellTypeByMarker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_gene_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Dict</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'scanpy'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'piaso'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'piaso'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'infog'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smooth_prediction</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_gdr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">k_nearest_neighbors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_confidence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_existing_adjacency_graph</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_faiss</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CellTypes_predicted'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extract_cell_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delimiter_cell_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.predictCellTypeByMarker" title="Link to this definition">#</a></dt>
<dd><p>Predict cell types using marker genes and optionally smooth predictions.</p>
<p>This function performs cell type prediction using marker genes in two steps:
1. Calculate gene set scores for marker genes
2. Optionally smooth the predictions using k-nearest neighbors</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – AnnData object containing single-cell data</p></li>
<li><p><strong>marker_gene_set</strong> (<em>list</em><em>, </em><em>dict</em><em>, or </em><em>pandas.DataFrame</em>) – Collection of marker genes for different cell types (pre-filtered/prepared)</p></li>
<li><p><strong>score_method</strong> (<em>{'scanpy'</em><em>, </em><em>'piaso'}</em><em>, </em><em>default='piaso'</em>) – Method to use for scoring marker genes</p></li>
<li><p><strong>score_layer</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>default='infog'</em>) – Layer of the AnnData object to use for scoring</p></li>
<li><p><strong>use_score</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to use scores (True) or p-values (False) for cell type prediction</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>default=None</em>) – Number of parallel workers for score calculation</p></li>
<li><p><strong>smooth_prediction</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to smooth predictions using k-nearest neighbors</p></li>
<li><p><strong>use_rep</strong> (<em>str</em><em>, </em><em>default='X_gdr'</em>) – Key in adata.obsm containing the low-dimensional embedding to use for neighbor search</p></li>
<li><p><strong>k_nearest_neighbors</strong> (<em>int</em><em>, </em><em>default=7</em>) – Number of neighbors to consider for smoothing</p></li>
<li><p><strong>return_confidence</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to return confidence scores for smoothed predictions</p></li>
<li><p><strong>use_existing_adjacency_graph</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Whether to use existing neighborhood graph if available</p></li>
<li><p><strong>use_faiss</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Whether to use FAISS for faster neighbor search</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>default='CellTypes_predicted'</em>) – Key to use for storing cell type predictions in adata.obs</p></li>
<li><p><strong>extract_cell_type</strong> (<em>bool</em><em>, </em><em>default=False</em>) – Whether to extract cell type name by removing suffix after delimiter</p></li>
<li><p><strong>delimiter_cell_type</strong> (<em>str</em><em>, </em><em>default='-'</em>) – Delimiter to use when extracting cell type names (only used if extract_cell_type=True)</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default=True</em>) – Whether to modify adata in place or return a copy</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>default=1927</em>) – Random seed for reproducibility</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>default=1</em>) – Level of verbosity (0=quiet, 1=basic info, 2=detailed)</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em><em>, </em><em>default=-1</em>) – Number of jobs for parallel processing during smoothing</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>If inplace=False</em> – AnnData: Copy of adata with cell type predictions added</p></li>
<li><p><em>If inplace=True</em> – None, but adata is modified in place</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Basic usage</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">predictCellTypeByMarker</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">marker_gene_set</span><span class="o">=</span><span class="n">cosgMarkerDB</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">score_method</span><span class="o">=</span><span class="s1">&#39;piaso&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">use_score</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">smooth_prediction</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.stitchSpace">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">stitchSpace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rep</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_stitch'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_cluster_key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_pruned_graph_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_use_global_markers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_leiden_resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_leiden_n_neighbors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_n_markers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_marker_overlap_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_cosg_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_cosg_mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_cosg_expressed_pct</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_cosg_remove_lowly_expressed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_bbknn_neighbors_within_batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_bbknn_trim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction_smooth_within_batch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correction_use_mutual_sqrt_weights</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">AnnData</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span></span><a class="headerlink" href="#piaso.tools.stitchSpace" title="Link to this definition">#</a></dt>
<dd><p>Performs a batch correction using a BBKNN graph that has been pruned based on marker gene overlap between batch-specific
clusters. Overlap check uses local markers and optionally global markers (controlled by <cite>filter_use_global_markers</cite>).</p>
<p>Clusters are identified internally using Leiden and stored in <cite>adata.obs[filter_cluster_key_added]</cite>.
Markers identified by COSG.
Intermediate results like the compatibility ‘hypergraph’, marker gene lists, and the pruned graph
structure are stored in <cite>adata.uns</cite> and <cite>adata.obsp</cite>/<cite>adata.uns</cite>.</p>
<p>The correction moves each cell towards the average position of its neighbors in the pruned graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – Annotated data matrix. Needs expression data for COSG (in .X or specified layer).</p></li>
<li><p><strong>batch_key</strong> – Key in <cite>adata.obs</cite> for batch information.</p></li>
<li><p><strong>use_rep</strong> – Representation in <cite>adata.obsm</cite> for BBKNN, clustering, and correction (e.g., ‘X_pca’).</p></li>
<li><p><strong>key_added</strong> – Base key for storing results. Corrected embedding will be in
<cite>adata.obsm[key_added]</cite>. Intermediate results stored in <cite>adata.uns</cite>.</p></li>
<li><p><strong>filter_cluster_key_added</strong> – Key in <cite>adata.obs</cite> where generated batch-cluster labels will be stored.
If None, a default key is generated (e.g., f”{batch_key}&#64;leiden&#64;res{res}”).</p></li>
<li><p><strong>filter_pruned_graph_key</strong> – Base key for storing the pruned graph structure in <cite>adata.obsp</cite> and <cite>adata.uns</cite>.
If None, defaults to “pruned_markers”. Connectivities/distances will be stored
as <cite>{filter_pruned_graph_key}_connectivities</cite>/<cite>_distances</cite>.</p></li>
<li><p><strong>filter_use_global_markers</strong> – If True, run global COSG and require BOTH local AND global marker overlap
for inter-batch cluster compatibility. If False (default), only local overlap is used.</p></li>
<li><p><strong>filter_leiden_resolution</strong> – Resolution parameter for internal within-batch Leiden clustering.</p></li>
<li><p><strong>filter_leiden_n_neighbors</strong> – KNN parameter for internal within-batch Leiden clustering’s graph.</p></li>
<li><p><strong>filter_n_markers</strong> – Number of top COSG markers to compare between clusters.</p></li>
<li><p><strong>filter_marker_overlap_threshold</strong> – Minimum Jaccard index for marker overlap to consider clusters compatible.</p></li>
<li><p><strong>filter_cosg_layer</strong> – Layer in <cite>adata.layers</cite> to use for COSG marker identification. If None (default), uses <cite>adata.X</cite>.</p></li>
<li><p><strong>filter_cosg_mu</strong> – <cite>mu</cite> parameter for COSG (default: 100.0). Higher values increase sparsity.</p></li>
<li><p><strong>filter_cosg_expressed_pct</strong> – <cite>expressed_pct</cite> parameter for COSG (default: 0.1). Minimum expression pct for a gene.</p></li>
<li><p><strong>filter_cosg_remove_lowly_expressed</strong> – <cite>remove_lowly_expressed</cite> parameter for COSG (default: True). Filter lowly expressed genes.</p></li>
<li><p><strong>filter_bbknn_neighbors_within_batch</strong> – <cite>neighbors_within_batch</cite> parameter for the initial <cite>bbknn.bbknn</cite> call.</p></li>
<li><p><strong>filter_bbknn_trim</strong> – Optional <cite>trim</cite> parameter passed to the initial <cite>bbknn.bbknn</cite> call.</p></li>
<li><p><strong>random_state</strong> – Seed for the random number generator used in Leiden clustering for reproducibility.
Default: 1927.</p></li>
<li><p><strong>correction_smooth_within_batch</strong> – If True, smooth the correction vector within batches using the pruned graph structure.</p></li>
<li><p><strong>correction_use_mutual_sqrt_weights</strong> – If True, applies symmetrization and sqrt weighting to the <em>pruned</em> graph
before the correction step.</p></li>
<li><p><strong>copy</strong> – If True, return a modified copy of <cite>adata</cite>. Otherwise, modify <cite>adata</cite> inplace.</p></li>
<li><p><strong>verbosity</strong> – Level of detail to print: 0 (minimal), 1 or higher (more progress messages
and intermediate storage locations). Default: 0. Controls BBKNN logging level.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>copy=True</cite>, returns the modified AnnData object.
Otherwise, modifies the input <cite>adata</cite> object inplace and returns None.
Adds/updates:
- <cite>adata.obsm[key_added]</cite>: The corrected embedding.
- <cite>adata.obs[filter_cluster_key_added]</cite>: Generated batch-cluster labels (using ‘&#64;’ delimiter).
- <cite>adata.uns[f’{key_added}_hypergraph_compatibility’]</cite>: Compatibility dict.
- <cite>adata.uns[f’{key_added}_local_markers’]</cite>: Local marker dict.
- <cite>adata.uns[f’{key_added}_global_markers’]</cite>: Global marker dict.
- <cite>adata.obsp[f’{filter_pruned_graph_key}_connectivities’]</cite>: Pruned graph connectivities.
- <cite>adata.obsp[f’{filter_pruned_graph_key}_distances’]</cite>: Pruned graph dummy distances.
- <cite>adata.uns[filter_pruned_graph_key]</cite>: Neighbors dictionary for pruned graph.
- <cite>adata.uns[f’{key_added}_params’]</cite>: Dictionary of parameters used.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData or None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">stitchSpaceModule</span> <span class="c1"># Assuming the code is saved as stitchSpaceModule.py</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">datasets</span><span class="o">.</span><span class="n">pbmc68k_reduced</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Simulate batches (replace with actual batch info)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span> <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;B&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Assume normalized data is in adata.layers[&#39;log1p&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;log1p&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Precompute PCA if not done</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Run correction using log1p layer for COSG, increased verbosity</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">stitchSpace</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;X_stitch_corrected&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">filter_cluster_key_added</span><span class="o">=</span><span class="s1">&#39;batch@cluster_stitch&#39;</span><span class="p">,</span> <span class="c1"># Optional: specify key</span>
<span class="gp">... </span>    <span class="n">filter_cosg_layer</span><span class="o">=</span><span class="s1">&#39;log1p&#39;</span><span class="p">,</span> <span class="c1"># Specify layer for COSG</span>
<span class="gp">... </span>    <span class="n">random_state</span><span class="o">=</span><span class="mi">1927</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">1</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Visualize results</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;X_stitch_corrected&#39;</span><span class="p">)</span> <span class="c1"># Compute neighbors on corrected embedding</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;batch@cluster_stitch&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Visualize UMAP based on the pruned graph itself</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">neighbors_key</span><span class="o">=</span><span class="s1">&#39;pruned_markers&#39;</span><span class="p">)</span> <span class="c1"># Use default key or specify if changed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span> <span class="s1">&#39;batch@cluster_stitch&#39;</span><span class="p">],</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;UMAP on Pruned Graph&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runSCALAR">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runSCALAR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specificity_matrix</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lr_pairs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ligand_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'ligand'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receptor_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'receptor'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sender_cell_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">receiver_cell_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_permutations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_nearest_neighbors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">42</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_by_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunk_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefilter_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefilter_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="headerlink" href="#piaso.tools.runSCALAR" title="Link to this definition">#</a></dt>
<dd><p>Calculates ligand-receptor interaction scores, computes permutation-based p-values
using a vectorized approach, and corrects for multiple testing using FDR for each
cell type-cell type pair independently.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> – AnnData object with gene expression data.</p></li>
<li><p><strong>specificity_matrix</strong> – DataFrame with genes as rows, cell types as columns,
and specificity scores as values.</p></li>
<li><p><strong>lr_pairs</strong> – DataFrame listing interacting gene pairs.</p></li>
<li><p><strong>ligand_col</strong> – Column name for ligands in lr_pairs.</p></li>
<li><p><strong>receptor_col</strong> – Column name for receptors in lr_pairs.</p></li>
<li><p><strong>annotation_col</strong> – Optional column in lr_pairs to carry over.</p></li>
<li><p><strong>sender_cell_types</strong> – List of cell types to use as senders. If None, all are used.</p></li>
<li><p><strong>receiver_cell_types</strong> – List of cell types to use as receivers. If None, all are used.</p></li>
<li><p><strong>n_permutations</strong> – Number of permutations for the null distribution.</p></li>
<li><p><strong>n_nearest_neighbors</strong> – Number of control genes to sample from.</p></li>
<li><p><strong>layer</strong> – Layer in adata to use for expression.</p></li>
<li><p><strong>random_seed</strong> – Seed for reproducibility.</p></li>
<li><p><strong>rank_by_score</strong> – If True, sorts the final output by interaction_score.</p></li>
<li><p><strong>chunk_size</strong> – The number of interactions to process in each vectorized chunk
to manage memory usage.</p></li>
<li><p><strong>prefilter_fdr</strong> – If True, interactions with scores &lt;= prefilter_threshold are
excluded from FDR calculation within each group and assigned an
FDR of 1.0.</p></li>
<li><p><strong>prefilter_threshold</strong> – The score threshold used for pre-filtering before FDR calculation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A pandas DataFrame with interaction scores, p-values, and FDR-corrected p-values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.queryPIASOmarkerDB">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">queryPIASOmarkerDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">study</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tissue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_studies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_cell_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#piaso.tools.queryPIASOmarkerDB" title="Link to this definition">#</a></dt>
<dd><p>Query PIASOmarkerDB for cell type marker genes.</p>
<p>This is the main entry point for accessing PIASOmarkerDB through PIASO.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Gene symbol(s) to filter by.</p></li>
<li><p><strong>cell_type</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Cell type(s) to filter by.</p></li>
<li><p><strong>study</strong> (<em>str</em><em>, </em><em>optional</em>) – Study/publication to filter by.</p></li>
<li><p><strong>species</strong> (<em>str</em><em>, </em><em>optional</em>) – Species to filter by (e.g., “Human”, “Mouse”).</p></li>
<li><p><strong>tissue</strong> (<em>str</em><em>, </em><em>optional</em>) – Tissue to filter by.</p></li>
<li><p><strong>condition</strong> (<em>str</em><em>, </em><em>optional</em>) – Condition to filter by.</p></li>
<li><p><strong>min_score</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum specificity score (&gt;= 0).</p></li>
<li><p><strong>max_score</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum specificity score (&gt;= 0).</p></li>
<li><p><strong>limit</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum results to return. Default: None (no limit).</p></li>
<li><p><strong>as_dict</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, also return {cell_type: [genes]} dictionary.
Returns tuple (DataFrame, dict). Default: False.</p></li>
<li><p><strong>list_studies</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return list of available study names instead of markers.
Default: False.</p></li>
<li><p><strong>list_cell_types</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return list of available cell types instead of markers.
Default: False.</p></li>
<li><p><strong>list_genes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return list of unique gene symbols instead of markers.
Default: False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>pd.DataFrame</em> – Marker query results (default).</p></li>
<li><p><em>tuple (pd.DataFrame, dict)</em> – If as_dict=True: (DataFrame, {cell_type: [genes]}).</p></li>
<li><p><em>list of str</em> – If list_studies=True, list_cell_types=True, or list_genes=True.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Query marker genes:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="s2">&quot;Foxp2&quot;</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Foxp2&quot;</span><span class="p">,</span> <span class="s2">&quot;Syt6&quot;</span><span class="p">,</span> <span class="s2">&quot;Tle4&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Get both DataFrame and marker dictionary:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="p">,</span> <span class="n">marker_dict</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">study</span><span class="o">=</span><span class="s2">&quot;AllenWholeMouseBrain_isocortex&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">as_dict</span><span class="o">=</span><span class="kc">True</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DataFrame shape: </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cell types in dict: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">marker_dict</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>List available studies:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">studies</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">list_studies</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total studies: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">studies</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>List cell types:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cell_types</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">list_cell_types</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#piaso.tools.analyzeMarkers" title="piaso.tools.analyzeMarkers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">analyzeMarkers</span></code></a></dt><dd><p>Analyze gene lists for cell type inference</p>
</dd>
<dt><a class="reference internal" href="#piaso.tools.PIASOmarkerDB" title="piaso.tools.PIASOmarkerDB"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PIASOmarkerDB</span></code></a></dt><dd><p>Client class for advanced usage</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>PIASOmarkerDB website: <a class="reference external" href="https://piaso.org/piasomarkerdb/">https://piaso.org/piasomarkerdb/</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.getMarkers">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">getMarkers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">study</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tissue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">condition</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">as_dict</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_studies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_cell_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">list_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#piaso.tools.getMarkers" title="Link to this definition">#</a></dt>
<dd><p>Query PIASOmarkerDB for cell type marker genes.</p>
<p>This is the main entry point for accessing PIASOmarkerDB through PIASO.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Gene symbol(s) to filter by.</p></li>
<li><p><strong>cell_type</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Cell type(s) to filter by.</p></li>
<li><p><strong>study</strong> (<em>str</em><em>, </em><em>optional</em>) – Study/publication to filter by.</p></li>
<li><p><strong>species</strong> (<em>str</em><em>, </em><em>optional</em>) – Species to filter by (e.g., “Human”, “Mouse”).</p></li>
<li><p><strong>tissue</strong> (<em>str</em><em>, </em><em>optional</em>) – Tissue to filter by.</p></li>
<li><p><strong>condition</strong> (<em>str</em><em>, </em><em>optional</em>) – Condition to filter by.</p></li>
<li><p><strong>min_score</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum specificity score (&gt;= 0).</p></li>
<li><p><strong>max_score</strong> (<em>float</em><em>, </em><em>optional</em>) – Maximum specificity score (&gt;= 0).</p></li>
<li><p><strong>limit</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum results to return. Default: None (no limit).</p></li>
<li><p><strong>as_dict</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, also return {cell_type: [genes]} dictionary.
Returns tuple (DataFrame, dict). Default: False.</p></li>
<li><p><strong>list_studies</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return list of available study names instead of markers.
Default: False.</p></li>
<li><p><strong>list_cell_types</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return list of available cell types instead of markers.
Default: False.</p></li>
<li><p><strong>list_genes</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return list of unique gene symbols instead of markers.
Default: False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>pd.DataFrame</em> – Marker query results (default).</p></li>
<li><p><em>tuple (pd.DataFrame, dict)</em> – If as_dict=True: (DataFrame, {cell_type: [genes]}).</p></li>
<li><p><em>list of str</em> – If list_studies=True, list_cell_types=True, or list_genes=True.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Query marker genes:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="s2">&quot;Foxp2&quot;</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Foxp2&quot;</span><span class="p">,</span> <span class="s2">&quot;Syt6&quot;</span><span class="p">,</span> <span class="s2">&quot;Tle4&quot;</span><span class="p">])</span>
</pre></div>
</div>
<p>Get both DataFrame and marker dictionary:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="p">,</span> <span class="n">marker_dict</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">study</span><span class="o">=</span><span class="s2">&quot;AllenWholeMouseBrain_isocortex&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">as_dict</span><span class="o">=</span><span class="kc">True</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DataFrame shape: </span><span class="si">{</span><span class="n">df</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cell types in dict: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">marker_dict</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>List available studies:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">studies</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">list_studies</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Total studies: </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">studies</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>List cell types:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cell_types</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">queryPIASOmarkerDB</span><span class="p">(</span><span class="n">list_cell_types</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#piaso.tools.analyzeMarkers" title="piaso.tools.analyzeMarkers"><code class="xref py py-obj docutils literal notranslate"><span class="pre">analyzeMarkers</span></code></a></dt><dd><p>Analyze gene lists for cell type inference</p>
</dd>
<dt><a class="reference internal" href="#piaso.tools.PIASOmarkerDB" title="piaso.tools.PIASOmarkerDB"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PIASOmarkerDB</span></code></a></dt><dd><p>Client class for advanced usage</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>PIASOmarkerDB website: <a class="reference external" href="https://piaso.org/piasomarkerdb/">https://piaso.org/piasomarkerdb/</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.analyzeMarkers">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">analyzeMarkers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tissue</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">studies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_cell_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_studies</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">DataFrame</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#piaso.tools.analyzeMarkers" title="Link to this definition">#</a></dt>
<dd><p>Analyze gene list(s) to infer potential cell types.</p>
<p>This function queries PIASOmarkerDB to find which cell types are associated
with the input genes, ranking results by matched gene count and specificity.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genes</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>pd.DataFrame</em><em>, or </em><em>dict</em>) – <p>Gene input. Supports three formats:</p>
<ul>
<li><p><strong>list of str</strong>: Single list of gene symbols.
Returns: pd.DataFrame with analysis results.</p></li>
<li><p><strong>pd.DataFrame</strong>: Columns are clusters/cell types, rows are genes.
Ideal for COSG output: <code class="docutils literal notranslate"><span class="pre">pd.DataFrame(adata.uns['cosg']['names']).head(50)</span></code>.
Returns: tuple (results_dict, top_hits_dict).</p></li>
<li><p><strong>dict</strong>: <code class="docutils literal notranslate"><span class="pre">{cluster_name:</span> <span class="pre">[gene_list]}</span></code>.
Returns: tuple (results_dict, top_hits_dict).</p></li>
</ul>
</p></li>
<li><p><strong>n_top_genes</strong> (<em>int</em><em>, </em><em>optional</em>) – For DataFrame/dict input: only use top N genes per column/key.
Default: 50. Useful for COSG results which may rank many genes.</p></li>
<li><p><strong>species</strong> (<em>str</em><em>, </em><em>optional</em>) – Filter results by species (e.g., “Human”, “Mouse”).</p></li>
<li><p><strong>tissue</strong> (<em>str</em><em>, </em><em>optional</em>) – Filter results by tissue.</p></li>
<li><p><strong>studies</strong> (<em>str</em><em> or </em><em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Study/studies to include in analysis. Only cell types from these
studies will be considered. Study names are validated against
PIASOmarkerDB. Cannot overlap with exclude_studies.
Default: None (use all studies).</p></li>
<li><p><strong>min_genes</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum number of genes that must match a cell type. Default: 1.</p></li>
<li><p><strong>exclude_cell_types</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Cell types to exclude from results.</p></li>
<li><p><strong>exclude_studies</strong> (<em>list</em><em> of </em><em>str</em><em>, </em><em>optional</em>) – Studies to exclude from results. Cannot overlap with studies.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><em>pd.DataFrame</em> – For single list input: DataFrame with columns:
cell_type, study_publication, species, tissue, condition,
matched_gene_count, matched_genes, avg_specificity.</p></li>
<li><p><em>tuple (dict, dict)</em> – For DataFrame or dict input:</p>
<ul>
<li><p><strong>results_dict</strong>: <code class="docutils literal notranslate"><span class="pre">{cluster_name:</span> <span class="pre">result_DataFrame}</span></code></p></li>
<li><p><strong>top_hits_dict</strong>: <code class="docutils literal notranslate"><span class="pre">{cluster_name:</span> <span class="pre">&quot;predicted_cell_type&quot;}</span></code>
(or “Unassigned” if no matches found)</p></li>
</ul>
</li>
</ul>
</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValidationError</strong> – If studies parameter contains invalid study names, or if
    studies and exclude_studies have overlapping values.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Single gene list (mouse cortex L6 markers):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">query_genes</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Syt6&quot;</span><span class="p">,</span> <span class="s2">&quot;Tle4&quot;</span><span class="p">,</span> <span class="s2">&quot;Hs3st4&quot;</span><span class="p">,</span> <span class="s2">&quot;Fezf2&quot;</span><span class="p">,</span> <span class="s2">&quot;Foxp2&quot;</span><span class="p">,</span> <span class="s2">&quot;Col12a1&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">analyzeMarkers</span><span class="p">(</span><span class="n">query_genes</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<span class="go"># Top hit: EN-L6-CT from WangKriegstein2025</span>
</pre></div>
</div>
<p>With specific study filter:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">results</span><span class="p">,</span> <span class="n">top_hits</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">analyzeMarkers</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">cosg_marker_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">min_genes</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">studies</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AllenWholeMouseBrain_isocortex&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">top_hits</span><span class="p">)</span>
<span class="go">{&#39;L2-3 IT&#39;: &#39;007 L2/3 IT CTX Glut&#39;, &#39;PV&#39;: &#39;052 Pvalb Gaba&#39;, ...}</span>
</pre></div>
</div>
<p>Dictionary input (microglia and L6 CT markers):</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gene_sets</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>    <span class="s1">&#39;Cluster_0&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Cx3cr1&#39;</span><span class="p">,</span> <span class="s1">&#39;P2ry12&#39;</span><span class="p">,</span> <span class="s1">&#39;Tmem119&#39;</span><span class="p">,</span> <span class="s1">&#39;Csf1r&#39;</span><span class="p">,</span> <span class="s1">&#39;Trem2&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="s1">&#39;Cluster_1&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;Syt6&#39;</span><span class="p">,</span> <span class="s1">&#39;Tle4&#39;</span><span class="p">,</span> <span class="s1">&#39;Hs3st4&#39;</span><span class="p">,</span> <span class="s1">&#39;Fezf2&#39;</span><span class="p">,</span> <span class="s1">&#39;Foxp2&#39;</span><span class="p">],</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span><span class="p">,</span> <span class="n">top_hits</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">analyzeMarkers</span><span class="p">(</span><span class="n">gene_sets</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">top_hits</span><span class="p">)</span>
<span class="go">{&#39;Cluster_0&#39;: &#39;Microglia&#39;, &#39;Cluster_1&#39;: &#39;EN-L6-CT&#39;}</span>
</pre></div>
</div>
<p>COSG integration workflow:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">cosg</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Run COSG</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cosg</span><span class="o">.</span><span class="n">cosg</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;cosg&#39;</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Get top 50 markers per cluster</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cosg_marker_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;cosg&#39;</span><span class="p">][</span><span class="s1">&#39;names&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Analyze with PIASOmarkerDB</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">results</span><span class="p">,</span> <span class="n">top_hits</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">analyzeMarkers</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">cosg_marker_df</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Add annotations to AnnData</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type_predicted&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">top_hits</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#piaso.tools.queryPIASOmarkerDB" title="piaso.tools.queryPIASOmarkerDB"><code class="xref py py-obj docutils literal notranslate"><span class="pre">queryPIASOmarkerDB</span></code></a></dt><dd><p>Direct marker queries</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>For COSG results, the DataFrame columns are cluster/cell type names and
rows contain the ranked marker genes. Only the top <code class="docutils literal notranslate"><span class="pre">n_top_genes</span></code> genes
per column are used for analysis.</p>
<p>When <code class="docutils literal notranslate"><span class="pre">studies</span></code> is provided, study names are validated against PIASOmarkerDB.
If an invalid study name is provided, a ValidationError is raised with
instructions to list available studies.</p>
<p>PIASOmarkerDB website: <a class="reference external" href="https://piaso.org/piasomarkerdb/">https://piaso.org/piasomarkerdb/</a></p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.PIASOmarkerDB">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">PIASOmarkerDB</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">base_url</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Path</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.PIASOmarkerDB" title="Link to this definition">#</a></dt>
<dd><p>Python client for accessing PIASOmarkerDB.</p>
<p>PIASOmarkerDB is a comprehensive database of cell type marker genes with
specificity scores across various tissues, species, studies, and conditions,
powered by PIASO (Precise Integrative Analysis of Single-cell Omics) methodologies.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>base_url</strong> (<em>str</em><em>, </em><em>optional</em>) – Base URL for the PIASOmarkerDB API.
Default: “<a class="reference external" href="https://piaso.org/piasomarkerdb">https://piaso.org/piasomarkerdb</a>”</p></li>
<li><p><strong>timeout</strong> (<em>int</em><em>, </em><em>optional</em>) – Request timeout in seconds. Default: 30</p></li>
<li><p><strong>cache_dir</strong> (<em>str</em><em> or </em><em>Path</em><em>, </em><em>optional</em>) – Directory for caching downloaded markers.
Default: ~/.piaso/markers</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Basic usage:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">piaso.tools</span> <span class="kn">import</span> <span class="n">PIASOmarkerDB</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">client</span> <span class="o">=</span> <span class="n">PIASOmarkerDB</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Query markers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">client</span><span class="o">.</span><span class="n">getMarkers</span><span class="p">(</span><span class="n">gene</span><span class="o">=</span><span class="s2">&quot;Foxp2&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Get as dict</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="p">,</span> <span class="n">marker_dict</span> <span class="o">=</span> <span class="n">client</span><span class="o">.</span><span class="n">getMarkers</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">study</span><span class="o">=</span><span class="s2">&quot;AllenWholeMouseBrain_isocortex&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">as_dict</span><span class="o">=</span><span class="kc">True</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Download markers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">client</span><span class="o">.</span><span class="n">downloadMarkers</span><span class="p">(</span><span class="s2">&quot;markers.csv&quot;</span><span class="p">,</span> <span class="n">species</span><span class="o">=</span><span class="s2">&quot;Mouse&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">piaso.tl.queryPIASOmarkerDB</span></code></dt><dd><p>Main query function</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">piaso.tl.analyzeMarkers</span></code></dt><dd><p>Gene list analysis</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>PIASOmarkerDB provides marker genes with specificity scores computed using
PIASO’s standardized methodology across multiple single-cell RNA-seq studies.</p>
<p>Website: <a class="reference external" href="https://piaso.org/piasomarkerdb/">https://piaso.org/piasomarkerdb/</a></p>
</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="piaso.preprocessing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">piaso.preprocessing package</p>
      </div>
    </a>
    <a class="right-next"
       href="references.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">References</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runSVD"><code class="docutils literal notranslate"><span class="pre">runSVD()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runSVDLazy"><code class="docutils literal notranslate"><span class="pre">runSVDLazy()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runGDR"><code class="docutils literal notranslate"><span class="pre">runGDR()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.calculateScoreParallel"><code class="docutils literal notranslate"><span class="pre">calculateScoreParallel()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.calculateScoreParallel_multiBatch"><code class="docutils literal notranslate"><span class="pre">calculateScoreParallel_multiBatch()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runGDRParallel"><code class="docutils literal notranslate"><span class="pre">runGDRParallel()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runCOSGParallel"><code class="docutils literal notranslate"><span class="pre">runCOSGParallel()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.leiden_local"><code class="docutils literal notranslate"><span class="pre">leiden_local()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.infog"><code class="docutils literal notranslate"><span class="pre">infog()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.score"><code class="docutils literal notranslate"><span class="pre">score()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.predictCellTypeByGDR"><code class="docutils literal notranslate"><span class="pre">predictCellTypeByGDR()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.smoothCellTypePrediction"><code class="docutils literal notranslate"><span class="pre">smoothCellTypePrediction()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.predictCellTypeByMarker"><code class="docutils literal notranslate"><span class="pre">predictCellTypeByMarker()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.stitchSpace"><code class="docutils literal notranslate"><span class="pre">stitchSpace()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runSCALAR"><code class="docutils literal notranslate"><span class="pre">runSCALAR()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.queryPIASOmarkerDB"><code class="docutils literal notranslate"><span class="pre">queryPIASOmarkerDB()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.getMarkers"><code class="docutils literal notranslate"><span class="pre">getMarkers()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.analyzeMarkers"><code class="docutils literal notranslate"><span class="pre">analyzeMarkers()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.PIASOmarkerDB"><code class="docutils literal notranslate"><span class="pre">PIASOmarkerDB()</span></code></a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Min Dai.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>