
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>piaso.tools package &#8212; PIASO 0.1.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="_static/custom.css?v=0330b353" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=01f34227"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'piaso.tools';</script>
    <script src="_static/sidebar_toggle.js?v=3b053ec6"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="References" href="references.html" />
    <link rel="prev" title="piaso.preprocessing package" href="piaso.preprocessing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class=" navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
    
    
      
    
    
    <img src="_static/PIASO_logo.png" class="logo__image only-light" alt="PIASO 0.1.0 documentation - Home"/>
    <img src="_static/PIASO_logo.png" class="logo__image only-dark pst-js-only" alt="PIASO 0.1.0 documentation - Home"/>
  
  
</a></div>
    
  </div>
  
  <div class=" navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tutorials.html">
    Tutorials
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="modules.html">
    API
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="references.html">
    References
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="releaseNotes.html">
    Release notes
  </a>
</li>

  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/genecell/PIASO" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="tutorials.html">
    Tutorials
  </a>
</li>


<li class="nav-item current active">
  <a class="nav-link nav-internal" href="modules.html">
    API
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="references.html">
    References
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="releaseNotes.html">
    Release notes
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/genecell/PIASO" title="GitHub" class="nav-link pst-navbar-icon" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><i class="fa-brands fa-square-github fa-lg" aria-hidden="true"></i>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<h3><a href="index.html">Table of Contents</a></h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="references.html">References</a></li>
<li class="toctree-l1"><a class="reference internal" href="releaseNotes.html">Release notes</a></li>
</ul>
</div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="modules.html" class="nav-link">API</a></li>
    
    
    <li class="breadcrumb-item"><a href="piaso.html" class="nav-link">piaso package</a></li>
    
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">piaso.tools package</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="module-piaso.tools">
<span id="piaso-tools-package"></span><h1>piaso.tools package<a class="headerlink" href="#module-piaso.tools" title="Link to this heading">#</a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.leiden_local">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">leiden_local</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">clustering_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'each'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Sequence</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Leiden_local'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dr_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_pca'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gdr_resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.leiden_local" title="Link to this definition">#</a></dt>
<dd><p>Perform Leiden clustering locally, i.e., on selected group(s), on an AnnData object.
This function enables flexible clustering within specified groups, supports batch effect handling, and stores results in the AnnData object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>)</p></li>
<li><p><strong>clustering_type</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'each'</em><em>)</em>) – Specifies the clustering approach:
- ‘each’: Perform clustering independently within each group.
- ‘all’: Perform clustering across all selected groups.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'Leiden'</em><em>)</em>) – The key in <cite>adata.obs</cite> specifying the cell labels to be used for selecting groups.</p></li>
<li><p><strong>groups</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – A list of specific group(s) to be clustered. If None, all groups in the <cite>groupby</cite> category will be used.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 0.25</em><em>)</em>) – Resolution parameter for the Leiden algorithm, controlling clustering granularity.
Higher values result in more clusters.</p></li>
<li><p><strong>batch_key</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Key in <cite>adata.obs</cite> specifying batch labels. If provided, it handles batch effects during clustering. If None, batch effects are ignored.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'Leiden_local'</em><em>)</em>) – The name of the key under which the local Leiden clustering results will be stored in <cite>adata.obs</cite>.</p></li>
<li><p><strong>dr_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'X_pca'</em><em>)</em>) – Dimensionality reduction method to be used for local clustering.
Allowed values are: ‘X_pca’, ‘X_gdr’, ‘X_pca_harmony’, ‘X_svd_full’, ‘X_svd_full_harmony’.</p></li>
<li><p><strong>gdr_resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – Resolution parameter for the GDR dimensionality reduction method if ‘dr_method’ is set to ‘X_gdr’.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – If False, the operation is performed in-place. If True, a copy of the <cite>adata</cite> object is returned with the clustering results added.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>If <cite>copy=True</cite>: Returns a new AnnData object with clustering results added to <cite>adata.obs[key_added]</cite>.</p></li>
<li><p>If <cite>copy=False</cite>: Modifies the input <cite>adata</cite> object in-place by adding clustering results to <cite>adata.obs[key_added]</cite>.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>AnnData or None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># Example usage</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">leiden_local</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">clustering_type</span><span class="o">=</span><span class="s1">&#39;each&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;Leiden&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groups</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="s1">&#39;1&#39;</span><span class="p">],</span>
<span class="gp">... </span>    <span class="n">resolution</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;Leiden_local&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">dr_method</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">copy</span><span class="o">=</span><span class="kc">False</span>
<span class="gp">... </span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.infog">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">infog</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'infog'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added_highly_variable_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'highly_variable'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.infog" title="Link to this definition">#</a></dt>
<dd><p>Performs INFOG normalization of single-cell RNA sequencing data based on “biological information”.</p>
<p>This function outputs the selected highly variable genes and normalized gene expression values based on the raw UMI counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – An AnnData object.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, returns a new AnnData object with the normalized data instead of modifying <cite>adata</cite> in place.</p></li>
<li><p><strong>n_top_genes</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=3000</em>) – The number of top highly variable genes to select.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='infog'</em>) – The key under which the normalized gene expression matrix is stored in <cite>adata.layers</cite>.</p></li>
<li><p><strong>key_added_highly_variable_gene</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='highly_variable'</em>) – The key under which the selection of highly variable genes is stored in <cite>adata.var</cite>.</p></li>
<li><p><strong>trim</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – If True, trim the normalized gene expression values.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=1</em>) – Controls the level of logging and output messages.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies which layer of <cite>adata</cite> to use for INFOG normalization. If None, <cite>adata.X</cite> is used. Note: the raw UMIs counts should be used.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>If <cite>copy</cite> is True, returns a modified AnnData object with the normalized expression matrix.</p></li>
<li><p>Otherwise, modifies <cite>adata</cite> in place.</p></li>
<li><p>The normalized gene expression values will be saved in <cite>adata.layers</cite> with key as <cite>infog</cite> by default or <cite>key_added</cite> if specified.</p></li>
</ul>
</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">infog</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;infog&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">trim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;raw&quot;</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the normalized data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s1">&#39;infog&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the highly variable genes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.score">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_weights</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_nearest_neighbors</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">leaf_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">40</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'infog'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_ctrl_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.score" title="Link to this definition">#</a></dt>
<dd><p>For a given gene set, compute gene expression enrichment scores and P values for all the cells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The AnnData object for the gene expression matrix.</p></li>
<li><p><strong>gene_list</strong> (<em>list</em><em> of </em><em>str</em>) – A list of gene names for which the score will be computed.</p></li>
<li><p><strong>gene_weights</strong> (<em>list</em><em> of </em><em>floats</em><em>, </em><em>optional</em>) – A list of weights corresponding to the genes in <cite>gene_list</cite>. The length of
<cite>gene_weights</cite> must match the length of <cite>gene_list</cite>. If None, all genes in
<cite>gene_list</cite> are weighted equally. Default is None.</p></li>
<li><p><strong>n_nearest_neighbors</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of nearest neighbors to consider for randomly selecting control gene sets based on the similarity of genes’ mean and variance among all cells.
Default is 30.</p></li>
<li><p><strong>leaf_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Leaf size for the KD-tree or Ball-tree used in nearest neighbor calculations. Default is 40.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – The name of the layer in <cite>adata.layers</cite> to use for gene expression values. Default is ‘infog’.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
<li><p><strong>n_ctrl_set</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of control gene sets to be used for calculating P values. Default is 100.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em>) – If provided, the computed scores will be stored in <cite>adata.obs[key_added]</cite>. The scores and P values will be stored in <cite>adata.uns[key_added]</cite> as well.
Default is None, and the <cite>INFOG_score</cite> will be used as the key.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity for logging information.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Modifies the <cite>adata</cite> object in-place, see <cite>key_added</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.calculateScoreParallel">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">calculateScoreParallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_set</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'scanpy'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'piaso'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.calculateScoreParallel" title="Link to this definition">#</a></dt>
<dd><p>Compute gene set scores in parallel using shared memory for efficiency.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The input AnnData object.</p></li>
<li><p><strong>gene_set</strong> (<em>dict</em><em>, </em><em>list</em><em> of </em><em>lists</em><em>, or </em><em>pandas.DataFrame</em>) – <dl class="simple">
<dt>A collection of gene sets, where each gene set is either:</dt><dd><ul>
<li><p>A dictionary: Keys are gene set names, values are lists of genes.</p></li>
<li><p>A list of lists: Each sublist contains genes in a gene set.</p></li>
<li><p>A pandas.DataFrame: Each column represents a gene set, and column names are gene set names.</p></li>
</ul>
</dd>
</dl>
</p></li>
<li><p><strong>score_method</strong> (<em>{'scanpy'</em><em>, </em><em>'piaso'}</em><em>, </em><em>optional</em>) – The method used for gene set scoring. Must be either ‘scanpy’ (default) or ‘piaso’.
- ‘scanpy’: Uses the Scanpy’s built-in gene set scoring method.
- ‘piaso’: Uses the PIASO’s gene set scoring method, which is more robust to sequencing depth variations.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
<li><p><strong>score_layer</strong> (<em>str</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Layer of the AnnData object to use. If None, <cite>adata.X</cite> is used.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em> or </em><em>None</em><em>, </em><em>optional</em>) – Number of parallel workers to use. Defaults to the number of CPUs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><dl class="simple">
<dt>np.ndarray: A 2D array where each column contains the scores for a gene set.</dt><dd><p>If <cite>gene_set</cite> is a dictionary, columns correspond to the keys of the dictionary.
If <cite>gene_set</cite> is a list of lists, columns correspond to the index of the sublists.
If <cite>gene_set</cite> is pandas.DataFrame, columns correspond to the index of the columns.</p>
</dd>
</dl>
</li>
<li><p>list: The names of the gene sets.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.calculateScoreParallel_multiBatch">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">calculateScoreParallel_multiBatch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">marker_gene_n_groups_indices</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'scanpy'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'piaso'</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.calculateScoreParallel_multiBatch" title="Link to this definition">#</a></dt>
<dd><p>Calculate gene set scores for each adata batch in parallel using shared memory. Different marker gene sets will be calculated in parallel as well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The key in <cite>adata.obs</cite> used to identify batches.</p></li>
<li><p><strong>marker_gene</strong> (<em>DataFrame</em>) – The marker gene DataFrame.</p></li>
<li><p><strong>marker_gene_n_groups_indices</strong> (<em>list</em>) – Indices specifying the marker gene set group boundaries, used for score normalization within each marker gene set group.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em>) – Maximum number of parallel workers to use.</p></li>
<li><p><strong>score_layer</strong> (<em>str</em>) – The layer of <cite>adata</cite> to use for scoring.</p></li>
<li><p><strong>score_method</strong> (<em>{'scanpy'</em><em>, </em><em>'piaso'}</em><em>, </em><em>optional</em>) – The method used for gene set scoring. Must be either ‘scanpy’ (default) or ‘piaso’.
- ‘scanpy’: Uses the Scanpy’s built-in gene set scoring method.
- ‘piaso’: Uses the PIASO’s gene set scoring method, which is more robust to sequencing depth variations.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>list: A list of normalized score arrays for each batch.</p></li>
<li><p>list: A list of cell barcodes for each batch.</p></li>
<li><p>list: A list of gene set names.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;example_data.h5ad&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">score_list</span><span class="p">,</span> <span class="n">cellbarcode_info</span><span class="p">,</span> <span class="n">gene_set_names</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">calculateScoreParallel_multiBatch</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">marker_gene</span><span class="o">=</span><span class="n">marker_gene</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">marker_gene_n_groups_indices</span><span class="o">=</span><span class="n">marker_gene_n_groups_indices</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">score_layer</span><span class="o">=</span><span class="s1">&#39;piaso&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">max_workers</span><span class="o">=</span><span class="mi">8</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">cellbarcode_info</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.predictCellTypeByGDR">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">predictCellTypeByGDR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adata_ref</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'log1p'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer_reference</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'log1p'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'CellTypes'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query_groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Leiden'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_integration</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoring_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.predictCellTypeByGDR" title="Link to this definition">#</a></dt>
<dd><p>Predicts cell types in a query dataset (<cite>adata</cite>) using the GDR dimensionality reduction method based on a reference dataset (<cite>adata_ref</cite>).
To use GDR for dimensionality reduction, please refer to <cite>piaso.tl.runGDR</cite> or <cite>piaso.tl.runGDRParallel</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – The query single-cell AnnData object for which cell types are to be predicted.</p></li>
<li><p><strong>adata_ref</strong> (<em>AnnData</em>) – The reference single-cell AnnData object with known cell type annotations.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'log1p'</em><em>)</em>) – The layer in <cite>adata</cite> to use for gene expression data. If <cite>None</cite>, uses the <cite>.X</cite> matrix.</p></li>
<li><p><strong>layer_reference</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'log1p'</em><em>)</em>) – The layer in <cite>adata_ref</cite> to use for reference gene expression data. If <cite>None</cite>, uses the <cite>.X</cite> matrix.</p></li>
<li><p><strong>reference_groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'CellTypes'</em><em>)</em>) – The column in <cite>adata_ref.obs</cite> used to define reference cell type groupings.</p></li>
<li><p><strong>query_groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: 'Leiden'</em><em>)</em>) – The column in <cite>adata.obs</cite> used to for GDR dimensionality reduction, such as clusters identified using Leiden or Louvain algorithms.</p></li>
<li><p><strong>mu</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 10.0</em><em>)</em>) – A regularization parameter for controlling the gene expression specificity, used in COSG (marker gene identification) and GDR.</p></li>
<li><p><strong>n_genes</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 15</em><em>)</em>) – The number of top specific genes per group, used in COSG and GDR.</p></li>
<li><p><strong>return_integration</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – If <cite>True</cite>, the function will return the integrated low-dimensional cell embeddings of the query dataset and reference dataset.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to use highly variable genes, used in GDR.</p></li>
<li><p><strong>n_highly_variable_genes</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 5000</em><em>)</em>) – The number of highly variable genes to select, if <cite>use_highly_variable</cite> is <cite>True</cite>, used in GDR.</p></li>
<li><p><strong>n_svd_dims</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 50</em><em>)</em>) – The number of dimensions to retain during SVD, used in GDR.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – Resolution parameter for clustering, used in GDR.</p></li>
<li><p><strong>scoring_method</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – The method used for gene set scoring, used in GDR.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – A key to add the predicted cell types or integration results to <cite>adata.obs</cite>. If <cite>None</cite>, <cite>CellTypes_gdr</cite> will be used.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – The level of logging output. Higher values produce more detailed logs for debugging and monitoring progress.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>return_integration</cite> is <cite>True</cite>, returns an AnnData object of merged reference and query datasets with integrated
cell embeddings and predicted cell types. Otherwise, updates <cite>adata</cite> in place with the predicted cell types.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None or AnnData</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load query dataset</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;query_data.h5ad&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Load reference dataset with known cell type annotations</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata_ref</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;reference_data.h5ad&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Predict cell types for the query dataset</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">predictCellTypeByGDR</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">adata_ref</span><span class="o">=</span><span class="n">adata_ref</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;log1p&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">layer_reference</span><span class="o">=</span><span class="s1">&#39;log1p&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">reference_groupby</span><span class="o">=</span><span class="s1">&#39;CellTypes&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">query_groupby</span><span class="o">=</span><span class="s1">&#39;Leiden&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">mu</span><span class="o">=</span><span class="mf">10.0</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">n_genes</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">return_integration</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">n_highly_variable_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">n_svd_dims</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">resolution</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;CellTypes_gdr&#39;</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the predicted cell types in the query dataset</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;CellTypes_gdr&#39;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runCOSGParallel">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runCOSGParallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_gene_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runCOSGParallel" title="Link to this definition">#</a></dt>
<dd><p>Run COSG on batches in parallel using shared memory and multiprocessing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em>) – The key in <cite>adata.obs</cite> used to identify batches.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – The key in <cite>adata.obs</cite> used to group observations for clustering. If None, clustering will be performed.</p></li>
<li><p><strong>n_svd_dims</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 50</em><em>)</em>) – Number of SVD components to compute.</p></li>
<li><p><strong>n_highly_variable_genes</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 5000</em><em>)</em>) – Number of highly variable genes to use for SVD.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 0</em><em>)</em>) – Level of verbosity for logging information.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – Resolution parameter for clustering.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – Layer of the <cite>adata</cite> object to use for COSG.</p></li>
<li><p><strong>infog_layer</strong> (<em>str</em><em>, </em><em>optional</em><em> (</em><em>default: None</em><em>)</em>) – If specified, the INFOG normalization will be calculated using this layer of <cite>adata.layers</cite>, which is expected to contain the UMI count matrix. Defaults to None.</p></li>
<li><p><strong>mu</strong> (<em>float</em><em>, </em><em>optional</em><em> (</em><em>default: 1.0</em><em>)</em>) – COSG parameter to control regularization.</p></li>
<li><p><strong>n_gene</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 30</em><em>)</em>) – Number of marker genes to compute for each cluster.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: True</em><em>)</em>) – Whether to use highly variable genes for SVD.</p></li>
<li><p><strong>return_gene_names</strong> (<em>bool</em><em>, </em><em>optional</em><em> (</em><em>default: False</em><em>)</em>) – Whether to return gene names instead of indices in the marker gene DataFrame.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em><em> (</em><em>default: 8</em><em>)</em>) – Maximum number of parallel workers to use. If None, defaults to the number of available CPU cores.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Combined marker gene DataFrame with batch-specific suffixes.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>DataFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;example_data.h5ad&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">marker_genes</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runCOSGParallel</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s1">&#39;batch&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_svd_dims</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_highly_variable_genes</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">resolution</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;log1p&#39;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">mu</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_gene</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">return_gene_names</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">max_workers</span><span class="o">=</span><span class="mi">4</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">marker_genes</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runGDR">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runGDR</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoring_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runGDR" title="Link to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runGDRParallel">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runGDRParallel</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groupby</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_gene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_highly_variable_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">5000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_svd_dims</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resolution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scoring_method</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_workers</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">calculate_score_multiBatch</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runGDRParallel" title="Link to this definition">#</a></dt>
<dd><p>Run GDR (marker Gene-guided dimensionality reduction) in parallel using multi-cores and shared memeory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – Annotated data matrix.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obs</cite> representing batch information. Defaults to None. If specified, different batches will be processed separately and in parallel, otherwise, the input data will be processed as one batch.</p></li>
<li><p><strong>groupby</strong> (<em>str</em><em>, </em><em>optional</em>) – Key in <cite>adata.obs</cite> to specify which cell group information to use. Defaults to None. If none, de novo clustering will be performed.</p></li>
<li><p><strong>n_gene</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of genes, parameter used in COSG. Defaults to 30.</p></li>
<li><p><strong>mu</strong> (<em>float</em><em>, </em><em>optional</em>) – Gene expression specificity parameter, used in COSG. Defaults to 1.0.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em>) – Layer in <cite>adata.layers</cite> to use for the analysis. Defaults to None, which uses <cite>adata.X</cite>.</p></li>
<li><p><strong>score_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – If specified, the gene scoring will be calculated using this layer of <cite>adata.layers</cite>. Defaults to None.</p></li>
<li><p><strong>infog_layer</strong> (<em>str</em><em>, </em><em>optional</em>) – If specified, the INFOG normalization will be calculated using this layer of <cite>adata.layers</cite>, which is expected to contain the UMI count matrix. Defaults to None.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use only highly variable genes when rerunning the dimensionality reduction. Defaults to True. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_highly_variable_genes</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of highly variable genes to use when <cite>use_highly_variable</cite> is True. Defaults to 5000. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>n_svd_dims</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of dimensions to use for SVD. Defaults to 50. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>resolution</strong> (<em>float</em><em>, </em><em>optional</em>) – Resolution parameter for de novo clustering. Defaults to 1.0. Only effective when <cite>groupby=None</cite>.</p></li>
<li><p><strong>scoring_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Specifies the gene set scoring method used to compute gene scores.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em>) – Key under which the GDR dimensionality reduction results will be stored in <cite>adata.obsm</cite>. If None, results will be saved to <cite>adata.obsm[X_gdr]</cite>.</p></li>
<li><p><strong>max_workers</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum number of workers to use for parallel computation. Defaults to 8.</p></li>
<li><p><strong>calculate_score_multiBatch</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to calculate gene scores across multiple adata batches (if <cite>batch_key</cite> is specified). Defaults to False.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em>) – Verbosity level of the function. Higher values provide more detailed logs. Defaults to 0.</p></li>
<li><p><strong>random_seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for reproducibility. Default is 1927.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function modifies <cite>adata</cite> in place by adding GDR dimensionality reduction result to <cite>adata.obsm[key_added]</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;example.h5ad&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runGDRParallel</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;batch&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;CellTypes&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">n_gene</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">max_workers</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">verbosity</span><span class="o">=</span><span class="mi">0</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_gdr&quot;</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runSVD">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runSVD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">AnnData</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_svd'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runSVD" title="Link to this definition">#</a></dt>
<dd><p>Performs Truncated Singular Value Decomposition (SVD) on the specified gene expression matrix (adata.X or a specified layer)
within an AnnData object and stores the resulting low-dimensional representation in <cite>adata.obsm</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – An AnnData object.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – If True, the decomposition is performed only on highly variable genes/features.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=50</em>) – The number of principal components to retain.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=10</em>) – A random seed to ensure reproducibility.</p></li>
<li><p><strong>scale_data</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, standardizes the input data before performing SVD.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='X_svd'</em>) – The key under which the resulting cell embeddings are stored in <cite>adata.obsm</cite>.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies which layer of <cite>adata</cite> to use for the transformation. If None, <cite>adata.X</cite> is used.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=0</em>) – Controls the verbosity of logging messages.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The function modifies <cite>adata</cite> in place, storing the cell embeddings in <cite>adata.obsm[key_added]</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runSVD</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
<span class="gp">... </span>       <span class="n">scale_data</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s1">&#39;X_svd&#39;</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the transformed data</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_svd&#39;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="piaso.tools.runSVDLazy">
<span class="sig-prename descclassname"><span class="pre">piaso.tools.</span></span><span class="sig-name descname"><span class="pre">runSVDLazy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">adata</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_components</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_highly_variable</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_top_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbosity</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_state</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1927</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scale_data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_trim</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_added</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'X_svd'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">infog_layer</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#piaso.tools.runSVDLazy" title="Link to this definition">#</a></dt>
<dd><p>Performs Truncated Singular Value Decomposition (SVD) in a “lazy” mode, based on the <cite>piaso.tl.runSVD</cite> function.</p>
<p>Compared to <cite>piaso.tl.runSVD</cite>, this function includes the step of highly variable gene section. If <cite>layer</cite> is set to <cite>infog</cite>,
both the highly variable genes and normalized gene expression values were taken from the INFOG normalization outputs.</p>
<p>This function performs on the specified gene expression matrix (adata.X or a specified layer) within an AnnData object
and stores the resulting low-dimensional representation in <cite>adata.obsm</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adata</strong> (<em>AnnData</em>) – An AnnData object.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, returns a copy of <cite>adata</cite> with the computed embeddings instead of modifying in place.</p></li>
<li><p><strong>n_components</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=50</em>) – The number of singular value decomposition (SVD) components to retain.</p></li>
<li><p><strong>use_highly_variable</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – If True, uses only highly variable genes for the decomposition.</p></li>
<li><p><strong>n_top_genes</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=3000</em>) – The number of top highly variable genes to retain before performing SVD.</p></li>
<li><p><strong>verbosity</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=0</em>) – Controls the verbosity of logging messages.</p></li>
<li><p><strong>batch_key</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies the key in <cite>adata.obs</cite> containing batch labels for <cite>batch_key</cite> used in the <cite>sc.pp.highly_variable_genes</cite> function.</p></li>
<li><p><strong>random_state</strong> (<em>int</em><em>, </em><em>optional</em><em>, </em><em>default=1927</em>) – A random seed to ensure reproducibility.</p></li>
<li><p><strong>scale_data</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=False</em>) – If True, standardizes the input data before performing SVD.</p></li>
<li><p><strong>infog_trim</strong> (<em>bool</em><em>, </em><em>optional</em><em>, </em><em>default=True</em>) – Used for the <cite>trim</cite> parameter in <cite>piaso.tl.infog</cite> function, effective only when <cite>layer</cite> set to <cite>infog</cite>.</p></li>
<li><p><strong>key_added</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default='X_svd'</em>) – The key under which the resulting cell embeddings are stored in <cite>adata.obsm</cite>.</p></li>
<li><p><strong>layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Specifies which layer of <cite>adata</cite> to use for the transformation. If None, <cite>adata.X</cite> is used.</p></li>
<li><p><strong>infog_layer</strong> (<em>str</em><em>, </em><em>optional</em><em>, </em><em>default=None</em>) – Used for the <cite>layer</cite> parameter in <cite>piaso.tl.infog</cite> function, effective only when <cite>layer</cite> set to <cite>infog</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>If <cite>copy</cite> is True, returns a modified AnnData object. Otherwise, modifies <cite>adata</cite> in place.</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">piaso</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span> <span class="o">=</span> <span class="n">piaso</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">runSVDLazy</span><span class="p">(</span>
<span class="gp">... </span>    <span class="n">adata</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">use_highly_variable</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;X_svd&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">layer</span><span class="o">=</span><span class="kc">None</span>
<span class="gp">... </span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Access the cell embedding</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_svd&#39;</span><span class="p">]</span>
</pre></div>
</div>
</dd></dl>

</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="piaso.preprocessing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">piaso.preprocessing package</p>
      </div>
    </a>
    <a class="right-next"
       href="references.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">References</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.leiden_local"><code class="docutils literal notranslate"><span class="pre">leiden_local()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.infog"><code class="docutils literal notranslate"><span class="pre">infog()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.score"><code class="docutils literal notranslate"><span class="pre">score()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.calculateScoreParallel"><code class="docutils literal notranslate"><span class="pre">calculateScoreParallel()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.calculateScoreParallel_multiBatch"><code class="docutils literal notranslate"><span class="pre">calculateScoreParallel_multiBatch()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.predictCellTypeByGDR"><code class="docutils literal notranslate"><span class="pre">predictCellTypeByGDR()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runCOSGParallel"><code class="docutils literal notranslate"><span class="pre">runCOSGParallel()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runGDR"><code class="docutils literal notranslate"><span class="pre">runGDR()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runGDRParallel"><code class="docutils literal notranslate"><span class="pre">runGDRParallel()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runSVD"><code class="docutils literal notranslate"><span class="pre">runSVD()</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#piaso.tools.runSVDLazy"><code class="docutils literal notranslate"><span class="pre">runSVDLazy()</span></code></a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2025, Min Dai.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>