Search.setIndex({"alltitles": {"API": [[2, "api"]], "Built-in color palettes in PIASO": [[8, "Built-in-color-palettes-in-PIASO"]], "Cell type prediction by GDR": [[4, "Cell-type-prediction-by-GDR"]], "Citation": [[0, "citation"]], "Cluster the data": [[7, "Cluster-the-data"]], "Cluster the data with Leiden algorithm": [[6, "Cluster-the-data-with-Leiden-algorithm"]], "Cluster the query data": [[4, "Cluster-the-query-data"]], "Combine categorical variables": [[3, "Combine-categorical-variables"]], "Continuous color palettes": [[8, "Continuous-color-palettes"]], "Count the number of cells in each category": [[3, "Count-the-number-of-cells-in-each-category"]], "Development version": [[1, "development-version"]], "Dimensionality reduction and visualization": [[7, "Dimensionality-reduction-and-visualization"]], "Dimensionality reduction with GDRParallel": [[5, "Dimensionality-reduction-with-GDRParallel"]], "Discrete color palettes": [[8, "Discrete-color-palettes"]], "Finer clustering on selected cluster(s)": [[6, "Finer-clustering-on-selected-cluster(s)"]], "GDR: marker Gene-guided Dimensionality Reduction": [[5, "GDR:-marker-Gene-guided-Dimensionality-Reduction"]], "INFOG normalization": [[5, "INFOG-normalization"]], "Identify marker genes for every cluster": [[7, "Identify-marker-genes-for-every-cluster"]], "Identify markger genes in subclusters": [[6, "Identify-markger-genes-in-subclusters"]], "Indices and tables": [[0, "indices-and-tables"]], "Installation": [[1, "installation"]], "Introduction": [[7, "Introduction"]], "Load the data": [[3, "Load-the-data"], [4, "Load-the-data"], [5, "Load-the-data"], [6, "Load-the-data"], [7, "Load-the-data"], [8, "Load-the-data"]], "Load the query dataset": [[4, "Load-the-query-dataset"]], "Load the reference dataset": [[4, "Load-the-reference-dataset"]], "Normalize and log-transform the data": [[7, "Normalize-and-log-transform-the-data"]], "PIASO: Precise Integrative Analysis of Single-cell Omics": [[0, "piaso-precise-integrative-analysis-of-single-cell-omics"]], "Perform clustering across all selected groups": [[6, "Perform-clustering-across-all-selected-groups"]], "Perform clustering independently within each group": [[6, "Perform-clustering-independently-within-each-group"]], "Plotting functions in PIASO": [[3, "Plotting-functions-in-PIASO"]], "Predict cell types by GDR": [[4, "Predict-cell-types-by-GDR"], [7, "Predict-cell-types-by-GDR"]], "References": [[13, "References"]], "Release 1.0.0 (released February 24, 2025)": [[14, "release-1-0-0-released-february-24-2025"]], "Release notes": [[14, "release-notes"]], "Run GDR for dimensionality reduction": [[7, "Run-GDR-for-dimensionality-reduction"]], "Run leiden_local on multiple clusters": [[6, "Run-leiden_local-on-multiple-clusters"]], "Run leiden_local on one cluster": [[6, "Run-leiden_local-on-one-cluster"]], "Save output plots": [[3, "Save-output-plots"]], "Split the UMAP by condition": [[3, "Split-the-UMAP-by-condition"]], "Stable version": [[1, "stable-version"]], "Subpackages": [[9, "subpackages"]], "Tutorials": [[15, "tutorials"]], "Violin plots of entire data": [[3, "Violin-plots-of-entire-data"]], "Violin plots of grouped data": [[3, "Violin-plots-of-grouped-data"]], "Visualize GDR results with UMAPs": [[5, "Visualize-GDR-results-with-UMAPs"]], "Visualize with PCA-based UMAP": [[5, "Visualize-with-PCA-based-UMAP"]], "Visualize with UMAP": [[6, "Visualize-with-UMAP"]], "Visualize with UMAPs": [[4, "Visualize-with-UMAPs"]], "Visualize with a continuous color map": [[3, "Visualize-with-a-continuous-color-map"]], "Visualize with a discrete color map": [[3, "Visualize-with-a-discrete-color-map"]], "Visualizing categorical variables (Cell subclass) across different experimental conditions": [[3, "Visualizing-categorical-variables-(Cell-subclass)-across-different-experimental-conditions"]], "Visualizing the expression of a specific gene across two different experimental conditions": [[3, "Visualizing-the-expression-of-a-specific-gene-across-two-different-experimental-conditions"]], "Visulaizing continuous variables (Number of UMIs) across different experimental conditions": [[3, "Visulaizing-continuous-variables-(Number-of-UMIs)-across-different-experimental-conditions"]], "piaso package": [[9, "module-piaso"]], "piaso.plotting package": [[10, "module-piaso.plotting"]], "piaso.preprocessing package": [[11, "module-piaso.preprocessing"]], "piaso.tools package": [[12, "module-piaso.tools"]]}, "docnames": ["index", "installation", "modules", "notebooks/01_showThePlotFunction", "notebooks/GDRPredictCellTypeTutorial", "notebooks/GDR_Tutorial", "notebooks/LeidenLocalTutorial", "notebooks/PIASO_tutorial", "notebooks/colorPalettes", "piaso", "piaso.plotting", "piaso.preprocessing", "piaso.tools", "references", "releaseNotes", "tutorials"], "envversion": {"nbsphinx": 4, "sphinx": 61, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinxcontrib.bibtex": 9}, "filenames": ["index.rst", "installation.md", "modules.rst", "notebooks/01_showThePlotFunction.ipynb", "notebooks/GDRPredictCellTypeTutorial.ipynb", "notebooks/GDR_Tutorial.ipynb", "notebooks/LeidenLocalTutorial.ipynb", "notebooks/PIASO_tutorial.ipynb", "notebooks/colorPalettes.ipynb", "piaso.rst", "piaso.plotting.rst", "piaso.preprocessing.rst", "piaso.tools.rst", "references.ipynb", "releaseNotes.rst", "tutorials.rst"], "indexentries": {"calculatescoreparallel() (in module piaso.tools)": [[12, "piaso.tools.calculateScoreParallel", false]], "calculatescoreparallel_multibatch() (in module piaso.tools)": [[12, "piaso.tools.calculateScoreParallel_multiBatch", false]], "createcustomcmapfromhex() (in module piaso.plotting)": [[10, "piaso.plotting.createCustomCmapFromHex", false]], "getcrosscategories() (in module piaso.preprocessing)": [[11, "piaso.preprocessing.getCrossCategories", false]], "infog() (in module piaso.tools)": [[12, "piaso.tools.infog", false]], "leiden_local() (in module piaso.tools)": [[12, "piaso.tools.leiden_local", false]], "module": [[9, "module-piaso", false], [10, "module-piaso.plotting", false], [11, "module-piaso.preprocessing", false], [12, "module-piaso.tools", false]], "piaso": [[9, "module-piaso", false]], "piaso.plotting": [[10, "module-piaso.plotting", false]], "piaso.preprocessing": [[11, "module-piaso.preprocessing", false]], "piaso.tools": [[12, "module-piaso.tools", false]], "plot_embeddings_split() (in module piaso.plotting)": [[10, "piaso.plotting.plot_embeddings_split", false]], "plot_features_violin() (in module piaso.plotting)": [[10, "piaso.plotting.plot_features_violin", false]], "predictcelltypebygdr() (in module piaso.tools)": [[12, "piaso.tools.predictCellTypeByGDR", false]], "runcosgparallel() (in module piaso.tools)": [[12, "piaso.tools.runCOSGParallel", false]], "rungdr() (in module piaso.tools)": [[12, "piaso.tools.runGDR", false]], "rungdrparallel() (in module piaso.tools)": [[12, "piaso.tools.runGDRParallel", false]], "runsvd() (in module piaso.tools)": [[12, "piaso.tools.runSVD", false]], "runsvdlazy() (in module piaso.tools)": [[12, "piaso.tools.runSVDLazy", false]], "score() (in module piaso.tools)": [[12, "piaso.tools.score", false]], "table() (in module piaso.preprocessing)": [[11, "piaso.preprocessing.table", false]]}, "objects": {"": [[9, 0, 0, "-", "piaso"]], "piaso": [[10, 0, 0, "-", "plotting"], [11, 0, 0, "-", "preprocessing"], [12, 0, 0, "-", "tools"]], "piaso.plotting": [[10, 1, 1, "", "createCustomCmapFromHex"], [10, 1, 1, "", "plot_embeddings_split"], [10, 1, 1, "", "plot_features_violin"]], "piaso.preprocessing": [[11, 1, 1, "", "getCrossCategories"], [11, 1, 1, "", "table"]], "piaso.tools": [[12, 1, 1, "", "calculateScoreParallel"], [12, 1, 1, "", "calculateScoreParallel_multiBatch"], [12, 1, 1, "", "infog"], [12, 1, 1, "", "leiden_local"], [12, 1, 1, "", "predictCellTypeByGDR"], [12, 1, 1, "", "runCOSGParallel"], [12, 1, 1, "", "runGDR"], [12, 1, 1, "", "runGDRParallel"], [12, 1, 1, "", "runSVD"], [12, 1, 1, "", "runSVDLazy"], [12, 1, 1, "", "score"]]}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "objtypes": {"0": "py:module", "1": "py:function"}, "terms": {"": [0, 4, 5, 7, 12, 13, 15], "0": [3, 4, 5, 6, 7, 8, 10, 12], "00": [4, 5, 6, 7], "01": [4, 6, 7], "01774": 13, "018": 4, "02": [5, 7], "024": 13, "03": 7, "04": 6, "05": [4, 5, 10], "06": 4, "060": 7, "07": [5, 7, 13], "08": 7, "09": [5, 7], "1": [1, 3, 4, 5, 6, 7, 8, 10, 12], "10": [3, 4, 5, 6, 7, 10, 12, 13], "100": [3, 5, 6, 7, 12], "1000": 12, "10206": 6, "1038": 13, "1080": 3, "1093": 13, "11": [3, 4, 5, 6, 7], "1101": [7, 13], "11039": 6, "12": [3, 4, 5, 6, 7], "120": 4, "1299": 3, "12gngenuh2g9if975u8prtx3vpav5rbnk": 5, "13": [3, 4, 5, 6, 7, 8], "14": [3, 4, 5, 6, 7, 10], "140": 3, "14346": 4, "15": [3, 4, 5, 6, 7, 8, 12], "1551": 3, "156": 7, "157": 3, "16": [3, 4, 5, 6, 7], "1663": 3, "17": [3, 4, 6, 7], "17412": 5, "1797": 3, "18": [3, 4, 6, 7], "19": [3, 4, 7, 8], "191": 6, "1927": 12, "1b2ybvwcuvih0987zdniezkdz138x6zti": 4, "1e4": 7, "1min": 7, "1nf0irtgffqcecm6hycpptiykxs65k2p4": 4, "1nh": [3, 7, 8], "1pdbigpveo": [4, 6], "2": [1, 3, 4, 5, 6, 7, 8, 10, 12, 13], "20": [3, 5, 7, 12, 13], "200": 7, "20000": [3, 4, 6, 7, 8], "2022": [7, 13], "2024": [0, 4, 6, 7, 13], "2025": [4, 7], "20k": [3, 4, 6, 7, 8], "21": [3, 5, 7], "22": [3, 4, 5, 7], "224": 3, "23": [3, 4, 5, 7, 13], "231": 3, "2366": 13, "2383": 13, "2388": 3, "24": [3, 4, 6, 7, 8], "25": [3, 4, 7, 12], "256": 10, "26": [3, 4, 5, 6, 7], "26205": 5, "27": [3, 4, 7, 13], "279": 7, "28": [3, 5, 7], "29": [3, 7], "299": 3, "2d": 12, "3": [3, 4, 5, 6, 7, 8], "30": [3, 6, 7, 12], "300": [3, 4, 5, 6, 7, 8], "3000": [5, 7, 12], "305": 3, "31": [3, 7], "32": [3, 4, 5, 7], "329": 3, "33": [3, 7], "34": [3, 6, 7], "35": [3, 5, 7], "353": 4, "354": 3, "36": 3, "36601": [3, 4, 6, 7, 8], "37": 3, "387": 4, "39": [3, 4, 5, 6, 7, 8], "3min": 7, "4": [3, 4, 5, 6, 7, 8, 12], "40": [5, 12], "41": 5, "42": 12, "423": 7, "43": 7, "44": 7, "451": 7, "46": 5, "479": 7, "48": 5, "4826": 3, "496": 3, "5": [3, 4, 5, 6, 7, 8, 10, 13], "50": [4, 5, 7, 12], "5000": [4, 5, 7, 12], "51": 7, "513": 7, "5168": 6, "525": 7, "55": 4, "564": 4, "572266": 10, "595": 7, "6": [3, 4, 5, 6, 7, 8, 10], "60": 5, "604399": 13, "604399v1": 7, "605": 3, "610": 4, "63": 5, "638": 3, "6493": 6, "653": 3, "661": 7, "692": 7, "7": [3, 4, 5, 6, 7, 8], "70": [5, 7], "72310972213745": 5, "73": 5, "76": 7, "76560": 4, "767": 6, "77": 7, "780": 3, "79": 5, "8": [3, 4, 5, 6, 7, 8, 12], "80": [3, 4, 5, 6, 7, 8, 10], "805": 4, "81": 7, "839": 4, "84": 7, "853": 6, "859": 6, "86": 3, "9": [3, 4, 5, 6, 7, 8, 10], "9092": 6, "91": 4, "92": [5, 6], "96": [5, 7, 8], "97": 5, "A": [10, 11, 12], "But": 7, "By": 3, "For": [7, 12], "IT": 3, "If": [0, 1, 7, 10, 11, 12], "In": [3, 4, 5, 6, 7], "No": 7, "The": [3, 4, 5, 6, 7, 8, 10, 11, 12], "To": [3, 4, 6, 7, 8, 12], "_cluster": 6, "_normal": [5, 7], "_rungdr": [4, 7], "about": [4, 6, 7], "absent": 3, "access": 12, "accord": [3, 10], "accur": 13, "accuraci": 4, "achiev": [6, 7], "across": 12, "action": [3, 5, 8], "activ": [0, 1], "actual": [1, 6, 7], "ad": [3, 4, 5, 6, 7, 8, 12], "ad_rna_mtg_subsample_excludereference_20k_piaso": [3, 4, 7, 8], "ad_rna_mtg_subsample_excludereference_20k_piaso_preprocess": [4, 6], "adata": [3, 4, 5, 6, 7, 8, 10, 12], "adata_combin": [4, 7], "adata_i": 6, "adata_path": 5, "adata_ref": [4, 7, 12], "adata_test": 7, "add": 12, "addit": [5, 10], "adgrg7": 3, "adjac": [5, 6, 7], "adjust": [3, 7], "adultcortexmultiomerna_integrated_anno": 5, "affect": [3, 4, 6, 7, 8], "african": [3, 4, 6, 7, 8], "after": [4, 7, 10], "ag": [3, 4, 6, 7, 8], "again": 7, "aibs_query_adata": 4, "aim": 8, "al": [0, 4, 6, 7, 13], "alaska": [3, 4, 6, 7, 8], "algorithm": 12, "align": [3, 4, 6, 7, 8], "all": [3, 4, 7, 12], "allen": [3, 8], "allow": 12, "along": 6, "alongsid": 3, "alreadi": [1, 7], "also": [3, 6, 7], "alter": [0, 13], "altern": 7, "alzheim": [3, 4, 6, 7, 8, 13], "american": [3, 4, 6, 7, 8], "among": 12, "amplified_quantity_ng": [3, 4, 6, 7, 8], "an": [1, 7, 10, 12], "analysi": [3, 4, 5, 6, 7, 8, 9, 12], "analyt": 7, "ani": [6, 7, 10], "anndata": [3, 4, 5, 6, 7, 8, 10, 12], "annot": [4, 6, 7, 10, 12], "ap": [3, 4, 6, 7, 8], "api": 0, "apo": [3, 4, 6, 7, 8], "apoe4": [3, 4, 6, 7, 8], "append": [3, 4, 5, 6, 7, 8], "appli": 10, "applic": [3, 7], "approach": 12, "ar": [4, 5, 6, 7, 10, 11, 12], "ar_id": [3, 4, 6, 7, 8], "aren": 7, "argument": 10, "arial": [4, 5, 6, 7], "around": 10, "arrai": 12, "arteriolosclerosi": [3, 4, 6, 7, 8], "as_datafram": [3, 11], "ascend": [3, 11], "asian": [3, 4, 6, 7, 8], "aspect": 10, "assess": 5, "assign": 10, "astrocyt": 3, "astyp": [4, 7], "atac_barcod": 5, "atac_chimeric_read": 5, "atac_confidently_mapped_read_pair": [3, 4, 6, 7, 8], "atac_dup_read": 5, "atac_fraction_of_genome_in_peak": [3, 4, 6, 7, 8], "atac_fraction_of_high_quality_fragments_in_cel": [3, 4, 6, 7, 8], "atac_fraction_of_high_quality_fragments_overlapping_peak": [3, 4, 6, 7, 8], "atac_fraction_of_high_quality_fragments_overlapping_tss": [3, 4, 6, 7, 8], "atac_fraction_of_transposition_events_in_peaks_in_cel": [3, 4, 6, 7, 8], "atac_frag": 5, "atac_lowmapq": 5, "atac_mean_raw_read_pairs_per_cel": [3, 4, 6, 7, 8], "atac_median_high_quality_fragments_per_cel": [3, 4, 6, 7, 8], "atac_mitochondrial_read": 5, "atac_non": [3, 4, 6, 7, 8], "atac_number_of_peak": [3, 4, 6, 7, 8], "atac_peak_region_cutsit": 5, "atac_peak_region_frag": 5, "atac_percent_dupl": [3, 4, 6, 7, 8], "atac_q30_bases_in_barcod": [3, 4, 6, 7, 8], "atac_q30_bases_in_read_1": [3, 4, 6, 7, 8], "atac_q30_bases_in_read_2": [3, 4, 6, 7, 8], "atac_q30_bases_in_sample_index_i1": [3, 4, 6, 7, 8], "atac_raw_read": 5, "atac_sequenced_read_pair": [3, 4, 6, 7, 8], "atac_tss_enrichment_scor": [3, 4, 6, 7, 8], "atac_tss_frag": 5, "atac_unmapped_read": 5, "atac_unmapped_read_pair": [3, 4, 6, 7, 8], "atac_valid_barcod": [3, 4, 6, 7, 8], "atherosclerosi": [3, 4, 6, 7, 8], "atla": [4, 6, 7, 13], "attribut": 10, "author": 7, "auto": [7, 10], "automat": 10, "autonotebook": 7, "avail": [0, 3, 4, 5, 6, 7, 8, 10, 12], "avg_size_bp": [3, 4, 6, 7, 8], "ax": [4, 7, 8, 10], "axi": [4, 7, 10], "b1257a": 10, "backend": [6, 7], "backup": 7, "ball": 12, "barcod": 12, "base": [6, 7, 10, 12], "bash": 7, "basi": [3, 4, 7, 8, 10], "batch": [4, 5, 12], "batch_condit": [3, 4, 6, 7, 8], "batch_kei": [5, 6, 7, 12], "batch_vendor_nam": [3, 4, 6, 7, 8], "bbab579": 13, "bc": [3, 4, 6, 7, 8], "been": 7, "befor": [7, 12], "benchmark": 5, "better": [3, 6], "between": 6, "bh_pval": 5, "bib": 13, "bin": 7, "binomi": 7, "bioinformat": 13, "biolog": 7, "biorxiv": [0, 7, 13], "black": [3, 4, 6, 7, 8], "bodi": [3, 4, 6, 7, 8], "bool": [10, 11, 12], "boolean": [7, 10], "both": [8, 10], "boundari": 12, "braak": [3, 4, 6, 7, 8], "braak_color": [3, 4, 6, 7, 8], "brain": [3, 4, 6, 7, 8], "brief": 13, "built": [6, 12, 15], "c": 8, "c_": 8, "c_color1": [3, 8], "c_color3": [4, 7], "c_color4": [3, 5, 7, 10], "caa": [3, 4, 6, 7, 8], "calb2": 3, "calcul": [4, 5, 7, 11, 12], "calculate_qc_metr": 7, "calculate_score_multibatch": [5, 12], "calculatescoreparallel": [2, 9, 12], "calculatescoreparallel_multibatch": [2, 9, 12], "can": [3, 4, 6, 7, 8, 10, 11], "captur": [6, 7], "car3": 3, "case": 5, "casi": [3, 4, 6, 7, 8], "cat": [4, 7], "categor": [4, 6, 7, 10, 11], "categori": [4, 5, 7, 8, 10, 11, 12], "categories_ord": [6, 7], "cdk9": 10, "ceil": [8, 10], "cell": [5, 6, 8, 9, 10, 12, 13, 15], "cell_depth": 5, "cell_prep_typ": [3, 4, 6, 7, 8], "cellbarcode_info": 12, "celltyp": [4, 5, 7, 12], "celltypes_color": 5, "celltypes_gdr": [4, 7, 12], "celltypes_tacco": 5, "celltypes_tacco_color": 5, "centroid": [4, 7], "cerad": [3, 4, 6, 7, 8], "chandeli": 3, "chang": [3, 4, 6, 7, 8], "change_color": [3, 4, 6, 7, 8], "check": 4, "chodl": 3, "choic": [3, 4, 6, 7, 8], "cite": 0, "clariti": 3, "class": [3, 4, 6, 7, 8], "clearli": 5, "clinic": [3, 4, 6, 7, 8], "close": 6, "cluster": [0, 5, 10, 12, 15], "cluster_color": 4, "clustering_typ": [6, 12], "cm": [4, 5, 6, 7], "cmap": [3, 4, 5, 6, 7, 8, 10], "code": 10, "cognit": [3, 4, 6, 7, 8], "col1": 11, "col2": 11, "col_siz": [8, 10], "collect": 12, "color": [4, 5, 6, 7, 10, 15], "color_map": [3, 4, 5, 6, 7, 8], "colorbar": [4, 5, 7], "colormap": 10, "column": [4, 7, 10, 11, 12], "com": [1, 3, 4, 6, 7, 8], "combin": [11, 12], "compar": [4, 6, 7], "complet": [4, 7], "compon": 12, "comput": [4, 5, 6, 7, 10, 12], "concat": [4, 7], "concaten": [4, 7], "conda": [1, 7], "condit": 0, "conditionxsubclass": 3, "confid": [3, 4, 6, 7, 8], "confusion_matrix": [4, 7], "confusion_matrix_df": [4, 7], "connect": [3, 4, 5, 6, 7, 8], "consid": [0, 12], "contain": [6, 10, 11, 12], "content": 7, "contin": 8, "continu": [4, 6, 7], "control": [3, 7, 11, 12], "converg": [4, 7], "convert": 10, "coordin": [5, 7, 10], "copi": [5, 6, 7, 12], "core": 12, "correspond": [10, 12], "cortex": 5, "cortic": [0, 13], "cosg": [4, 5, 6, 7, 12, 13], "could": 3, "count": [5, 7, 11, 12], "counts_sum": 5, "cpu": [4, 5, 6, 7, 12], "cratqfxj5pavpy8_huqn1nricakq2": [3, 7, 8], "creat": [1, 10], "createcustomcmapfromhex": [2, 9, 10], "cross": 11, "cross_species_clust": 4, "cross_species_cluster_color": 4, "cst3_embeddingssplit": 10, "ct": 3, "current": 0, "custom": [10, 11], "d": [3, 4, 6, 7, 8], "d96998": 10, "d_": 8, "d_color1": [3, 8], "d_color3": 3, "d_color4": [4, 5, 6, 7], "dai": [7, 13], "data": [0, 10, 12], "data1": [3, 7, 8], "data_dir": 4, "datafram": [3, 4, 6, 7, 10, 11, 12], "dataset": [0, 3, 5, 6, 7, 10, 12], "de": 12, "death": [3, 4, 6, 7, 8], "debug": 12, "decomposit": 12, "default": [4, 6, 7, 10, 11, 12], "default_embed": [3, 4, 6, 7, 8], "defin": [10, 11, 12], "del": 7, "delimit": 11, "demonstr": [3, 7], "dendrogram": [6, 7], "dendrogram_": [6, 7], "dendrogram_leiden": [5, 7], "dendrogram_leiden_loc": 5, "dendrogram_leiden_local_on": 6, "denot": 10, "deprec": [4, 7], "depth": 12, "deriv": 7, "descend": 11, "describ": [4, 6, 7], "destin": [4, 5, 7], "detail": [4, 6, 7, 12], "detect": [3, 4, 6, 7, 8, 10], "determin": [3, 10], "df": 11, "df_tmp": [6, 7], "diag": 5, "dict": [3, 10, 11, 12], "dictionari": [11, 12], "diff": 7, "differ": [6, 12], "differenti": 6, "dimens": [7, 12], "dimension": [0, 6, 12, 15], "direct": [6, 7], "directli": 1, "directori": 7, "disabl": [4, 6, 7], "diseas": [3, 4, 6, 7, 8, 13], "disease_color": [3, 4, 6, 7, 8], "displai": [8, 10], "distanc": [3, 4, 5, 6, 7, 8], "distinct": 6, "distribut": 3, "divers": 8, "divid": [5, 7, 10], "do": 7, "doe": 10, "doi": 13, "don": [1, 7], "done": 7, "donor": [3, 4, 6, 7, 8], "donor_color": 4, "dot": [3, 10], "dotplot": [3, 6, 7], "doublet": [3, 4, 6, 7, 8, 10], "download": [4, 5, 6, 7], "downstream": 7, "dpi": [3, 4, 5, 6, 7, 8, 10], "dpi_sav": [3, 4, 5, 6, 7, 8], "dr_method": [6, 12], "drive": [3, 4, 6, 7, 8], "drive_link": [3, 4, 6, 7, 8], "dure": 12, "e": [0, 6, 7, 10, 12], "e8aebc": 10, "each": [5, 7, 10, 12], "educ": [3, 4, 6, 7, 8], "effect": [5, 12], "effici": 12, "either": [6, 12], "elimin": 7, "els": 8, "embed": [0, 3, 4, 5, 7, 8, 10, 12], "en": [4, 7], "enabl": 12, "encount": [5, 7], "end_tim": 5, "endotheli": 3, "enhanc": 3, "enrich": 12, "ensur": [7, 12], "enumer": 7, "env": 7, "environ": [1, 7], "equal": [10, 12], "et": [0, 3, 4, 6, 7, 13], "eval": 7, "exampl": [10, 12], "example_data": 12, "exce": 7, "excluded_reason": 5, "exhibit": 6, "exist": 11, "exp_component_vendor_nam": [3, 4, 6, 7, 8], "expand": [3, 4, 6, 7, 8], "expc_cell_captur": [3, 4, 6, 7, 8], "expect": 12, "experiment": 7, "experiment_component_fail": [3, 4, 6, 7, 8], "explor": [3, 4, 6, 7, 8], "express": [6, 7, 10, 12], "expressed_pct": [6, 7], "extract": 7, "facecolor": [3, 4, 5, 6, 7, 8], "facs_population_plan": [3, 4, 6, 7, 8], "factor": 10, "faefef": 10, "fals": [3, 4, 5, 6, 7, 8, 10, 11, 12], "famili": [4, 5, 6, 7], "fan": [3, 4, 6, 7, 8], "fast": 13, "fearur": 7, "featur": [0, 3, 7, 10, 12], "feature_list": [3, 10], "feature_typ": 5, "fezf2": 3, "fig": 8, "figdir": [3, 8], "figsiz": [3, 4, 5, 6, 7, 8, 10], "figur": [3, 4, 5, 6, 7, 8, 10], "file": [3, 4, 5, 6, 7, 8, 10], "filter": [6, 7], "filter_cel": 7, "finer": 15, "finish": [4, 5, 6, 7], "first": [1, 5, 6, 7, 11], "fit": 7, "fix": 10, "fix_coordinate_ratio": 10, "flavor": [4, 6, 7], "flexibl": 12, "float": [6, 7, 10, 12], "follow": 7, "font": [4, 5, 6, 7, 10], "font_manag": [4, 6, 7], "format": 10, "found": [4, 6, 7], "fraction": [3, 4, 6, 7, 8], "frameon": [3, 4, 5, 6, 7, 8], "frequent": 3, "fresh": [3, 4, 6, 7, 8], "from": [1, 3, 4, 5, 6, 7, 8, 10, 11, 12], "function": [0, 6, 8, 10, 11, 12, 15], "further": 7, "futur": [4, 6, 7], "futurewarn": [3, 4, 5, 6, 7, 8], "g": [0, 6, 7, 10], "gabitto": [4, 6, 7, 13], "gad1": [3, 8], "gad2": 3, "gdr": [0, 12, 15], "gdr_resolut": 12, "gdrive": [4, 5, 7], "gender": [3, 4, 6, 7, 8], "gene": [0, 4, 8, 10, 12, 13, 15], "gene_depth": 5, "gene_id": [3, 4, 5, 6, 7, 8], "gene_list": 12, "gene_set": 12, "gene_set_nam": 12, "gene_weight": 12, "genecel": [1, 7], "gener": [10, 11], "genom": [3, 4, 5, 6, 7, 8], "genotyp": [3, 4, 6, 7, 8], "get_n_split": 7, "getattr": 8, "getcrosscategori": [2, 3, 9, 11], "getlogg": [4, 6, 7], "gex_barcod": 5, "gex_conf_exonic_antisense_read": 5, "gex_conf_exonic_dup_read": 5, "gex_conf_exonic_read": 5, "gex_conf_exonic_unique_read": 5, "gex_conf_intergenic_read": 5, "gex_conf_intronic_antisense_read": 5, "gex_conf_intronic_dup_read": 5, "gex_conf_intronic_read": 5, "gex_conf_intronic_unique_read": 5, "gex_conf_txomic_unique_read": 5, "gex_estimated_number_of_cel": [3, 4, 6, 7, 8], "gex_exonic_umi": 5, "gex_fraction_of_transcriptomic_reads_in_cel": [3, 4, 6, 7, 8], "gex_genes_count": 5, "gex_intronic_umi": 5, "gex_mapped_read": 5, "gex_mean_raw_reads_per_cel": [3, 4, 6, 7, 8], "gex_median_genes_per_cel": [3, 4, 6, 7, 8], "gex_median_umi_counts_per_cel": [3, 4, 6, 7, 8], "gex_number_of_read": [3, 4, 6, 7, 8], "gex_percent_dupl": [3, 4, 6, 7, 8], "gex_q30_bases_in_barcod": [3, 4, 6, 7, 8], "gex_q30_bases_in_read_2": [3, 4, 6, 7, 8], "gex_q30_bases_in_sample_index_i1": [3, 4, 6, 7, 8], "gex_q30_bases_in_sample_index_i2": [3, 4, 6, 7, 8], "gex_q30_bases_in_umi": [3, 4, 6, 7, 8], "gex_raw_read": 5, "gex_reads_mapped_antisense_to_gen": [3, 4, 6, 7, 8], "gex_reads_mapped_confidently_to_exonic_region": [3, 4, 6, 7, 8], "gex_reads_mapped_confidently_to_genom": [3, 4, 6, 7, 8], "gex_reads_mapped_confidently_to_intergenic_region": [3, 4, 6, 7, 8], "gex_reads_mapped_confidently_to_intronic_region": [3, 4, 6, 7, 8], "gex_reads_mapped_confidently_to_transcriptom": [3, 4, 6, 7, 8], "gex_reads_mapped_to_genom": [3, 4, 6, 7, 8], "gex_reads_with_tso": [3, 4, 6, 7, 8], "gex_sequenced_read_pair": [3, 4, 6, 7, 8], "gex_sequencing_satur": [3, 4, 6, 7, 8], "gex_total_genes_detect": [3, 4, 6, 7, 8], "gex_umis_count": 5, "gex_valid_barcod": [3, 4, 6, 7, 8], "gex_valid_umi": [3, 4, 6, 7, 8], "git": 1, "github": 1, "github_test": [3, 8], "given": [6, 7, 11, 12], "go": 7, "googl": [3, 4, 6, 7, 8], "gradient": 10, "granular": 12, "graph": 6, "great": [3, 4, 6, 7, 8], "greatapesmtg_rna_integrated_raw_onlyorthogonal_gdr_humansubset_aib": 4, "grid": [3, 4, 7, 10], "grossli": [3, 4, 6, 7, 8], "group": [5, 7, 10, 12], "groupbi": [3, 5, 6, 7, 10, 12], "gt": 7, "guid": [7, 12, 15], "h5ad": [3, 4, 5, 6, 7, 8, 12], "ha": 7, "handl": 12, "harmonypi": [4, 7], "have": [1, 3, 7, 8], "haven": 7, "hawaiian": [3, 4, 6, 7, 8], "head": 12, "heatmap": [4, 7], "height": [3, 10], "height_singl": [3, 10], "help": 5, "here": [3, 6, 7], "heterogen": 6, "hex": 10, "hex_color": 10, "hexadecim": 10, "higher": [7, 12], "highest": [3, 4, 6, 7, 8], "highli": [0, 5, 7, 12], "highly_vari": [4, 5, 6, 7], "highly_variable_gen": 7, "highly_variable_infog": 5, "highly_variable_intersect": [5, 7], "highly_variable_nbatch": [5, 7], "highly_variable_rank": [4, 5, 6, 7], "hispan": [3, 4, 6, 7, 8], "home": [3, 4, 5, 6, 7, 8], "horizont": 3, "how": [3, 4, 5, 6, 7], "html": [4, 7], "http": [1, 3, 4, 6, 7, 8, 13], "human": 4, "hvg": [4, 5, 6], "i": [0, 3, 4, 5, 6, 7, 8, 10, 11, 12, 13], "id": [3, 4, 6, 7, 8], "ident": [0, 7, 13], "identif": [12, 13], "identifi": [4, 5, 12], "idx": [6, 7], "ignor": [3, 5, 8, 12], "igraph": [4, 5, 6, 7], "implement": [6, 7], "impli": 7, "import": [3, 4, 5, 6, 7, 8, 10, 12], "importlib": [3, 8], "imshow": 10, "inch": 10, "includ": 0, "independ": 12, "index": [0, 4, 7, 12], "indian": [3, 4, 6, 7, 8], "indic": [10, 12], "infer": [0, 5], "info": [4, 6, 7], "info_factor": 5, "infog": [0, 2, 9, 12], "infog_lay": [5, 12], "infog_scor": 12, "infog_trim": 12, "inform": [5, 12], "initi": [4, 7], "inplac": 7, "input": [7, 12], "instal": [0, 7], "instead": [4, 6, 7, 12], "int": [7, 8, 10, 12], "integ": 10, "integr": [5, 9, 12, 13], "interest": 3, "interneuron": [0, 13], "interv": [3, 4, 6, 7, 8], "intregr": 7, "introduct": [4, 6, 15], "invalid": 7, "io": [4, 7], "iprogress": 7, "ipywidget": 7, "is_cel": 5, "isinst": [6, 7], "island": [3, 4, 6, 7, 8], "isn": 5, "item": [6, 7], "iter": [4, 5, 6, 7, 11], "iterate_by_second_column": 11, "iterrow": [6, 7], "j": [8, 13], "jitter": 10, "join": [5, 11], "jupyt": 7, "k": [6, 7, 13], "kd": 12, "kei": [10, 12], "key_ad": [4, 5, 6, 7, 12], "keyword": 10, "kmean": [4, 7], "knn": [5, 7], "know": [4, 7], "known": [4, 12], "kwarg": 10, "l2": 3, "l4": 3, "l5": 3, "l6": 3, "l6b": 3, "label": [4, 6, 7, 10, 12], "lamp5": [3, 6], "last": [3, 4, 6, 7, 8], "late": [3, 4, 6, 7, 8], "late_color": [3, 4, 6, 7, 8], "latest": [4, 7], "latino": [3, 4, 6, 7, 8], "layer": [3, 4, 5, 6, 7, 8, 10, 12], "layer_refer": [4, 7, 12], "lead": 7, "leaf": 12, "leaf_siz": 12, "learn": [4, 5, 6, 7], "legend": [3, 10], "legend_fontoutlin": [3, 4, 6, 7, 10], "legend_fonts": [3, 4, 6, 7, 10], "legend_fontweight": 3, "legend_loc": [3, 4, 6, 7, 8, 10], "legend_marker_s": 10, "leiden": [4, 5, 7, 12], "leiden_color": [4, 5], "leiden_last": 5, "leiden_last2": 5, "leiden_last2_color": 5, "leiden_last_color": 5, "leiden_loc": [2, 5, 9, 12], "leiden_local_color": 5, "leiden_local_many_al": 6, "leiden_local_many_each": 6, "leiden_local_on": 6, "leiden_sampl": 5, "leidenalg": [6, 7], "len": 8, "length": [8, 12], "less": [6, 7], "let": [6, 7], "level": [3, 4, 6, 7, 8, 12], "lewi": [3, 4, 6, 7, 8], "lhx6": [3, 6], "lib": 7, "librari": 10, "library_input_ng": [3, 4, 6, 7, 8], "library_prep": [3, 4, 6, 7, 8], "library_prep_pass_fail": [3, 4, 6, 7, 8], "limit": 10, "line": [3, 10], "linearsegmentedcolormap": 10, "linspac": 10, "list": [3, 6, 7, 10, 11, 12], "liter": 12, "load": [10, 12], "load_nam": [3, 4, 6, 7, 8], "local": [6, 12], "locat": 10, "log": [4, 5, 6, 12], "log1p": [4, 5, 6, 7, 12], "logarithm": 7, "louvain": [4, 5, 6, 10, 12], "low": [7, 12], "lower": 10, "lowli": 7, "lt": [5, 7], "lyse": 7, "m": [4, 6, 7, 13], "machin": 7, "made": 6, "mai": 6, "maintain": 3, "make": 7, "malat1": 3, "map": [4, 6, 7, 8], "mapping_dict": 3, "margin": 10, "margin_ratio": 10, "marker": [4, 6, 10, 12, 13, 15], "marker_gen": [7, 12], "marker_gene_n_groups_indic": 12, "marker_genes_list": [6, 7], "match": [4, 7, 12], "matplotlib": [3, 4, 5, 6, 7, 8, 10], "matrix": [5, 6, 7, 10, 12], "max_work": [5, 12], "maximum": [10, 12], "mean": [4, 5, 6, 7, 10, 12], "mean_count": [4, 5, 6], "memeori": 12, "memori": 12, "merfish": 0, "merg": 12, "metadata": [3, 4, 6, 7, 8], "method": [0, 3, 4, 5, 6, 7, 8, 12], "metric": [0, 4, 7, 10], "microglia": 3, "microinfarct": [3, 4, 6, 7, 8], "might": 7, "min_gen": 7, "minimum": 10, "mitochondri": [3, 4, 6, 7, 8], "mmse": [3, 4, 6, 7, 8], "moca": [3, 4, 6, 7, 8], "model": 7, "model_select": 7, "moder": 3, "modifi": [4, 6, 12], "modul": 0, "monitor": 12, "month": [3, 4, 6, 7, 8], "more": [4, 6, 7, 12], "mrna": 7, "mt": [4, 5, 6, 7], "mu": [4, 5, 6, 7, 12], "multi": 12, "multimap": [3, 4, 6, 7, 8], "multimod": 13, "multiom": 5, "multiome_feature_linkages_detect": [3, 4, 6, 7, 8], "multiome_linked_gen": [3, 4, 6, 7, 8], "multiome_linked_peak": [3, 4, 6, 7, 8], "multipl": [0, 3, 10, 12], "multiprocess": 12, "must": [6, 7, 12], "myenv": 1, "n": [3, 4, 5, 6, 8, 10], "n_cell": 5, "n_cells_by_count": [4, 5, 6], "n_compon": [7, 12], "n_ctrl_set": 12, "n_gene": [4, 5, 6, 7, 12], "n_genes_by_count": [4, 5, 6, 7], "n_genes_us": [6, 7], "n_highly_variable_gen": [4, 5, 7, 12], "n_iter": [4, 6, 7], "n_nearest_neighbor": 12, "n_neighbor": [5, 7], "n_ob": [3, 4, 5, 6, 7, 8, 10], "n_split": 7, "n_svd_dim": [4, 5, 7, 12], "n_top_gen": [5, 7, 12], "n_var": [3, 4, 5, 6, 7, 8, 10], "name": [1, 3, 4, 6, 7, 8, 10, 11, 12], "nat": 13, "nativ": [3, 4, 6, 7, 8], "natur": 6, "ncol": [3, 4, 5, 6, 7, 8, 10], "ncount_rna": 4, "ndarrai": 12, "nearest": 12, "neg": 7, "neighbor": [3, 4, 5, 6, 7, 8, 12], "neighborhood": 4, "network": 0, "neun": [3, 4, 6, 7, 8], "neuron": [0, 13], "neuropatholog": [3, 4, 6, 7, 8], "neurosci": 13, "neurotyp": [3, 4, 6, 7, 8], "new": [11, 12], "nfeature_rna": 4, "non": [3, 4, 6, 7, 8], "none": [3, 4, 5, 6, 7, 8, 10, 12], "normal": [0, 4, 6, 12], "normalize_pearson_residu": 7, "normalize_tot": 7, "normalized_cf_matrix_df": [4, 7], "note": [0, 7], "notebook": [3, 8], "notebook_tqdm": 7, "novel": 0, "novo": 12, "now": [4, 7], "np": [3, 4, 5, 6, 7, 8, 10, 12], "nrow": 8, "nuclear_read_pair": [3, 4, 6, 7, 8], "number": [4, 6, 7, 8, 10, 12], "numpi": [3, 4, 5, 6, 7, 8, 10], "o": [4, 5, 6, 7], "ob": [3, 4, 5, 6, 7, 8, 10, 12], "object": [3, 4, 5, 6, 7, 8, 10, 12], "observ": [3, 4, 6, 7, 8, 10, 12], "obsm": [3, 4, 5, 6, 7, 8, 10, 12], "obsp": [3, 4, 5, 6, 7, 8], "obtain": [5, 10], "off": 10, "oligodendrocyt": 3, "omic": [7, 9], "one": [1, 12], "onli": [5, 12], "onto": 3, "opc": 3, "oper": 12, "option": [4, 5, 6, 7, 10, 11, 12], "order": [3, 11], "org": [3, 4, 6, 7, 8, 13], "organ": [3, 4, 6, 7, 8], "origin": [3, 4, 6, 7, 8], "other": [0, 3, 4, 6, 7, 8], "otherwis": 12, "our": [4, 6, 7], "out": [6, 7], "outlin": 10, "output": 12, "over": 7, "overal": [3, 4, 6, 7, 8], "overdispers": 7, "overwrit": 4, "p": 12, "p57": 5, "pacif": [3, 4, 6, 7, 8], "packag": [2, 3, 4, 5, 6, 7, 8], "page": 0, "pair": [5, 6, 7], "palett": [3, 4, 5, 6, 7, 15], "palette_valu": 8, "panda": [4, 5, 6, 7, 11, 12], "panel": 10, "parallel": [5, 12], "parallel_tim": 5, "paramet": [3, 5, 6, 7, 10, 11, 12], "part": 7, "particularli": 4, "pass": [6, 7, 10], "path": [3, 4, 5, 6, 7, 8, 10], "pax6": [3, 6], "pbmc3k": 10, "pcr_cycl": [3, 4, 6, 7, 8], "pct_counts_mt": [4, 5, 6, 7], "pct_counts_ribo": [4, 5, 6, 7], "pct_dropout_by_count": [4, 5, 6], "pd": [4, 5, 6, 7, 11], "pdf": [3, 10], "pdgfra": 3, "pearson": 7, "pei": 13, "per": [3, 7, 10, 12], "percenatg": 7, "percent_cdna_longer_than_400bp": [3, 4, 6, 7, 8], "percent_top": 7, "percentag": 7, "perform": [0, 4, 7, 12], "ph": [3, 4, 6, 7, 8], "piaso": [1, 2, 4, 5, 6, 7, 15], "piaso_test": 7, "piaso_umap_split_by_condit": 3, "pip": [1, 7], "pl": [3, 4, 5, 6, 7, 8, 10], "place": 12, "platt": 8, "pleas": [0, 4, 6, 7, 12], "plot": [0, 2, 5, 7, 8, 9, 15], "plot_embeddings_split": [2, 3, 8, 9, 10], "plot_features_violin": [2, 3, 7, 9, 10], "plt": [3, 4, 5, 6, 7, 8, 10], "pmi": [3, 4, 6, 7, 8], "point": 10, "portal": [3, 4, 6, 7, 8], "pos_cmap": [4, 6], "posit": [3, 4, 6, 7, 8], "pp": [3, 5, 7], "pr_offset": 7, "pre": [4, 6, 7], "precis": [7, 9], "predict": [12, 15], "predictcelltypebygdr": [2, 4, 7, 9, 12], "predicted_doublet": 5, "prefix": [3, 8], "preprocess": [0, 2, 4, 6, 9], "presenc": 5, "present": 11, "primari": [3, 4, 6, 7, 8], "princip": 12, "print": [5, 12], "print_head": [4, 5, 6, 7], "prior": 7, "proceed": 7, "process": [4, 6, 7, 12], "produc": [10, 12], "progress": [3, 4, 6, 7, 8, 12], "project": [3, 4, 6, 7, 8], "proportion": [0, 13], "protein": 7, "provid": [8, 10, 12], "pseudo": [3, 4, 6, 7, 8], "pt": 10, "public": [3, 4, 5, 6, 7, 8], "purpl": [4, 7], "purpos": 7, "pvalb": 3, "pvm": 3, "py": [4, 5, 6, 7], "pynndesc": [4, 5, 6, 7], "pyplot": [3, 4, 5, 6, 7, 8, 10], "pyramid": [0, 13], "python": [3, 4, 5, 6, 7, 8], "python3": 7, "qc_var": 7, "qualiti": 7, "quantification_fmol": [3, 4, 6, 7, 8], "queri": [7, 12], "query_data": 12, "query_groupbi": [4, 7, 12], "r1_index": [3, 4, 6, 7, 8], "race": [3, 4, 6, 7, 8], "rachleff": 13, "random": 12, "random_se": 12, "random_st": [5, 7, 12], "randomli": 12, "rang": [8, 10], "rank": [3, 11], "ratio": 10, "ravel": [5, 7], "raw": [4, 5, 6, 7, 10], "rc": [4, 7], "rcparam": [3, 4, 5, 6, 7, 8], "re": 7, "read": [3, 4, 5, 6, 7, 8], "read_h5ad": 12, "readthedoc": [4, 7], "real": [4, 7], "reason": 7, "reduc": 7, "reduct": [0, 6, 12, 15], "ref_adata": 4, "refer": [0, 3, 6, 7, 8, 12], "reference_data": 12, "reference_groupbi": [4, 7, 12], "reflect": 7, "region": [3, 4, 6, 7, 8], "regular": 12, "regulatori": 0, "reindex": [6, 7], "rel": 10, "releas": 0, "reli": 0, "remain": 7, "remov": [4, 7], "remove_lowly_express": [6, 7], "reorder": 4, "reorder_categori": [4, 7], "replac": 1, "repres": 12, "represent": 12, "reproduc": 12, "requir": 5, "rerun": [6, 12], "research": 0, "reshap": 10, "residu": 7, "residual_vari": [4, 5, 6, 7], "resolut": [4, 5, 6, 7, 12], "respect": 11, "restrict_to": 6, "result": [3, 8, 10, 11, 12], "retain": 12, "return": [10, 11, 12], "return_gene_nam": 12, "return_integr": [4, 7, 12], "reveal": 6, "rgb": 10, "ribo": 7, "ribo_cel": 7, "ribosom": 7, "right": 10, "rin": [3, 4, 6, 7, 8], "rna": [5, 7], "rna_amplif": [3, 4, 6, 7, 8], "rna_amplification_pass_fail": [3, 4, 6, 7, 8], "rnaseq": 4, "robust": 12, "root": 10, "row": [3, 6, 7, 10], "row_siz": [8, 10], "rp": 7, "rpl": 7, "run": [0, 4, 5, 12], "runcosgparallel": [2, 9, 12], "rungdr": [2, 5, 7, 9, 12], "rungdrparallel": [2, 5, 9, 12], "runsvd": [2, 7, 9, 12], "runsvdlazi": [2, 9, 12], "runtimewarn": [5, 7], "s41593": 13, "same": 8, "sampl": [3, 4, 5], "sample_color": [4, 5], "sample_id": [3, 4, 6, 7, 8], "sample_nam": [3, 4, 6, 7, 8], "sample_quantity_count": [3, 4, 6, 7, 8], "san": [4, 5, 6, 7], "satb2": 3, "save": [4, 5, 7, 10, 12], "save_dir": [3, 8], "sb": 5, "sbuimehrvvhnf7tfu7h6xi": [4, 6], "sc": [3, 4, 5, 6, 7, 8, 10, 12], "scale": 10, "scale_data": 12, "scanpi": [3, 4, 5, 6, 7, 8, 10, 12], "scatac": 0, "scib_metr": 5, "scikit": [4, 5, 6, 7], "scipi": [3, 4, 5, 6, 7, 8], "score": [0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12], "score_color": [3, 4, 6, 7, 8], "score_lay": [5, 7, 12], "score_list": 12, "score_method": 12, "scoring_method": [4, 5, 7, 12], "scratch": [3, 4, 5, 6, 8], "screen": [3, 4, 6, 7, 8], "scrna": [0, 4, 6, 7], "scrublet_cluster_scor": 5, "scrublet_scor": 5, "sea": [3, 4, 6, 7, 8], "seaborn": [4, 5, 6, 7], "search": 0, "seattl": [3, 4, 6, 7, 8], "second": 11, "secondari": [3, 4, 6, 7, 8], "section": [3, 4, 6, 7, 8], "see": [4, 7, 12], "seed": 12, "select": [0, 7, 8, 12, 15], "separ": [5, 10, 12], "seq": [0, 3, 4, 6, 7, 8], "sequenc": [10, 12], "seri": 11, "serif": [4, 5, 6, 7], "serpine1": 3, "set": [0, 3, 4, 5, 6, 7, 8, 10, 11, 12], "set_figure_param": [3, 4, 5, 6, 7, 8], "set_siz": 8, "set_styl": [4, 7], "set_vis": 8, "sever": [3, 4, 6, 7, 8], "sex": [3, 4, 6, 7, 8], "sex_color": [3, 4, 6, 7, 8], "shape": 10, "share": 12, "should": 10, "show": [3, 4, 5, 7, 8, 10], "show_axis_tick": 10, "show_figur": 10, "show_grid": [3, 10], "shown": 10, "side": [0, 10], "similar": [6, 12], "simplefilt": [3, 5, 8], "sinc": [4, 7], "singl": [3, 7, 8, 9, 12], "singular": 12, "site": 7, "size": [3, 4, 5, 6, 7, 8, 10, 12], "sklearn": [4, 7], "smaller": 8, "sn": [4, 5, 6, 7], "sncg": [3, 6], "snrna": [3, 8], "so": [6, 10], "softwar": [4, 5, 7], "some": [7, 8], "sort": 11, "space": [7, 10], "spars": [3, 5, 8], "speci": 4, "species_tech": 4, "species_tech_color": 4, "specif": [0, 12, 13], "specifi": [3, 4, 6, 7, 8, 10, 11, 12], "spectral_r": [3, 6, 7, 8], "split": [0, 7, 10], "splitbi": [3, 10], "sqrt": [7, 8], "squar": 10, "sst": 3, "stabl": 7, "stack": 0, "stand": 7, "standard": [5, 12], "standard_scal": [6, 7], "start": 7, "start_tim": 5, "startswith": [7, 8], "statsmodel": [4, 5, 6, 7], "statu": [3, 4, 6, 7, 8], "status_color": [3, 4, 6, 7, 8], "std": 7, "step": 7, "store": [3, 6, 7, 12], "str": [7, 8, 10, 11, 12], "stratifi": 7, "stratified_split": 7, "stratifiedshufflesplit": 7, "string": 10, "studi": [3, 4, 6, 7, 8], "sub": 6, "subclass": [4, 6, 7, 8], "subclass_color": [3, 4, 6, 7, 8], "subclass_df": 3, "subclassxcondit": 3, "sublist": 12, "subpackag": 2, "subplot": [3, 8, 10], "subsampl": [3, 4, 6, 7, 8], "subset": 4, "subtyp": [0, 13], "successfulli": [5, 6], "suffix": 12, "suggest": 6, "sum": [4, 7], "supertyp": [3, 4, 6, 7, 8], "supertype_color": [3, 4, 6, 7, 8], "support": 12, "suptitl": 8, "sure": 7, "surviv": [0, 13], "svd": [7, 12], "swap_ax": [3, 6, 7], "switch": 3, "sy": [3, 4, 5, 6, 7, 8], "t": [1, 5, 6, 7], "tabl": [2, 3, 9, 11], "target_sum": 7, "tbr1": 3, "tech": 4, "test": [4, 7], "test_index": 7, "test_siz": 7, "thal": [3, 4, 6, 7, 8], "thal_color": [3, 4, 6, 7, 8], "than": [6, 7, 8], "thei": 6, "them": 4, "therefor": [4, 7], "theta": 7, "thi": [3, 4, 5, 6, 7, 8, 10, 12], "through": 7, "tick": 10, "tight_layout": 8, "time": [4, 5, 6, 7], "titl": [3, 4, 6, 7, 8], "tl": [4, 5, 6, 7, 12], "togeth": [3, 6], "tool": [1, 2, 4, 5, 6, 7, 9], "top": [7, 12], "total": [3, 4, 5, 6, 7, 8], "total_count": [4, 5, 6, 7], "total_counts_mt": [4, 5, 6], "tqdm": [4, 6, 7], "tqdmwarn": 7, "train": 7, "train_index": 7, "transform": 12, "travaglini": 13, "tree": 12, "trim": 12, "true": [3, 4, 5, 6, 7, 10, 11, 12], "truncat": 12, "tupl": 12, "turori": 5, "tutori": [0, 4, 6, 7], "two": 11, "type": [0, 3, 5, 6, 10, 11, 12, 15], "u": [6, 7], "umap": [7, 8], "umap_1": 4, "umap_2": 4, "umi": [4, 6, 7, 8, 12], "un": [3, 4, 5, 6, 7, 8, 12], "under": [10, 12], "union": [4, 5, 6, 7], "uniqu": [10, 11], "unknown": [3, 4, 6, 7, 8], "unmap": [3, 4, 6, 7, 8], "unreport": [3, 4, 6, 7, 8], "upcom": 0, "updat": [7, 12], "upper": 10, "us": [0, 3, 4, 5, 6, 7, 8, 10, 11, 12], "usag": 12, "use_highly_vari": [4, 5, 7, 12], "use_raw": [6, 7, 10], "use_rep": [5, 6, 7], "user": [3, 4, 5, 6, 7, 8], "user_instal": 7, "usp": [3, 4, 6, 7, 8], "usual": 7, "uw": [3, 4, 6, 7, 8], "v": [3, 4, 5, 6, 7, 8, 13], "vallari": [3, 7, 8], "valu": [4, 6, 7, 10, 11, 12], "var": [3, 4, 5, 6, 7, 8], "var_nam": [3, 4, 7], "vari": 3, "variabl": [0, 5, 7, 10, 12], "varianc": [4, 5, 6, 7, 12], "variat": 12, "variou": 3, "varm": 5, "vas744": [3, 4, 5, 6, 7, 8], "vector": 7, "verbos": [4, 5, 6, 7, 12], "verifi": 4, "version": [4, 6, 7], "view": [3, 4, 6, 7, 8], "violin": [0, 7, 10], "violin_plot_by_piaso": 10, "violin_plot_piaso": 3, "vip": [3, 6], "viridi": [3, 4, 5, 6, 7, 8], "visual": 10, "vlmc": 3, "vmax": 10, "vmin": 10, "vv": 7, "wa": [4, 6], "wall": [4, 5, 6, 7], "wang": 13, "want": [6, 7], "warn": [3, 5, 8], "we": [3, 4, 5, 6, 7], "weight": [3, 4, 5, 6, 7, 8, 12], "well": [7, 12], "were": [4, 5, 7], "when": [5, 11, 12], "where": [10, 12], "whether": [10, 12], "which": [3, 6, 7, 10, 11, 12], "while": 3, "white": [3, 4, 5, 6, 7, 8], "whitegrid": [4, 7], "whole": 6, "widht": 3, "width": 10, "width_singl": [3, 10], "within": 12, "without": 5, "work": [5, 6, 7], "worker": 12, "workflow": 7, "wu": [0, 13], "www": 7, "x": [4, 5, 6, 7, 10, 12, 13], "x_gdr": [4, 5, 7, 12], "x_gdr_harmoni": 4, "x_max": 10, "x_min": 10, "x_normal": [3, 4, 6, 7, 8], "x_pca": [5, 10, 12], "x_pca_harmoni": [5, 12], "x_scvi": [3, 4, 6, 7, 8], "x_svd": [4, 6, 7, 12], "x_svd_full": [6, 12], "x_svd_full_harmoni": 12, "x_umap": [3, 4, 5, 6, 7, 8, 10], "x_umap_backup": [6, 7], "x_umap_gdr": 4, "x_umap_gdr_harmoni": 4, "x_umap_pca": 5, "x_umap_publish": 4, "x_umap_raw": [5, 7], "x_umap_svd": 7, "xenium": 0, "xlabel": 4, "xticklabel": [4, 7], "y": [7, 10], "y_max": 10, "y_min": 10, "year": [3, 4, 6, 7, 8], "yet": [1, 7], "ylabel": 4, "you": [1, 3, 6, 7, 8], "your": [0, 1, 7], "yticklabel": [4, 7], "zero": [5, 7], "zip": 3}, "titles": ["PIASO: Precise Integrative Analysis of Single-cell Omics", "Installation", "API", "Plotting functions in PIASO", "Cell type prediction by GDR", "GDR: marker Gene-guided Dimensionality Reduction", "Finer clustering on selected cluster(s)", "Introduction", "Built-in color palettes in PIASO", "piaso package", "piaso.plotting package", "piaso.preprocessing package", "piaso.tools package", "References", "Release notes", "Tutorials"], "titleterms": {"": 6, "0": 14, "1": 14, "2025": 14, "24": 14, "across": [3, 6], "algorithm": 6, "all": 6, "analysi": 0, "api": 2, "base": 5, "built": 8, "categor": 3, "categori": 3, "cell": [0, 3, 4, 7], "citat": 0, "cluster": [4, 6, 7], "color": [3, 8], "combin": 3, "condit": 3, "continu": [3, 8], "count": 3, "data": [3, 4, 5, 6, 7, 8], "dataset": 4, "develop": 1, "differ": 3, "dimension": [5, 7], "discret": [3, 8], "each": [3, 6], "entir": 3, "everi": 7, "experiment": 3, "express": 3, "februari": 14, "finer": 6, "function": 3, "gdr": [4, 5, 7], "gdrparallel": 5, "gene": [3, 5, 6, 7], "group": [3, 6], "guid": 5, "identifi": [6, 7], "independ": 6, "indic": 0, "infog": 5, "instal": 1, "integr": 0, "introduct": 7, "leiden": 6, "leiden_loc": 6, "load": [3, 4, 5, 6, 7, 8], "log": 7, "map": 3, "marker": [5, 7], "markger": 6, "multipl": 6, "normal": [5, 7], "note": 14, "number": 3, "omic": 0, "one": 6, "output": 3, "packag": [9, 10, 11, 12], "palett": 8, "pca": 5, "perform": 6, "piaso": [0, 3, 8, 9, 10, 11, 12], "plot": [3, 10], "precis": 0, "predict": [4, 7], "preprocess": 11, "queri": 4, "reduct": [5, 7], "refer": [4, 13], "releas": 14, "result": 5, "run": [6, 7], "save": 3, "select": 6, "singl": 0, "specif": 3, "split": 3, "stabl": 1, "subclass": 3, "subclust": 6, "subpackag": 9, "tabl": 0, "tool": 12, "transform": 7, "tutori": 15, "two": 3, "type": [4, 7], "umap": [3, 4, 5, 6], "umi": 3, "variabl": 3, "version": 1, "violin": 3, "visual": [3, 4, 5, 6, 7], "visula": 3, "within": 6}})