Search.setIndex({"alltitles": {"API": [[3, "api"]], "Cell type prediction by GDR": [[5, "Cell-type-prediction-by-GDR"]], "Citation": [[0, "citation"]], "Cluster the data": [[8, "Cluster-the-data"]], "Cluster the data with Leiden algorithm": [[7, "Cluster-the-data-with-Leiden-algorithm"]], "Cluster the query data": [[5, "Cluster-the-query-data"]], "Color pallettes in PIASO": [[9, "Color-pallettes-in-PIASO"]], "Combine categorical variables": [[4, "Combine-categorical-variables"]], "Continuous color palettes": [[9, "Continuous-color-palettes"]], "Count the number of cells in each category": [[4, "Count-the-number-of-cells-in-each-category"]], "Development version": [[2, "development-version"]], "Dimensionality reduction and visualization": [[8, "Dimensionality-reduction-and-visualization"]], "Dimensionality reduction with GDRParallel": [[6, "Dimensionality-reduction-with-GDRParallel"]], "Discrete color palettes": [[9, "Discrete-color-palettes"]], "Finer clustering on selected cluster(s)": [[7, "Finer-clustering-on-selected-cluster(s)"]], "GDR: marker Gene-guided Dimensionality Reduction": [[6, "GDR:-marker-Gene-guided-Dimensionality-Reduction"]], "INFOG Normalization": [[6, "INFOG-Normalization"]], "Identify marker genes for every cluster": [[8, "Identify-marker-genes-for-every-cluster"]], "Identify markger genes in subclusters": [[7, "Identify-markger-genes-in-subclusters"]], "Indices and tables": [[0, "indices-and-tables"]], "Installation": [[2, "installation"]], "Introduction": [[8, "Introduction"]], "Load the data": [[4, "Load-the-data"], [5, "Load-the-data"], [6, "Load-the-data"], [7, "Load-the-data"], [8, "Load-the-data"], [9, "Load-the-data"]], "Load the query dataset": [[5, "Load-the-query-dataset"]], "Load the reference dataset": [[5, "Load-the-reference-dataset"]], "Normalize and log-transform the data": [[8, "Normalize-and-log-transform-the-data"]], "PIASO: Precise Integrative Analysis of Single-cell Omics": [[0, "piaso-precise-integrative-analysis-of-single-cell-omics"]], "Perform clustering across all selected groups": [[7, "Perform-clustering-across-all-selected-groups"]], "Perform clustering independently within each group": [[7, "Perform-clustering-independently-within-each-group"]], "Plotting functions in PIASO": [[4, "Plotting-functions-in-PIASO"]], "Predict cell types by GDR": [[5, "Predict-cell-types-by-GDR"], [8, "Predict-cell-types-by-GDR"]], "References": [[1, "references"], [14, "References"]], "Release 1.0.0 (released February 24, 2025)": [[15, "release-1-0-0-released-february-24-2025"]], "Release notes": [[15, "release-notes"]], "Run GDR for dimensionality reduction": [[8, "Run-GDR-for-dimensionality-reduction"]], "Run leiden_local on multiple clusters": [[7, "Run-leiden_local-on-multiple-clusters"]], "Run leiden_local on one cluster": [[7, "Run-leiden_local-on-one-cluster"]], "Save output plots": [[4, "Save-output-plots"]], "Split the UMAP by condition": [[4, "Split-the-UMAP-by-condition"]], "Stable version": [[2, "stable-version"]], "Subpackages": [[10, "subpackages"]], "Tutorials": [[17, "tutorials"]], "Violin plots of entire data": [[4, "Violin-plots-of-entire-data"]], "Violin plots of grouped data": [[4, "Violin-plots-of-grouped-data"]], "Visualize GDR results with UMAPs": [[6, "Visualize-GDR-results-with-UMAPs"]], "Visualize with PCA-based UMAP": [[6, "Visualize-with-PCA-based-UMAP"]], "Visualize with UMAP": [[7, "Visualize-with-UMAP"]], "Visualize with UMAPs": [[5, "Visualize-with-UMAPs"]], "Visualize with a continuous color map": [[4, "Visualize-with-a-continuous-color-map"]], "Visualize with a discrete color map": [[4, "Visualize-with-a-discrete-color-map"]], "Visualizing categorical variables (Cell subclass) across different experimental conditions": [[4, "Visualizing-categorical-variables-(Cell-subclass)-across-different-experimental-conditions"]], "Visualizing the expression of a specific gene across two different experimental conditions": [[4, "Visualizing-the-expression-of-a-specific-gene-across-two-different-experimental-conditions"]], "Visulaizing continuous variables (Number of UMIs) across different experimental conditions": [[4, "Visulaizing-continuous-variables-(Number-of-UMIs)-across-different-experimental-conditions"]], "piaso package": [[10, "module-piaso"]], "piaso.plotting package": [[11, "module-piaso.plotting"]], "piaso.preprocessing package": [[12, "module-piaso.preprocessing"]], "piaso.tools package": [[13, "module-piaso.tools"]], "setup module": [[16, "setup-module"]]}, "docnames": ["index", "input", "installation", "modules", "notebooks/01_showThePlotFunction", "notebooks/GDRPredictCellTypeTutorial", "notebooks/GDR_Tutorial", "notebooks/LeidenLocalTutorial", "notebooks/PIASO_tutorial", "notebooks/colorPalettes", "piaso", "piaso.plotting", "piaso.preprocessing", "piaso.tools", "references", "releaseNotes", "setup", "tutorials"], "envversion": {"nbsphinx": 4, "sphinx": 61, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.intersphinx": 1, "sphinxcontrib.bibtex": 9}, "filenames": ["index.rst", "input.md", "installation.md", "modules.rst", "notebooks/01_showThePlotFunction.ipynb", "notebooks/GDRPredictCellTypeTutorial.ipynb", "notebooks/GDR_Tutorial.ipynb", "notebooks/LeidenLocalTutorial.ipynb", "notebooks/PIASO_tutorial.ipynb", "notebooks/colorPalettes.ipynb", "piaso.rst", "piaso.plotting.rst", "piaso.preprocessing.rst", "piaso.tools.rst", "references.ipynb", "releaseNotes.rst", "setup.rst", "tutorials.rst"], "indexentries": {"calculatescoreparallel() (in module piaso.tools)": [[13, "piaso.tools.calculateScoreParallel", false]], "calculatescoreparallel_multibatch() (in module piaso.tools)": [[13, "piaso.tools.calculateScoreParallel_multiBatch", false]], "createcustomcmapfromhex() (in module piaso.plotting)": [[11, "piaso.plotting.createCustomCmapFromHex", false]], "getcrosscategories() (in module piaso.preprocessing)": [[12, "piaso.preprocessing.getCrossCategories", false]], "infog() (in module piaso.tools)": [[13, "piaso.tools.infog", false]], "leiden_local() (in module piaso.tools)": [[13, "piaso.tools.leiden_local", false]], "module": [[10, "module-piaso", false], [11, "module-piaso.plotting", false], [12, "module-piaso.preprocessing", false], [13, "module-piaso.tools", false]], "piaso": [[10, "module-piaso", false]], "piaso.plotting": [[11, "module-piaso.plotting", false]], "piaso.preprocessing": [[12, "module-piaso.preprocessing", false]], "piaso.tools": [[13, "module-piaso.tools", false]], "plot_embeddings_split() (in module piaso.plotting)": [[11, "piaso.plotting.plot_embeddings_split", false]], "plot_features_violin() (in module piaso.plotting)": [[11, "piaso.plotting.plot_features_violin", false]], "predictcelltypebygdr() (in module piaso.tools)": [[13, "piaso.tools.predictCellTypeByGDR", false]], "runcosgparallel() (in module piaso.tools)": [[13, "piaso.tools.runCOSGParallel", false]], "rungdr() (in module piaso.tools)": [[13, "piaso.tools.runGDR", false]], "rungdrparallel() (in module piaso.tools)": [[13, "piaso.tools.runGDRParallel", false]], "runsvd() (in module piaso.tools)": [[13, "piaso.tools.runSVD", false]], "runsvdlazy() (in module piaso.tools)": [[13, "piaso.tools.runSVDLazy", false]], "score() (in module piaso.tools)": [[13, "piaso.tools.score", false]], "table() (in module piaso.preprocessing)": [[12, "piaso.preprocessing.table", false]]}, "objects": {"": [[10, 0, 0, "-", "piaso"]], "piaso": [[11, 0, 0, "-", "plotting"], [12, 0, 0, "-", "preprocessing"], [13, 0, 0, "-", "tools"]], "piaso.plotting": [[11, 1, 1, "", "createCustomCmapFromHex"], [11, 1, 1, "", "plot_embeddings_split"], [11, 1, 1, "", "plot_features_violin"]], "piaso.preprocessing": [[12, 1, 1, "", "getCrossCategories"], [12, 1, 1, "", "table"]], "piaso.tools": [[13, 1, 1, "", "calculateScoreParallel"], [13, 1, 1, "", "calculateScoreParallel_multiBatch"], [13, 1, 1, "", "infog"], [13, 1, 1, "", "leiden_local"], [13, 1, 1, "", "predictCellTypeByGDR"], [13, 1, 1, "", "runCOSGParallel"], [13, 1, 1, "", "runGDR"], [13, 1, 1, "", "runGDRParallel"], [13, 1, 1, "", "runSVD"], [13, 1, 1, "", "runSVDLazy"], [13, 1, 1, "", "score"]]}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "function", "Python function"]}, "objtypes": {"0": "py:module", "1": "py:function"}, "terms": {"": [0, 5, 6, 8, 13, 14, 17], "0": [4, 5, 6, 7, 8, 9, 11, 13], "00": [5, 6, 7, 8], "01": [5, 7, 8], "01774": 14, "018": 5, "02": [6, 8], "024": 14, "03": 8, "04": 7, "05": [5, 6, 11], "06": 5, "060": 8, "07": [6, 8, 14], "08": 8, "09": [6, 8], "1": [2, 4, 5, 6, 7, 8, 9, 11, 13], "10": [4, 5, 6, 7, 8, 9, 11, 13, 14], "100": [4, 6, 7, 8, 13], "1000": 13, "10206": 7, "1038": 14, "1080": 4, "1093": 14, "11": [4, 5, 6, 7, 8], "1101": [8, 14], "11039": 7, "12": [4, 5, 6, 7, 8], "120": 5, "1299": 4, "12gngenuh2g9if975u8prtx3vpav5rbnk": 6, "13": [4, 5, 6, 7, 8, 9], "14": [4, 5, 6, 7, 8, 11], "140": 4, "14346": 5, "15": [4, 5, 6, 7, 8, 9, 13], "1551": 4, "156": 8, "157": 4, "16": [4, 5, 6, 7, 8], "1663": 4, "17": [4, 5, 7, 8], "17412": 6, "1797": 4, "18": [4, 5, 7, 8], "19": [4, 5, 8, 9], "191": 7, "1927": 13, "1b2ybvwcuvih0987zdniezkdz138x6zti": 5, "1e4": 8, "1min": 8, "1nf0irtgffqcecm6hycpptiykxs65k2p4": 5, "1nh": [4, 8, 9], "1pdbigpveo": [5, 7], "2": [2, 4, 5, 6, 7, 8, 9, 11, 13, 14], "20": [4, 6, 8, 13, 14], "200": 8, "20000": [4, 5, 7, 8, 9], "2022": [8, 14], "2024": [0, 5, 7, 8, 14], "2025": [5, 8], "20k": [4, 5, 7, 8, 9], "21": [4, 6, 8], "22": [4, 5, 6, 8], "224": 4, "23": [4, 5, 6, 8, 14], "231": 4, "2366": 14, "2383": 14, "2388": 4, "24": [4, 5, 7, 8, 9], "25": [4, 5, 8, 13], "256": 11, "26": [4, 5, 6, 7, 8], "26205": 6, "27": [4, 5, 8, 14], "279": 8, "28": [4, 6, 8], "29": [4, 8], "299": 4, "2d": 13, "3": [4, 5, 6, 7, 8, 9], "30": [4, 7, 8, 9, 13], "300": [4, 5, 6, 7, 8, 9], "3000": [6, 8, 13], "305": 4, "31": [4, 8], "32": [4, 5, 6, 8], "329": 4, "33": [4, 8], "34": [4, 7, 8], "35": [4, 6, 8], "353": 5, "354": 4, "36": 4, "36601": [4, 5, 7, 8, 9], "37": 4, "387": 5, "39": [4, 5, 6, 7, 8, 9], "3min": 8, "4": [4, 5, 6, 7, 8, 9, 13], "40": [6, 13], "41": 6, "42": 13, "423": 8, "43": 8, "44": 8, "451": 8, "46": 6, "479": 8, "48": 6, "4826": 4, "496": 4, "5": [4, 5, 6, 7, 8, 9, 11, 14], "50": [5, 6, 8, 13], "5000": [5, 6, 8, 13], "51": 8, "513": 8, "5168": 7, "525": 8, "55": 5, "564": 5, "572266": 11, "595": 8, "6": [4, 5, 6, 7, 8, 9, 11], "60": 6, "604399": 14, "604399v1": 8, "605": 4, "610": 5, "63": 6, "638": 4, "6493": 7, "653": 4, "661": 8, "692": 8, "7": [4, 5, 6, 7, 8, 9], "70": [6, 8], "72310972213745": 6, "73": 6, "76": 8, "76560": 5, "767": 7, "77": 8, "780": 4, "79": 6, "8": [4, 5, 6, 7, 8, 9, 13], "80": [4, 5, 6, 7, 8, 9, 11], "805": 5, "81": 8, "839": 5, "84": 8, "853": 7, "859": 7, "86": 4, "9": [4, 5, 6, 7, 8, 11], "9092": 7, "91": 5, "92": [6, 7], "96": [6, 8, 9], "97": 6, "A": [11, 12, 13], "But": 8, "By": 4, "For": [8, 13], "IT": 4, "If": [0, 2, 8, 11, 12, 13], "In": [4, 5, 6, 7, 8], "No": 8, "The": [4, 5, 6, 7, 8, 9, 11, 12, 13], "To": [4, 5, 7, 8, 13], "_cluster": 7, "_normal": [6, 8], "_rungdr": [5, 8], "about": [5, 7, 8], "absent": 4, "access": 13, "accord": [4, 11], "accur": 14, "accuraci": 5, "achiev": [7, 8], "across": 13, "action": [4, 6, 9], "activ": [0, 2], "actual": [2, 7, 8], "ad": [4, 5, 6, 7, 8, 9, 13], "ad_rna_mtg_subsample_excludereference_20k_piaso": [4, 5, 8, 9], "ad_rna_mtg_subsample_excludereference_20k_piaso_preprocess": [5, 7], "adata": [4, 5, 6, 7, 8, 9, 11, 13], "adata_combin": [5, 8], "adata_i": 7, "adata_path": 6, "adata_ref": [5, 8, 13], "adata_test": 8, "add": 13, "addit": [6, 11], "adgrg7": 4, "adjac": [6, 7, 8], "adjust": [4, 8], "adultcortexmultiomerna_integrated_anno": 6, "affect": [4, 5, 7, 8, 9], "african": [4, 5, 7, 8, 9], "after": [5, 8, 11], "ag": [4, 5, 7, 8, 9], "again": 8, "aibs_query_adata": 5, "aim": 9, "al": [0, 5, 7, 8, 14], "alaska": [4, 5, 7, 8, 9], "algorithm": 13, "align": [4, 5, 7, 8, 9], "all": [4, 5, 8, 13], "allen": [4, 9], "allow": 13, "along": 7, "alongsid": 4, "alreadi": [2, 8], "also": [4, 7, 8], "alter": [0, 14], "altern": 8, "alzheim": [4, 5, 7, 8, 9, 14], "american": [4, 5, 7, 8, 9], "among": 13, "amplified_quantity_ng": [4, 5, 7, 8, 9], "an": [2, 8, 11, 13], "analysi": [4, 5, 6, 7, 8, 9, 10, 13], "analyt": 8, "ani": [7, 8, 11], "anndata": [4, 5, 6, 7, 8, 9, 11, 13], "annot": [5, 7, 8, 11, 13], "ap": [4, 5, 7, 8, 9], "api": 0, "apo": [4, 5, 7, 8, 9], "apoe4": [4, 5, 7, 8, 9], "append": [4, 5, 6, 7, 8, 9], "appli": 11, "applic": [4, 8], "approach": 13, "ar": [5, 6, 7, 8, 11, 12, 13], "ar_id": [4, 5, 7, 8, 9], "aren": 8, "argument": 11, "arial": [5, 6, 7, 8], "around": 11, "arrai": 13, "arteriolosclerosi": [4, 5, 7, 8, 9], "as_datafram": [4, 12], "ascend": [4, 12], "asian": [4, 5, 7, 8, 9], "aspect": 11, "assess": 6, "assign": 11, "astrocyt": 4, "astyp": [5, 8], "atac_barcod": 6, "atac_chimeric_read": 6, "atac_confidently_mapped_read_pair": [4, 5, 7, 8, 9], "atac_dup_read": 6, "atac_fraction_of_genome_in_peak": [4, 5, 7, 8, 9], "atac_fraction_of_high_quality_fragments_in_cel": [4, 5, 7, 8, 9], "atac_fraction_of_high_quality_fragments_overlapping_peak": [4, 5, 7, 8, 9], "atac_fraction_of_high_quality_fragments_overlapping_tss": [4, 5, 7, 8, 9], "atac_fraction_of_transposition_events_in_peaks_in_cel": [4, 5, 7, 8, 9], "atac_frag": 6, "atac_lowmapq": 6, "atac_mean_raw_read_pairs_per_cel": [4, 5, 7, 8, 9], "atac_median_high_quality_fragments_per_cel": [4, 5, 7, 8, 9], "atac_mitochondrial_read": 6, "atac_non": [4, 5, 7, 8, 9], "atac_number_of_peak": [4, 5, 7, 8, 9], "atac_peak_region_cutsit": 6, "atac_peak_region_frag": 6, "atac_percent_dupl": [4, 5, 7, 8, 9], "atac_q30_bases_in_barcod": [4, 5, 7, 8, 9], "atac_q30_bases_in_read_1": [4, 5, 7, 8, 9], "atac_q30_bases_in_read_2": [4, 5, 7, 8, 9], "atac_q30_bases_in_sample_index_i1": [4, 5, 7, 8, 9], "atac_raw_read": 6, "atac_sequenced_read_pair": [4, 5, 7, 8, 9], "atac_tss_enrichment_scor": [4, 5, 7, 8, 9], "atac_tss_frag": 6, "atac_unmapped_read": 6, "atac_unmapped_read_pair": [4, 5, 7, 8, 9], "atac_valid_barcod": [4, 5, 7, 8, 9], "atherosclerosi": [4, 5, 7, 8, 9], "atla": [5, 7, 8, 14], "attribut": 11, "author": 8, "auto": [8, 11], "automat": 11, "autonotebook": 8, "avail": [0, 4, 5, 6, 7, 8, 9, 11, 13], "avg_size_bp": [4, 5, 7, 8, 9], "ax": [5, 8, 9, 11], "axi": [5, 8, 11], "b1257a": 11, "backend": [7, 8], "backup": 8, "ball": 13, "barcod": 13, "base": [7, 8, 11, 13], "bash": 8, "basi": [4, 5, 8, 9, 11], "batch": [5, 6, 13], "batch_condit": [4, 5, 7, 8, 9], "batch_kei": [6, 7, 8, 13], "batch_vendor_nam": [4, 5, 7, 8, 9], "bbab579": 14, "bc": [4, 5, 7, 8, 9], "been": 8, "befor": [8, 13], "benchmark": 6, "better": [4, 7], "between": 7, "bh_pval": 6, "bib": 14, "bin": 8, "binomi": 8, "bioinformat": 14, "biolog": 8, "biorxiv": [0, 8, 14], "black": [4, 5, 7, 8, 9], "bodi": [4, 5, 7, 8, 9], "bool": [11, 12, 13], "boolean": [8, 11], "both": [9, 11], "boundari": 13, "braak": [4, 5, 7, 8, 9], "braak_color": [4, 5, 7, 8, 9], "brain": [4, 5, 7, 8, 9], "brief": 14, "built": [7, 13], "c": 9, "c_": 9, "c_color1": 4, "c_color3": [5, 8], "c_color4": [4, 6, 8, 11], "caa": [4, 5, 7, 8, 9], "calb2": 4, "calcul": [5, 6, 8, 12, 13], "calculate_qc_metr": 8, "calculate_score_multibatch": [6, 13], "calculatescoreparallel": [3, 10, 13], "calculatescoreparallel_multibatch": [3, 10, 13], "can": [4, 5, 7, 8, 11, 12], "captur": [7, 8], "car3": 4, "case": 6, "casi": [4, 5, 7, 8, 9], "cat": [5, 8], "categor": [5, 7, 8, 11, 12], "categori": [5, 6, 8, 9, 11, 12, 13], "categories_ord": [7, 8], "cdk9": 11, "ceil": [9, 11], "cell": [6, 7, 9, 10, 11, 13, 14, 17], "cell_depth": 6, "cell_prep_typ": [4, 5, 7, 8, 9], "cellbarcode_info": 13, "celltyp": [5, 6, 8, 13], "celltypes_color": 6, "celltypes_gdr": [5, 8, 13], "celltypes_tacco": 6, "celltypes_tacco_color": 6, "centroid": [5, 8], "cerad": [4, 5, 7, 8, 9], "chandeli": 4, "chang": [4, 5, 7, 8, 9], "change_color": [4, 5, 7, 8, 9], "check": 5, "chodl": 4, "choic": [4, 5, 7, 8, 9], "cite": 0, "clariti": 4, "class": [4, 5, 7, 8, 9], "clearli": 6, "clinic": [4, 5, 7, 8, 9], "close": 7, "cluster": [0, 6, 11, 13, 17], "cluster_color": 5, "clustering_typ": [7, 13], "cm": [5, 6, 7, 8], "cmap": [4, 5, 6, 7, 8, 9, 11], "code": 11, "cognit": [4, 5, 7, 8, 9], "col1": 12, "col2": 12, "col_siz": [9, 11], "collect": 13, "color": [5, 6, 7, 8, 11, 17], "color_map": [4, 5, 6, 7, 8, 9], "colorbar": [5, 6, 8], "colormap": 11, "column": [5, 8, 11, 12, 13], "com": [2, 4, 5, 7, 8, 9], "combin": [12, 13], "compar": [5, 7, 8], "complet": [5, 8], "compon": 13, "comput": [5, 6, 7, 8, 11, 13], "concat": [5, 8], "concaten": [5, 8], "conda": [2, 8], "condit": 0, "conditionxsubclass": 4, "confid": [4, 5, 7, 8, 9], "confusion_matrix": [5, 8], "confusion_matrix_df": [5, 8], "connect": [4, 5, 6, 7, 8, 9], "consid": [0, 13], "contain": [7, 11, 12, 13], "content": 8, "continu": [5, 7, 8], "control": [4, 8, 12, 13], "converg": [5, 8], "convert": 11, "coordin": [6, 8, 11], "copi": [6, 7, 8, 13], "core": 13, "correspond": [11, 13], "cortex": 6, "cortic": [0, 14], "cosg": [5, 6, 7, 8, 13, 14], "could": 4, "count": [6, 8, 12, 13], "counts_sum": 6, "cpu": [5, 6, 7, 8, 13], "cratqfxj5pavpy8_huqn1nricakq2": [4, 8, 9], "creat": [2, 11], "createcustomcmapfromhex": [3, 10, 11], "cross": 12, "cross_species_clust": 5, "cross_species_cluster_color": 5, "cst3_embeddingssplit": 11, "ct": 4, "current": 0, "custom": [11, 12], "d": [4, 5, 7, 8, 9], "d96998": 11, "d_": 9, "d_color1": 4, "d_color3": 4, "d_color4": [5, 6, 7, 8], "dai": [8, 14], "dai2022": 1, "data": [0, 11, 13], "data1": [4, 8, 9], "data_dir": 5, "datafram": [4, 5, 7, 8, 11, 12, 13], "dataset": [0, 4, 6, 7, 8, 11, 13], "de": 13, "death": [4, 5, 7, 8, 9], "debug": 13, "decomposit": 13, "default": [5, 7, 8, 11, 12, 13], "default_embed": [4, 5, 7, 8, 9], "defin": [11, 12, 13], "del": 8, "delimit": 12, "demonstr": [4, 8], "dendrogram": [7, 8], "dendrogram_": [7, 8], "dendrogram_leiden": [6, 8], "dendrogram_leiden_loc": 6, "dendrogram_leiden_local_on": 7, "denot": 11, "deprec": [5, 8], "depth": 13, "deriv": 8, "descend": 12, "describ": [5, 7, 8], "destin": [5, 6, 8], "detail": [5, 7, 8, 13], "detect": [4, 5, 7, 8, 9, 11], "determin": [4, 11], "df": 12, "df_tmp": [7, 8], "diag": 6, "dict": [4, 11, 12, 13], "dictionari": [12, 13], "diff": 8, "differ": [7, 13], "differenti": 7, "dimens": [8, 13], "dimension": [0, 7, 13, 17], "direct": [7, 8], "directli": 2, "directori": 8, "disabl": [5, 7, 8], "diseas": [4, 5, 7, 8, 9, 14], "disease_color": [4, 5, 7, 8, 9], "displai": [9, 11], "distanc": [4, 5, 6, 7, 8, 9], "distinct": 7, "distribut": 4, "divers": 9, "divid": [6, 8, 11], "do": 8, "doe": 11, "doi": 14, "don": [2, 8], "done": 8, "donor": [4, 5, 7, 8, 9], "donor_color": 5, "dot": [4, 11], "dotplot": [4, 7, 8], "doublet": [4, 5, 7, 8, 9, 11], "download": [5, 6, 7, 8], "downstream": 8, "dpi": [4, 5, 6, 7, 8, 9, 11], "dpi_sav": [4, 5, 6, 7, 8, 9], "dr_method": [7, 13], "drive": [4, 5, 7, 8, 9], "drive_link": [4, 5, 7, 8, 9], "dure": 13, "e": [0, 7, 8, 11, 13], "e8aebc": 11, "each": [6, 8, 11, 13], "educ": [4, 5, 7, 8, 9], "effect": [6, 13], "effici": 13, "either": [7, 13], "elimin": 8, "els": 9, "embed": [0, 4, 5, 6, 8, 9, 11, 13], "en": [5, 8], "enabl": 13, "encount": [6, 8], "end_tim": 6, "endotheli": 4, "enhanc": 4, "enrich": 13, "ensur": [8, 13], "enumer": 8, "env": 8, "environ": [2, 8], "equal": [11, 13], "et": [0, 4, 5, 7, 8, 14], "eval": 8, "exampl": [11, 13], "example_data": 13, "exce": 8, "excluded_reason": 6, "exhibit": 7, "exist": 12, "exp_component_vendor_nam": [4, 5, 7, 8, 9], "expand": [4, 5, 7, 8, 9], "expc_cell_captur": [4, 5, 7, 8, 9], "expect": 13, "experiment": 8, "experiment_component_fail": [4, 5, 7, 8, 9], "explor": [4, 5, 7, 8, 9], "express": [7, 8, 11, 13], "expressed_pct": [7, 8], "extract": 8, "facecolor": [4, 5, 6, 7, 8, 9], "facs_population_plan": [4, 5, 7, 8, 9], "factor": 11, "faefef": 11, "fals": [4, 5, 6, 7, 8, 9, 11, 12, 13], "famili": [5, 6, 7, 8], "fan": [4, 5, 7, 8, 9], "fast": 14, "fearur": 8, "featur": [0, 4, 8, 11, 13], "feature_list": [4, 11], "feature_typ": 6, "fezf2": 4, "fig": 9, "figdir": [4, 9], "figsiz": [4, 5, 6, 7, 8, 9, 11], "figur": [4, 5, 6, 7, 8, 9, 11], "file": [4, 5, 6, 7, 8, 9, 11], "filter": [7, 8], "filter_cel": 8, "finer": 17, "finish": [5, 6, 7, 8], "first": [2, 6, 7, 8, 12], "fit": 8, "fix": 11, "fix_coordinate_ratio": 11, "flavor": [5, 7, 8], "flexibl": 13, "float": [7, 8, 11, 13], "follow": 8, "font": [5, 6, 7, 8, 11], "font_manag": [5, 7, 8], "format": 11, "found": [5, 7, 8], "fraction": [4, 5, 7, 8, 9], "frameon": [4, 5, 6, 7, 8, 9], "frequent": 4, "fresh": [4, 5, 7, 8, 9], "from": [2, 4, 5, 6, 7, 8, 9, 11, 12, 13], "function": [0, 7, 11, 12, 13, 17], "further": 8, "futur": [5, 7, 8], "futurewarn": [4, 5, 6, 7, 8, 9], "g": [0, 7, 8, 11], "gabitto": [5, 7, 8, 14], "gad1": [4, 9], "gad2": 4, "gdr": [0, 13, 17], "gdr_resolut": 13, "gdrive": [5, 6, 8], "gender": [4, 5, 7, 8, 9], "gene": [0, 5, 9, 11, 13, 14, 17], "gene_depth": 6, "gene_id": [4, 5, 6, 7, 8, 9], "gene_list": 13, "gene_set": 13, "gene_set_nam": 13, "gene_weight": 13, "genecel": [2, 8], "gener": [11, 12], "genom": [4, 5, 6, 7, 8, 9], "genotyp": [4, 5, 7, 8, 9], "get_n_split": 8, "getattr": 9, "getcrosscategori": [3, 4, 10, 12], "getlogg": [5, 7, 8], "gex_barcod": 6, "gex_conf_exonic_antisense_read": 6, "gex_conf_exonic_dup_read": 6, "gex_conf_exonic_read": 6, "gex_conf_exonic_unique_read": 6, "gex_conf_intergenic_read": 6, "gex_conf_intronic_antisense_read": 6, "gex_conf_intronic_dup_read": 6, "gex_conf_intronic_read": 6, "gex_conf_intronic_unique_read": 6, "gex_conf_txomic_unique_read": 6, "gex_estimated_number_of_cel": [4, 5, 7, 8, 9], "gex_exonic_umi": 6, "gex_fraction_of_transcriptomic_reads_in_cel": [4, 5, 7, 8, 9], "gex_genes_count": 6, "gex_intronic_umi": 6, "gex_mapped_read": 6, "gex_mean_raw_reads_per_cel": [4, 5, 7, 8, 9], "gex_median_genes_per_cel": [4, 5, 7, 8, 9], "gex_median_umi_counts_per_cel": [4, 5, 7, 8, 9], "gex_number_of_read": [4, 5, 7, 8, 9], "gex_percent_dupl": [4, 5, 7, 8, 9], "gex_q30_bases_in_barcod": [4, 5, 7, 8, 9], "gex_q30_bases_in_read_2": [4, 5, 7, 8, 9], "gex_q30_bases_in_sample_index_i1": [4, 5, 7, 8, 9], "gex_q30_bases_in_sample_index_i2": [4, 5, 7, 8, 9], "gex_q30_bases_in_umi": [4, 5, 7, 8, 9], "gex_raw_read": 6, "gex_reads_mapped_antisense_to_gen": [4, 5, 7, 8, 9], "gex_reads_mapped_confidently_to_exonic_region": [4, 5, 7, 8, 9], "gex_reads_mapped_confidently_to_genom": [4, 5, 7, 8, 9], "gex_reads_mapped_confidently_to_intergenic_region": [4, 5, 7, 8, 9], "gex_reads_mapped_confidently_to_intronic_region": [4, 5, 7, 8, 9], "gex_reads_mapped_confidently_to_transcriptom": [4, 5, 7, 8, 9], "gex_reads_mapped_to_genom": [4, 5, 7, 8, 9], "gex_reads_with_tso": [4, 5, 7, 8, 9], "gex_sequenced_read_pair": [4, 5, 7, 8, 9], "gex_sequencing_satur": [4, 5, 7, 8, 9], "gex_total_genes_detect": [4, 5, 7, 8, 9], "gex_umis_count": 6, "gex_valid_barcod": [4, 5, 7, 8, 9], "gex_valid_umi": [4, 5, 7, 8, 9], "git": 2, "github": 2, "github_test": [4, 9], "given": [7, 8, 12, 13], "go": 8, "googl": [4, 5, 7, 8, 9], "gradient": 11, "granular": 13, "graph": 7, "great": [4, 5, 7, 8, 9], "greatapesmtg_rna_integrated_raw_onlyorthogonal_gdr_humansubset_aib": 5, "grid": [4, 5, 8, 11], "grossli": [4, 5, 7, 8, 9], "group": [6, 8, 11, 13], "groupbi": [4, 6, 7, 8, 11, 13], "gt": 8, "guid": [8, 13, 17], "h5ad": [4, 5, 6, 7, 8, 9, 13], "ha": 8, "handl": 13, "harmonypi": [5, 8], "have": [2, 4, 8, 9], "haven": 8, "hawaiian": [4, 5, 7, 8, 9], "head": 13, "heatmap": [5, 8], "height": [4, 11], "height_singl": [4, 11], "help": 6, "here": [4, 7, 8], "heterogen": 7, "hex": 11, "hex_color": 11, "hexadecim": 11, "higher": [8, 13], "highest": [4, 5, 7, 8, 9], "highli": [0, 6, 8, 13], "highly_vari": [5, 6, 7, 8], "highly_variable_gen": 8, "highly_variable_infog": 6, "highly_variable_intersect": [6, 8], "highly_variable_nbatch": [6, 8], "highly_variable_rank": [5, 6, 7, 8], "hispan": [4, 5, 7, 8, 9], "home": [4, 5, 6, 7, 8, 9], "horizont": 4, "how": [4, 5, 6, 7, 8], "html": [5, 8], "http": [2, 4, 5, 7, 8, 9, 14], "human": 5, "hvg": [5, 6, 7], "i": [0, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14], "id": [4, 5, 7, 8, 9], "ident": [0, 8, 14], "identif": [13, 14], "identifi": [5, 6, 13], "idx": [7, 8], "ignor": [4, 6, 9, 13], "igraph": [5, 6, 7, 8], "implement": [7, 8], "impli": 8, "import": [4, 5, 6, 7, 8, 9, 11, 13], "importlib": [4, 9], "imshow": 11, "inch": 11, "includ": 0, "independ": 13, "index": [0, 5, 8, 13], "indian": [4, 5, 7, 8, 9], "indic": [11, 13], "infer": [0, 6], "info": [5, 7, 8], "info_factor": 6, "infog": [0, 3, 10, 13], "infog_lay": [6, 13], "infog_scor": 13, "infog_trim": 13, "inform": [6, 13], "initi": [5, 8], "inplac": 8, "input": [8, 13], "instal": [0, 8], "instead": [5, 7, 8, 13], "int": [8, 9, 11, 13], "integ": 11, "integr": [6, 10, 13, 14], "interest": 4, "interneuron": [0, 14], "interv": [4, 5, 7, 8, 9], "intregr": 8, "introduct": [5, 7, 17], "invalid": 8, "io": [5, 8], "iprogress": 8, "ipywidget": 8, "is_cel": 6, "isinst": [7, 8], "island": [4, 5, 7, 8, 9], "isn": 6, "item": [7, 8], "iter": [5, 6, 7, 8, 12], "iterate_by_second_column": 12, "iterrow": [7, 8], "j": [9, 14], "jitter": 11, "join": [6, 12], "jupyt": 8, "k": [7, 8, 14], "kd": 13, "kei": [11, 13], "key_ad": [5, 6, 7, 8, 13], "keyword": 11, "kmean": [5, 8], "knn": [6, 8], "know": [5, 8], "known": [5, 13], "kwarg": 11, "l2": 4, "l4": 4, "l5": 4, "l6": 4, "l6b": 4, "label": [5, 7, 8, 11, 13], "lamp5": [4, 7], "last": [4, 5, 7, 8, 9], "late": [4, 5, 7, 8, 9], "late_color": [4, 5, 7, 8, 9], "latest": [5, 8], "latino": [4, 5, 7, 8, 9], "layer": [4, 5, 6, 7, 8, 9, 11, 13], "layer_refer": [5, 8, 13], "lead": 8, "leaf": 13, "leaf_siz": 13, "learn": [5, 6, 7, 8], "legend": [4, 11], "legend_fontoutlin": [4, 5, 7, 8, 11], "legend_fonts": [4, 5, 7, 8, 11], "legend_fontweight": 4, "legend_loc": [4, 5, 7, 8, 9, 11], "legend_marker_s": 11, "leiden": [5, 6, 8, 13], "leiden_color": [5, 6], "leiden_last": 6, "leiden_last2": 6, "leiden_last2_color": 6, "leiden_last_color": 6, "leiden_loc": [3, 6, 10, 13], "leiden_local_color": 6, "leiden_local_many_al": 7, "leiden_local_many_each": 7, "leiden_local_on": 7, "leiden_sampl": 6, "leidenalg": [7, 8], "len": 9, "length": [9, 13], "less": [7, 8], "let": [7, 8], "level": [4, 5, 7, 8, 9, 13], "lewi": [4, 5, 7, 8, 9], "lhx6": [4, 7], "lib": 8, "librari": 11, "library_input_ng": [4, 5, 7, 8, 9], "library_prep": [4, 5, 7, 8, 9], "library_prep_pass_fail": [4, 5, 7, 8, 9], "limit": 11, "line": [4, 11], "linearsegmentedcolormap": 11, "linspac": 11, "list": [4, 7, 8, 11, 12, 13], "liter": 13, "load": [11, 13], "load_nam": [4, 5, 7, 8, 9], "local": [7, 13], "locat": 11, "log": [5, 6, 7, 13], "log1p": [5, 6, 7, 8, 13], "logarithm": 8, "louvain": [5, 6, 7, 11, 13], "low": [8, 13], "lower": 11, "lowli": 8, "lt": [6, 8], "luecken2021": 1, "lyse": 8, "m": [5, 7, 8, 14], "machin": 8, "made": 7, "mai": 7, "maintain": 4, "make": 8, "malat1": 4, "map": [5, 7, 8, 9], "mapping_dict": 4, "margin": 11, "margin_ratio": 11, "marker": [5, 7, 11, 13, 14, 17], "marker_gen": [8, 13], "marker_gene_n_groups_indic": 13, "marker_genes_list": [7, 8], "match": [5, 8, 13], "matplotlib": [4, 5, 6, 7, 8, 9, 11], "matrix": [6, 7, 8, 11, 13], "max_work": [6, 13], "maximum": [11, 13], "mean": [5, 6, 7, 8, 11, 13], "mean_count": [5, 6, 7], "memeori": 13, "memori": 13, "merfish": 0, "merg": 13, "metadata": [4, 5, 7, 8, 9], "method": [0, 4, 5, 6, 7, 8, 9, 13], "metric": [0, 5, 8, 11], "microglia": 4, "microinfarct": [4, 5, 7, 8, 9], "might": 8, "min_gen": 8, "minimum": 11, "mitochondri": [4, 5, 7, 8, 9], "mmse": [4, 5, 7, 8, 9], "moca": [4, 5, 7, 8, 9], "model": 8, "model_select": 8, "moder": 4, "modifi": [5, 7, 13], "modul": 0, "monitor": 13, "month": [4, 5, 7, 8, 9], "more": [5, 7, 8, 13], "mrna": 8, "mt": [5, 6, 7, 8], "mu": [5, 6, 7, 8, 13], "multi": 13, "multimap": [4, 5, 7, 8, 9], "multimod": 14, "multiom": 6, "multiome_feature_linkages_detect": [4, 5, 7, 8, 9], "multiome_linked_gen": [4, 5, 7, 8, 9], "multiome_linked_peak": [4, 5, 7, 8, 9], "multipl": [0, 4, 11, 13], "multiprocess": 13, "must": [7, 8, 13], "myenv": 2, "n": [4, 5, 6, 7, 9, 11], "n_cell": 6, "n_cells_by_count": [5, 6, 7], "n_compon": [8, 13], "n_ctrl_set": 13, "n_gene": [5, 6, 7, 8, 13], "n_genes_by_count": [5, 6, 7, 8], "n_genes_us": [7, 8], "n_highly_variable_gen": [5, 6, 8, 13], "n_iter": [5, 7, 8], "n_nearest_neighbor": 13, "n_neighbor": [6, 8], "n_ob": [4, 5, 6, 7, 8, 9, 11], "n_split": 8, "n_svd_dim": [5, 6, 8, 13], "n_top_gen": [6, 8, 13], "n_var": [4, 5, 6, 7, 8, 9, 11], "name": [2, 4, 5, 7, 8, 9, 11, 12, 13], "nat": 14, "nativ": [4, 5, 7, 8, 9], "natur": 7, "ncol": [4, 5, 6, 7, 8, 9, 11], "ncount_rna": 5, "ndarrai": 13, "nearest": 13, "neg": 8, "neighbor": [4, 5, 6, 7, 8, 9, 13], "neighborhood": 5, "network": 0, "neun": [4, 5, 7, 8, 9], "neuron": [0, 14], "neuropatholog": [4, 5, 7, 8, 9], "neurosci": 14, "neurotyp": [4, 5, 7, 8, 9], "new": [12, 13], "nfeature_rna": 5, "non": [4, 5, 7, 8, 9], "none": [4, 5, 6, 7, 8, 9, 11, 13], "normal": [0, 5, 7, 13], "normalize_pearson_residu": 8, "normalize_tot": 8, "normalized_cf_matrix_df": [5, 8], "note": [0, 8], "notebook": [4, 9], "notebook_tqdm": 8, "novel": 0, "novo": 13, "now": [5, 8], "np": [4, 5, 6, 7, 8, 9, 11, 13], "nrow": 9, "nuclear_read_pair": [4, 5, 7, 8, 9], "number": [5, 7, 8, 9, 11, 13], "numpi": [4, 5, 6, 7, 8, 9, 11], "o": [5, 6, 7, 8], "ob": [4, 5, 6, 7, 8, 9, 11, 13], "object": [4, 5, 6, 7, 8, 9, 11, 13], "observ": [4, 5, 7, 8, 9, 11, 13], "obsm": [4, 5, 6, 7, 8, 9, 11, 13], "obsp": [4, 5, 6, 7, 8, 9], "obtain": [6, 11], "off": 11, "offer": 9, "oligodendrocyt": 4, "omic": [8, 10], "one": [2, 13], "onli": [6, 13], "onto": 4, "opc": 4, "oper": 13, "option": [5, 6, 7, 8, 9, 11, 12, 13], "order": [4, 12], "org": [4, 5, 7, 8, 9, 14], "organ": [4, 5, 7, 8, 9], "origin": [4, 5, 7, 8, 9], "other": [0, 4, 5, 7, 8, 9], "otherwis": 13, "our": [5, 7, 8], "out": [7, 8], "outlin": 11, "output": 13, "over": 8, "overal": [4, 5, 7, 8, 9], "overdispers": 8, "overwrit": 5, "p": 13, "p57": 6, "pacif": [4, 5, 7, 8, 9], "packag": [3, 4, 5, 6, 7, 8, 9], "page": 0, "pair": [6, 7, 8], "palett": [4, 5, 6, 7, 8], "palette_valu": 9, "pallett": 17, "panda": [5, 6, 7, 8, 12, 13], "panel": 11, "parallel": [6, 13], "parallel_tim": 6, "paramet": [4, 6, 7, 8, 11, 12, 13], "part": 8, "particularli": 5, "pass": [7, 8, 11], "path": [4, 5, 6, 7, 8, 9, 11], "pax6": [4, 7], "pbmc3k": 11, "pcr_cycl": [4, 5, 7, 8, 9], "pct_counts_mt": [5, 6, 7, 8], "pct_counts_ribo": [5, 6, 7, 8], "pct_dropout_by_count": [5, 6, 7], "pd": [5, 6, 7, 8, 12], "pdf": [4, 11], "pdgfra": 4, "pearson": 8, "pei": 14, "per": [4, 8, 11, 13], "percenatg": 8, "percent_cdna_longer_than_400bp": [4, 5, 7, 8, 9], "percent_top": 8, "percentag": 8, "perform": [0, 5, 8, 13], "ph": [4, 5, 7, 8, 9], "piaso": [2, 3, 5, 6, 7, 8, 17], "piaso_test": 8, "piaso_umap_split_by_condit": 4, "pip": [2, 8], "pl": [4, 5, 6, 7, 8, 9, 11], "place": 13, "pleas": [0, 5, 7, 8, 13], "plot": [0, 3, 6, 8, 9, 10, 17], "plot_embeddings_split": [3, 4, 10, 11], "plot_features_violin": [3, 4, 8, 10, 11], "plt": [4, 5, 6, 7, 8, 9, 11], "pmi": [4, 5, 7, 8, 9], "point": 11, "portal": [4, 5, 7, 8, 9], "pos_cmap": [5, 7], "posit": [4, 5, 7, 8, 9], "pp": [4, 6, 8], "pr_offset": 8, "pre": [5, 7, 8], "precis": [8, 10], "predict": [13, 17], "predictcelltypebygdr": [3, 5, 8, 10, 13], "predicted_doublet": 6, "prefix": [4, 9], "preprocess": [0, 3, 5, 7, 10], "presenc": 6, "present": 12, "primari": [4, 5, 7, 8, 9], "princip": 13, "print": [6, 13], "print_head": [5, 6, 7, 8], "prior": 8, "proceed": 8, "process": [5, 7, 8, 13], "produc": [11, 13], "progress": [4, 5, 7, 8, 9, 13], "project": [4, 5, 7, 8, 9], "proportion": [0, 14], "protein": 8, "provid": [11, 13], "pseudo": [4, 5, 7, 8, 9], "pt": 11, "public": [4, 5, 6, 7, 8, 9], "purpl": [5, 8], "purpos": 8, "pvalb": 4, "pvm": 4, "py": [5, 6, 7, 8], "pynndesc": [5, 6, 7, 8], "pyplot": [4, 5, 6, 7, 8, 9, 11], "pyramid": [0, 14], "python": [4, 5, 6, 7, 8, 9], "python3": 8, "qc_var": 8, "qualiti": 8, "quantification_fmol": [4, 5, 7, 8, 9], "queri": [8, 13], "query_data": 13, "query_groupbi": [5, 8, 13], "r1_index": [4, 5, 7, 8, 9], "race": [4, 5, 7, 8, 9], "rachleff": 14, "random": 13, "random_se": 13, "random_st": [6, 8, 13], "randomli": 13, "rang": [9, 11], "rank": [4, 12], "ratio": 11, "ravel": [6, 8], "raw": [5, 6, 7, 8, 11], "rc": [5, 8], "rcparam": [4, 5, 6, 7, 8, 9], "re": 8, "read": [4, 5, 6, 7, 8, 9], "read_h5ad": 13, "readthedoc": [5, 8], "real": [5, 8], "reason": 8, "reduc": 8, "reduct": [0, 7, 13, 17], "ref_adata": 5, "refer": [0, 4, 7, 8, 9, 13], "reference_data": 13, "reference_groupbi": [5, 8, 13], "reflect": 8, "region": [4, 5, 7, 8, 9], "regular": 13, "regulatori": 0, "reindex": [7, 8], "rel": 11, "releas": 0, "reli": 0, "remain": 8, "remov": [5, 8], "remove_lowly_express": [7, 8], "reorder": 5, "reorder_categori": [5, 8], "replac": 2, "repres": 13, "represent": 13, "reproduc": 13, "requir": 6, "rerun": [7, 13], "research": 0, "reshap": 11, "residu": 8, "residual_vari": [5, 6, 7, 8], "resolut": [5, 6, 7, 8, 13], "respect": 12, "restrict_to": 7, "result": [4, 9, 11, 12, 13], "retain": 13, "return": [11, 12, 13], "return_gene_nam": 13, "return_integr": [5, 8, 13], "reveal": 7, "rgb": 11, "ribo": 8, "ribo_cel": 8, "ribosom": 8, "right": 11, "rin": [4, 5, 7, 8, 9], "rna": [6, 8], "rna_amplif": [4, 5, 7, 8, 9], "rna_amplification_pass_fail": [4, 5, 7, 8, 9], "rnaseq": 5, "robust": 13, "root": 11, "row": [4, 7, 8, 11], "row_siz": [9, 11], "rp": 8, "rpl": 8, "run": [0, 5, 6, 13], "runcosgparallel": [3, 10, 13], "rungdr": [3, 6, 8, 10, 13], "rungdrparallel": [3, 6, 10, 13], "runsvd": [3, 8, 10, 13], "runsvdlazi": [3, 10, 13], "runtimewarn": [6, 8], "s41593": 14, "same": 9, "sampl": [4, 5, 6], "sample_color": [5, 6], "sample_id": [4, 5, 7, 8, 9], "sample_nam": [4, 5, 7, 8, 9], "sample_quantity_count": [4, 5, 7, 8, 9], "san": [5, 6, 7, 8], "satb2": 4, "save": [5, 6, 8, 11, 13], "save_dir": [4, 9], "sb": 6, "sbuimehrvvhnf7tfu7h6xi": [5, 7], "sc": [4, 5, 6, 7, 8, 9, 11, 13], "scale": 11, "scale_data": 13, "scanpi": [4, 5, 6, 7, 8, 9, 11, 13], "scatac": 0, "scib_metr": 6, "scikit": [5, 6, 7, 8], "scipi": [4, 5, 6, 7, 8, 9], "score": [0, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13], "score_color": [4, 5, 7, 8, 9], "score_lay": [6, 8, 13], "score_list": 13, "score_method": 13, "scoring_method": [5, 6, 8, 13], "scratch": [4, 5, 6, 7, 9], "screen": [4, 5, 7, 8, 9], "scrna": [0, 5, 7, 8], "scrublet_cluster_scor": 6, "scrublet_scor": 6, "sea": [4, 5, 7, 8, 9], "seaborn": [5, 6, 7, 8], "search": 0, "seattl": [4, 5, 7, 8, 9], "second": 12, "secondari": [4, 5, 7, 8, 9], "section": [4, 5, 7, 8, 9], "see": [5, 8, 13], "seed": 13, "select": [0, 8, 9, 13, 17], "separ": [6, 11, 13], "seq": [0, 4, 5, 7, 8, 9], "sequenc": [11, 13], "seri": 12, "serif": [5, 6, 7, 8], "serpine1": 4, "set": [0, 4, 5, 6, 7, 8, 9, 11, 12, 13], "set_figure_param": [4, 5, 6, 7, 8, 9], "set_styl": [5, 8], "set_vis": 9, "sever": [4, 5, 7, 8, 9], "sex": [4, 5, 7, 8, 9], "sex_color": [4, 5, 7, 8, 9], "shape": 11, "share": 13, "should": 11, "show": [4, 5, 6, 8, 9, 11], "show_axis_tick": 11, "show_figur": 11, "show_grid": [4, 11], "shown": 11, "side": [0, 11], "similar": [7, 13], "simplefilt": [4, 6, 9], "sinc": [5, 8], "singl": [4, 8, 9, 10, 13], "singular": 13, "site": 8, "size": [4, 5, 6, 7, 8, 9, 11, 13], "sklearn": [5, 8], "smaller": 9, "sn": [5, 6, 7, 8], "sncg": [4, 7], "snrna": [4, 9], "so": [7, 11], "softwar": [5, 6, 8], "some": [8, 9], "sort": 12, "space": [8, 11], "spars": [4, 6, 9], "speci": 5, "species_tech": 5, "species_tech_color": 5, "specif": [0, 13, 14], "specifi": [4, 5, 7, 8, 9, 11, 12, 13], "spectral_r": [4, 7, 8, 9], "split": [0, 8, 11], "splitbi": [4, 11], "sqrt": [8, 9], "squar": 11, "sst": 4, "stabl": 8, "stack": 0, "stand": 8, "standard": [6, 13], "standard_scal": [7, 8], "start": 8, "start_tim": 6, "startswith": [8, 9], "statsmodel": [5, 6, 7, 8], "statu": [4, 5, 7, 8, 9], "status_color": [4, 5, 7, 8, 9], "std": 8, "step": 8, "store": [4, 7, 8, 13], "str": [8, 9, 11, 12, 13], "stratifi": 8, "stratified_split": 8, "stratifiedshufflesplit": 8, "string": 11, "studi": [4, 5, 7, 8, 9], "sub": 7, "subclass": [5, 7, 8, 9], "subclass_color": [4, 5, 7, 8, 9], "subclass_df": 4, "subclassxcondit": 4, "sublist": 13, "subpackag": 3, "subplot": [4, 9, 11], "subsampl": [4, 5, 7, 8, 9], "subset": 5, "subtyp": [0, 14], "successfulli": [6, 7], "suffix": 13, "suggest": 7, "sum": [5, 8], "supertyp": [4, 5, 7, 8, 9], "supertype_color": [4, 5, 7, 8, 9], "support": 13, "suptitl": 9, "sure": 8, "surviv": [0, 14], "svd": [8, 13], "swap_ax": [4, 7, 8], "switch": 4, "sy": [4, 5, 6, 7, 8, 9], "t": [2, 6, 7, 8], "tabl": [3, 4, 10, 12], "target_sum": 8, "tbr1": 4, "tech": 5, "test": [5, 8], "test_index": 8, "test_siz": 8, "thal": [4, 5, 7, 8, 9], "thal_color": [4, 5, 7, 8, 9], "than": [7, 8, 9], "thei": 7, "them": 5, "therefor": [5, 8], "theta": 8, "thi": [4, 5, 6, 7, 8, 9, 11, 13], "through": 8, "tick": 11, "time": [5, 6, 7, 8], "titl": [4, 5, 7, 8, 9], "tl": [5, 6, 7, 8, 13], "togeth": [4, 7], "tool": [2, 3, 5, 6, 7, 8, 10], "top": [8, 13], "total": [4, 5, 6, 7, 8, 9], "total_count": [5, 6, 7, 8], "total_counts_mt": [5, 6, 7], "tqdm": [5, 7, 8], "tqdmwarn": 8, "train": 8, "train_index": 8, "transform": 13, "travaglini": 14, "tree": 13, "trim": 13, "true": [4, 5, 6, 7, 8, 11, 12, 13], "truncat": 13, "tupl": 13, "turori": 6, "tutori": [0, 5, 7, 8], "two": 12, "type": [0, 4, 6, 7, 11, 12, 13, 17], "u": [7, 8], "umap": [8, 9], "umap_1": 5, "umap_2": 5, "umi": [5, 7, 8, 9, 13], "un": [4, 5, 6, 7, 8, 9, 13], "under": [11, 13], "union": [5, 6, 7, 8], "uniqu": [11, 12], "unknown": [4, 5, 7, 8, 9], "unmap": [4, 5, 7, 8, 9], "unreport": [4, 5, 7, 8, 9], "upcom": 0, "updat": [8, 13], "upper": 11, "us": [0, 4, 5, 6, 7, 8, 9, 11, 12, 13], "usag": 13, "use_highly_vari": [5, 6, 8, 13], "use_raw": [7, 8, 11], "use_rep": [6, 7, 8], "user": [4, 5, 6, 7, 8, 9], "user_instal": 8, "usp": [4, 5, 7, 8, 9], "usual": 8, "uw": [4, 5, 7, 8, 9], "v": [4, 5, 6, 7, 8, 9, 14], "vallari": [4, 8, 9], "valu": [5, 7, 8, 11, 12, 13], "var": [4, 5, 6, 7, 8, 9], "var_nam": [4, 5, 8], "vari": 4, "variabl": [0, 6, 8, 11, 13], "varianc": [5, 6, 7, 8, 13], "variat": 13, "variou": 4, "varm": 6, "vas744": [4, 5, 6, 7, 8, 9], "vector": 8, "verbos": [5, 6, 7, 8, 13], "verifi": 5, "version": [5, 7, 8], "view": [4, 5, 7, 8, 9], "violin": [0, 8, 11], "violin_plot_by_piaso": 11, "violin_plot_piaso": 4, "vip": [4, 7], "viridi": [4, 5, 6, 7, 8, 9], "visual": 11, "vlmc": 4, "vmax": 11, "vmin": 11, "vv": 8, "wa": [5, 7], "wall": [5, 6, 7, 8], "wang": 14, "want": [7, 8], "warn": [4, 6, 9], "we": [4, 5, 6, 7, 8], "weight": [4, 5, 6, 7, 8, 9, 13], "well": [8, 13], "were": [5, 6, 8], "when": [6, 12, 13], "where": [11, 13], "whether": [11, 13], "which": [4, 7, 8, 11, 12, 13], "while": 4, "white": [4, 5, 6, 7, 8, 9], "whitegrid": [5, 8], "whole": 7, "widht": 4, "width": 11, "width_singl": [4, 11], "within": 13, "without": 6, "work": [6, 7, 8], "worker": 13, "workflow": 8, "wu": [0, 14], "www": 8, "x": [5, 6, 7, 8, 11, 13, 14], "x_gdr": [5, 6, 8, 13], "x_gdr_harmoni": 5, "x_max": 11, "x_min": 11, "x_normal": [4, 5, 7, 8, 9], "x_pca": [6, 11, 13], "x_pca_harmoni": [6, 13], "x_scvi": [4, 5, 7, 8, 9], "x_svd": [5, 7, 8, 13], "x_svd_full": [7, 13], "x_svd_full_harmoni": 13, "x_umap": [4, 5, 6, 7, 8, 9, 11], "x_umap_backup": [7, 8], "x_umap_gdr": 5, "x_umap_gdr_harmoni": 5, "x_umap_pca": 6, "x_umap_publish": 5, "x_umap_raw": [6, 8], "x_umap_svd": 8, "xenium": 0, "xlabel": 5, "xticklabel": [5, 8], "y": [8, 11], "y_max": 11, "y_min": 11, "year": [4, 5, 7, 8, 9], "yet": [2, 8], "ylabel": 5, "you": [2, 4, 7, 8], "your": [0, 2, 8], "yticklabel": [5, 8], "zero": [6, 8], "zip": 4}, "titles": ["PIASO: Precise Integrative Analysis of Single-cell Omics", "References", "Installation", "API", "Plotting functions in PIASO", "Cell type prediction by GDR", "GDR: marker Gene-guided Dimensionality Reduction", "Finer clustering on selected cluster(s)", "Introduction", "Color pallettes in PIASO", "piaso package", "piaso.plotting package", "piaso.preprocessing package", "piaso.tools package", "References", "Release notes", "setup module", "Tutorials"], "titleterms": {"": 7, "0": 15, "1": 15, "2025": 15, "24": 15, "across": [4, 7], "algorithm": 7, "all": 7, "analysi": 0, "api": 3, "base": 6, "categor": 4, "categori": 4, "cell": [0, 4, 5, 8], "citat": 0, "cluster": [5, 7, 8], "color": [4, 9], "combin": 4, "condit": 4, "continu": [4, 9], "count": 4, "data": [4, 5, 6, 7, 8, 9], "dataset": 5, "develop": 2, "differ": 4, "dimension": [6, 8], "discret": [4, 9], "each": [4, 7], "entir": 4, "everi": 8, "experiment": 4, "express": 4, "februari": 15, "finer": 7, "function": 4, "gdr": [5, 6, 8], "gdrparallel": 6, "gene": [4, 6, 7, 8], "group": [4, 7], "guid": 6, "identifi": [7, 8], "independ": 7, "indic": 0, "infog": 6, "instal": 2, "integr": 0, "introduct": 8, "leiden": 7, "leiden_loc": 7, "load": [4, 5, 6, 7, 8, 9], "log": 8, "map": 4, "marker": [6, 8], "markger": 7, "modul": 16, "multipl": 7, "normal": [6, 8], "note": 15, "number": 4, "omic": 0, "one": 7, "output": 4, "packag": [10, 11, 12, 13], "palett": 9, "pallett": 9, "pca": 6, "perform": 7, "piaso": [0, 4, 9, 10, 11, 12, 13], "plot": [4, 11], "precis": 0, "predict": [5, 8], "preprocess": 12, "queri": 5, "reduct": [6, 8], "refer": [1, 5, 14], "releas": 15, "result": 6, "run": [7, 8], "save": 4, "select": 7, "setup": 16, "singl": 0, "specif": 4, "split": 4, "stabl": 2, "subclass": 4, "subclust": 7, "subpackag": 10, "tabl": 0, "tool": 13, "transform": 8, "tutori": 17, "two": 4, "type": [5, 8], "umap": [4, 5, 6, 7], "umi": 4, "variabl": 4, "version": 2, "violin": 4, "visual": [4, 5, 6, 7, 8], "visula": 4, "within": 7}})